{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import scipy\n",
    "import matplotlib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from collections import defaultdict\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "\n",
    "from subprocess import call\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set_style('whitegrid')\n",
    "pd.set_option('display.max_rows', 100)\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "pd.set_option('display.max_columns', 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def SRA_to_bioproject(accession):\n",
    "    '''\n",
    "    From an SRA (like SRR1883283) return a biosample\n",
    "    '''\n",
    "    from Bio import Entrez\n",
    "    Entrez.email = \"mattolm@gmail.com\"\n",
    "\n",
    "    try:\n",
    "        # GO FROM SRA TO ID\n",
    "        #accession = 'SRX1883283'\n",
    "        handle = Entrez.esearch(db='sra', term=accession, retmode='text')\n",
    "        record = Entrez.read(handle)\n",
    "        handle.close()\n",
    "        ID = record['IdList'][0]\n",
    "        #return ID\n",
    "\n",
    "        # ELINK FROM SRA ID TO BIOSAMPLE ID\n",
    "        accession = ID\n",
    "        handle = Entrez.elink(dbfrom=\"sra\", retmax=10, id=accession, linkname=\"sra_bioproject\")\n",
    "        record = Entrez.read(handle)\n",
    "        handle.close()\n",
    "        BID = [link[\"Id\"] for link in record[0][\"LinkSetDb\"][0][\"Link\"]][0]\n",
    "        #return BID\n",
    "\n",
    "        # SUMMARY FROM BIOSAMPLE ID TO TERM\n",
    "        handle = Entrez.esummary(db=\"bioproject\", id=BID)\n",
    "        record = Entrez.read(handle)\n",
    "        handle.close()\n",
    "        BioSample = record['DocumentSummarySet']['DocumentSummary'][0]['Project_Acc']\n",
    "\n",
    "        return BioSample\n",
    "    \n",
    "    except:\n",
    "        return 'Fail'\n",
    "    \n",
    "def SRA_to_library(accession):\n",
    "    '''\n",
    "    From an SRA (like SRR1883283) return a biosample\n",
    "    '''\n",
    "    from Bio import Entrez\n",
    "    Entrez.email = \"mattolm@gmail.com\"\n",
    "\n",
    "    lib = 'UNK'\n",
    "    handle = Entrez.efetch(db=\"sra\", id=accession, retmode='text')\n",
    "    for line in handle.readlines():\n",
    "        line = line.strip()\n",
    "        if 'LIBRARY_LAYOUT' in line:\n",
    "            lib = line.split('LIBRARY_LAYOUT')[1][2:].split('/')[0].split()[0]\n",
    "    return lib\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Link</th>\n",
       "      <th>Group</th>\n",
       "      <th>Center_ID</th>\n",
       "      <th>Method</th>\n",
       "      <th>IndividualLabels?</th>\n",
       "      <th>Publication link</th>\n",
       "      <th>Publication title</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA615032</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>Icahn School of Medicine at Mount Sina</td>\n",
       "      <td>Icahn_NY</td>\n",
       "      <td>IRRELEVANT</td>\n",
       "      <td>No</td>\n",
       "      <td>https://www.biorxiv.org/content/10.1101/2020.0...</td>\n",
       "      <td>SARS-CoV-2 launches a unique transcriptional s...</td>\n",
       "      <td>Evaluating lots of things; worth following up....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA610428</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>UNIVERSITY OF WASHINGTON</td>\n",
       "      <td>UW_WA</td>\n",
       "      <td>IRRELEVANT</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>The are isolating the virus first using human ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA613958</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>The Peter Doherty Institute for Infection and ...</td>\n",
       "      <td>Doherty_Melbourne</td>\n",
       "      <td>PCR_ARCTIC</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA614546</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>Paragon Genomics</td>\n",
       "      <td>Paragon_CA</td>\n",
       "      <td>Mixed</td>\n",
       "      <td>No</td>\n",
       "      <td>https://www.biorxiv.org/content/10.1101/2020.0...</td>\n",
       "      <td>High sensitivity detection of coronavirus SARS...</td>\n",
       "      <td>Evaluating lots of things; worth following up....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA616446</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>Hubei Provincial Center for Disease Control an...</td>\n",
       "      <td>Hubei_China</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>No</td>\n",
       "      <td>https://www.biorxiv.org/content/10.1101/2020.0...</td>\n",
       "      <td>Genome-wide data inferring the evolution and p...</td>\n",
       "      <td>Sequencing genomes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    BioProject                                               Link  \\\n",
       "0  PRJNA615032  https://www.ncbi.nlm.nih.gov/bioproject/?term=...   \n",
       "1  PRJNA610428  https://www.ncbi.nlm.nih.gov/bioproject/?term=...   \n",
       "2  PRJNA613958  https://www.ncbi.nlm.nih.gov/bioproject/?term=...   \n",
       "3  PRJNA614546  https://www.ncbi.nlm.nih.gov/bioproject/?term=...   \n",
       "4  PRJNA616446  https://www.ncbi.nlm.nih.gov/bioproject/?term=...   \n",
       "\n",
       "                                               Group          Center_ID  \\\n",
       "0             Icahn School of Medicine at Mount Sina           Icahn_NY   \n",
       "1                           UNIVERSITY OF WASHINGTON              UW_WA   \n",
       "2  The Peter Doherty Institute for Infection and ...  Doherty_Melbourne   \n",
       "3                                   Paragon Genomics         Paragon_CA   \n",
       "4  Hubei Provincial Center for Disease Control an...        Hubei_China   \n",
       "\n",
       "       Method IndividualLabels?  \\\n",
       "0  IRRELEVANT                No   \n",
       "1  IRRELEVANT                No   \n",
       "2  PCR_ARCTIC                No   \n",
       "3       Mixed                No   \n",
       "4     RNA-seq                No   \n",
       "\n",
       "                                    Publication link  \\\n",
       "0  https://www.biorxiv.org/content/10.1101/2020.0...   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3  https://www.biorxiv.org/content/10.1101/2020.0...   \n",
       "4  https://www.biorxiv.org/content/10.1101/2020.0...   \n",
       "\n",
       "                                   Publication title  \\\n",
       "0  SARS-CoV-2 launches a unique transcriptional s...   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3  High sensitivity detection of coronavirus SARS...   \n",
       "4  Genome-wide data inferring the evolution and p...   \n",
       "\n",
       "                                         Description  \n",
       "0  Evaluating lots of things; worth following up....  \n",
       "1  The are isolating the virus first using human ...  \n",
       "2                                                NaN  \n",
       "3  Evaluating lots of things; worth following up....  \n",
       "4                                 Sequencing genomes  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val_loc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/validated_downloads.txt'\n",
    "meta_loc = '/home/mattolm/user_data/Covid_19/Pipeline/Metadata/metadata.xlsx'\n",
    "\n",
    "THREADS = 6\n",
    "\n",
    "VALIDATED_RUNS = []\n",
    "with open(val_loc, 'r') as o:\n",
    "    for line in o.readlines():\n",
    "        VALIDATED_RUNS.append(line.strip())\n",
    "VALIDATED_RUNS = set(VALIDATED_RUNS)\n",
    "\n",
    "BioMeta = pd.read_excel(meta_loc, sheet_name='BioProjects')\n",
    "RunMeta = pd.read_excel(meta_loc, sheet_name='IndividualSamples')\n",
    "BioMeta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get a list of all SRA samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/mattolm/miniconda3/lib/python3.7/site-packages/tqdm/std.py:658: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version\n",
      "  from pandas import Panel\n",
      "/home/mattolm/miniconda3/lib/python3.7/site-packages/tqdm/std.py:658: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version\n",
      "  from pandas import Panel\n"
     ]
    }
   ],
   "source": [
    "from datetime import date\n",
    "today = date.today()\n",
    "d4 = today.strftime(\"%m%d%Y\")\n",
    "dloc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA.tsv'.format(d4)\n",
    "\n",
    "SEARCH_STRING = '(\\\\\"Severe acute respiratory syndrome coronavirus 2\\\\\"[Organism] OR SARS-CoV-2[All Fields]) AND \\\\\"platform illumina\\\\\"[Properties]'\n",
    "\n",
    "! pysradb search \"$SEARCH_STRING\" --detailed --saveto \"$dloc\"\n",
    "\n",
    "SRdb = pd.read_csv(dloc, sep='\\t')\n",
    "SRdb['Run'] = SRdb['run_accession']\n",
    "SRdb = SRdb[~SRdb['run_accession'].isna()]\n",
    "\n",
    "# Add BioProject and layout\n",
    "SRdb['BioProject'] = [SRA_to_bioproject(r) for r in SRdb['Run']]\n",
    "SRdb['TrueLibraryLayout'] = [SRA_to_library(r) for r in SRdb['Run']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4 new BioProjects with no info\n",
      "PRJNA624358\n",
      "Fail\n",
      "PRJNA624792\n",
      "PRJNA616147\n"
     ]
    }
   ],
   "source": [
    "# Any unknown BioProjects?\n",
    "unk = set(SRdb['BioProject'].tolist()) - set(BioMeta['BioProject'].tolist())\n",
    "if len(unk) == 0:\n",
    "    print(\"All BioProjects are known\")\n",
    "else:\n",
    "    print(\"{0} new BioProjects with no info\".format(len(unk)))\n",
    "    print(\"\\n\".join(list(unk)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>experiment_desc</th>\n",
       "      <th>organism_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>sample_title_x</th>\n",
       "      <th>instrument</th>\n",
       "      <th>total_spots</th>\n",
       "      <th>total_size</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>run_total_spots</th>\n",
       "      <th>run_total_bases</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>sra_url_alt</th>\n",
       "      <th>sra_url</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>isolate</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>host</th>\n",
       "      <th>host_disease</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>lat_lon</th>\n",
       "      <th>BioSampleModel</th>\n",
       "      <th>passage_history</th>\n",
       "      <th>strain</th>\n",
       "      <th>source_name</th>\n",
       "      <th>subject status</th>\n",
       "      <th>tissue/cell type</th>\n",
       "      <th>treatment</th>\n",
       "      <th>time after treatment</th>\n",
       "      <th>sub_species</th>\n",
       "      <th>cell line</th>\n",
       "      <th>description</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>host_taxid</th>\n",
       "      <th>isol_growth_condt</th>\n",
       "      <th>propagation</th>\n",
       "      <th>sample_title_y</th>\n",
       "      <th>seq_methods</th>\n",
       "      <th>source_uvig</th>\n",
       "      <th>virus_enrich_appr</th>\n",
       "      <th>culture_collection</th>\n",
       "      <th>genotype</th>\n",
       "      <th>host_age</th>\n",
       "      <th>host_description</th>\n",
       "      <th>host_disease_outcome</th>\n",
       "      <th>host_disease_stage</th>\n",
       "      <th>host_health_state</th>\n",
       "      <th>host_sex</th>\n",
       "      <th>host_subject_id</th>\n",
       "      <th>host_tissue_sampled</th>\n",
       "      <th>pathotype</th>\n",
       "      <th>serotype</th>\n",
       "      <th>serovar</th>\n",
       "      <th>specimen_voucher</th>\n",
       "      <th>subgroup</th>\n",
       "      <th>subtype</th>\n",
       "      <th>cell type</th>\n",
       "      <th>time point</th>\n",
       "      <th>identification_method</th>\n",
       "      <th>sample type</th>\n",
       "      <th>Laboratory Host</th>\n",
       "      <th>Extraction Method</th>\n",
       "      <th>ref_biomaterial</th>\n",
       "      <th>link_addit_analys</th>\n",
       "      <th>country</th>\n",
       "      <th>sra_url_alt1</th>\n",
       "      <th>sra_url_alt2</th>\n",
       "      <th>ena_fastq_url</th>\n",
       "      <th>ena_fastq_ftp</th>\n",
       "      <th>Run</th>\n",
       "      <th>BioProject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRP255993</td>\n",
       "      <td>SRX8095881</td>\n",
       "      <td>WGS of SARS-CoV-2 R03006_2020_2</td>\n",
       "      <td>WGS of SARS-CoV-2 R03006_2020_2</td>\n",
       "      <td>2697049.0</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS6462475</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2971431.0</td>\n",
       "      <td>1.079579e+09</td>\n",
       "      <td>SRR11524818</td>\n",
       "      <td>2971431.0</td>\n",
       "      <td>1.787565e+09</td>\n",
       "      <td>R03006_2_2020_S2_L001_R1_001.fastq.gz</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>National Institute for Communicable Diseases, ...</td>\n",
       "      <td>2020-03-07</td>\n",
       "      <td>South Africa: KwaZulu-Natal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Combined nasopharyngeal and oropharyngeal swab</td>\n",
       "      <td>28.5306 S 30.8958 E</td>\n",
       "      <td>Pathogen.cl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R03006_2020_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR11524818</td>\n",
       "      <td>PRJNA624358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRP255993</td>\n",
       "      <td>SRX8095880</td>\n",
       "      <td>WGS of SARS-CoV-2 R03006_2020_1</td>\n",
       "      <td>WGS of SARS-CoV-2 R03006_2020_1</td>\n",
       "      <td>2697049.0</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS6462474</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>1731908.0</td>\n",
       "      <td>6.862530e+08</td>\n",
       "      <td>SRR11524819</td>\n",
       "      <td>1731908.0</td>\n",
       "      <td>1.042411e+09</td>\n",
       "      <td>R03006_1_2020_S2_L001_R1_001.fastq.gz</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>National Institute for Communicable Diseases, ...</td>\n",
       "      <td>2020-03-07</td>\n",
       "      <td>South Africa: KwaZulu-Natal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Combined nasopharyngeal and oropharyngeal swab</td>\n",
       "      <td>28.5306 S 30.8958 E</td>\n",
       "      <td>Pathogen.cl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R03006_2020_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR11524819</td>\n",
       "      <td>PRJNA624358</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession experiment_accession                 experiment_title  \\\n",
       "6       SRP255993           SRX8095881  WGS of SARS-CoV-2 R03006_2020_2   \n",
       "7       SRP255993           SRX8095880  WGS of SARS-CoV-2 R03006_2020_1   \n",
       "\n",
       "                   experiment_desc  organism_taxid   \\\n",
       "6  WGS of SARS-CoV-2 R03006_2020_2        2697049.0   \n",
       "7  WGS of SARS-CoV-2 R03006_2020_1        2697049.0   \n",
       "\n",
       "                                     organism_name library_strategy  \\\n",
       "6  Severe acute respiratory syndrome coronavirus 2              WGS   \n",
       "7  Severe acute respiratory syndrome coronavirus 2              WGS   \n",
       "\n",
       "  library_source library_selection sample_accession  sample_title_x  \\\n",
       "6        GENOMIC            RANDOM       SRS6462475             NaN   \n",
       "7        GENOMIC            RANDOM       SRS6462474             NaN   \n",
       "\n",
       "       instrument  total_spots    total_size run_accession  run_total_spots  \\\n",
       "6  Illumina MiSeq    2971431.0  1.079579e+09   SRR11524818        2971431.0   \n",
       "7  Illumina MiSeq    1731908.0  6.862530e+08   SRR11524819        1731908.0   \n",
       "\n",
       "   run_total_bases                              run_alias  \\\n",
       "6     1.787565e+09  R03006_2_2020_S2_L001_R1_001.fastq.gz   \n",
       "7     1.042411e+09  R03006_1_2020_S2_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                         sra_url_alt  \\\n",
       "6  https://sra-download.ncbi.nlm.nih.gov/traces/s...   \n",
       "7  https://sra-download.ncbi.nlm.nih.gov/traces/s...   \n",
       "\n",
       "                                             sra_url experiment_alias isolate  \\\n",
       "6  https://sra-download.ncbi.nlm.nih.gov/traces/s...              NaN     NaN   \n",
       "7  https://sra-download.ncbi.nlm.nih.gov/traces/s...              NaN     NaN   \n",
       "\n",
       "                                        collected_by collection_date  \\\n",
       "6  National Institute for Communicable Diseases, ...      2020-03-07   \n",
       "7  National Institute for Communicable Diseases, ...      2020-03-07   \n",
       "\n",
       "                  geo_loc_name          host host_disease  \\\n",
       "6  South Africa: KwaZulu-Natal  Homo sapiens     COVID-19   \n",
       "7  South Africa: KwaZulu-Natal  Homo sapiens     COVID-19   \n",
       "\n",
       "                                 isolation_source              lat_lon  \\\n",
       "6  Combined nasopharyngeal and oropharyngeal swab  28.5306 S 30.8958 E   \n",
       "7  Combined nasopharyngeal and oropharyngeal swab  28.5306 S 30.8958 E   \n",
       "\n",
       "  BioSampleModel passage_history         strain source_name subject status  \\\n",
       "6    Pathogen.cl             NaN  R03006_2020_2         NaN            NaN   \n",
       "7    Pathogen.cl             NaN  R03006_2020_1         NaN            NaN   \n",
       "\n",
       "  tissue/cell type treatment time after treatment sub_species cell line  \\\n",
       "6              NaN       NaN                  NaN         NaN       NaN   \n",
       "7              NaN       NaN                  NaN         NaN       NaN   \n",
       "\n",
       "  description env_broad_scale env_local_scale env_medium  host_taxid  \\\n",
       "6         NaN             NaN             NaN        NaN         NaN   \n",
       "7         NaN             NaN             NaN        NaN         NaN   \n",
       "\n",
       "  isol_growth_condt propagation sample_title_y seq_methods source_uvig  \\\n",
       "6               NaN         NaN            NaN         NaN         NaN   \n",
       "7               NaN         NaN            NaN         NaN         NaN   \n",
       "\n",
       "  virus_enrich_appr culture_collection genotype host_age host_description  \\\n",
       "6               NaN                NaN      NaN      NaN              NaN   \n",
       "7               NaN                NaN      NaN      NaN              NaN   \n",
       "\n",
       "  host_disease_outcome host_disease_stage host_health_state host_sex  \\\n",
       "6                  NaN                NaN               NaN      NaN   \n",
       "7                  NaN                NaN               NaN      NaN   \n",
       "\n",
       "  host_subject_id host_tissue_sampled pathotype serotype serovar  \\\n",
       "6             NaN                 NaN       NaN      NaN     NaN   \n",
       "7             NaN                 NaN       NaN      NaN     NaN   \n",
       "\n",
       "  specimen_voucher subgroup subtype cell type time point  \\\n",
       "6              NaN      NaN     NaN       NaN        NaN   \n",
       "7              NaN      NaN     NaN       NaN        NaN   \n",
       "\n",
       "  identification_method sample type Laboratory Host Extraction Method  \\\n",
       "6                   NaN         NaN             NaN               NaN   \n",
       "7                   NaN         NaN             NaN               NaN   \n",
       "\n",
       "  ref_biomaterial link_addit_analys country sra_url_alt1 sra_url_alt2  \\\n",
       "6             NaN               NaN     NaN          NaN          NaN   \n",
       "7             NaN               NaN     NaN          NaN          NaN   \n",
       "\n",
       "  ena_fastq_url  ena_fastq_ftp          Run   BioProject  \n",
       "6           NaN            NaN  SRR11524818  PRJNA624358  \n",
       "7           NaN            NaN  SRR11524819  PRJNA624358  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SRdb[SRdb['BioProject'] == 'PRJNA624358']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# d4 = '04202020'\n",
    "# dloc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA.tsv'.format(d4)\n",
    "# SRdb = pd.read_csv(dloc, sep='\\t')\n",
    "# SRdb['Run'] = SRdb['run_accession']\n",
    "# SRdb = SRdb[~SRdb['run_accession'].isna()]\n",
    "\n",
    "# # Add BioProject\n",
    "# SRdb['BioProject'] = [SRA_to_bioproject(r) for r in SRdb['Run']]\n",
    "# SRdb['TrueLibraryLayout'] = [SRA_to_library(r) for r in SRdb['Run']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All BioProjects are known\n"
     ]
    }
   ],
   "source": [
    "# Fix that one\n",
    "SRdb['BioProject'] = ['PRJNA614995' if ((c == 'Utah Public Health Laboratory') & (b == 'Fail')) else b for b, c in zip(SRdb['BioProject'], SRdb['collected_by'])]\n",
    "\n",
    "# Any unknown BioProjects?\n",
    "unk = set(SRdb['BioProject'].tolist()) - set(BioMeta['BioProject'].tolist())\n",
    "if len(unk) == 0:\n",
    "    print(\"All BioProjects are known\")\n",
    "else:\n",
    "    print(\"{0} new BioProjects with no info\".format(len(unk)))\n",
    "    print(\"\\n\".join(list(unk)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 new samples (and 773 already processed = 459 total (including irrelevant already processed))\n"
     ]
    }
   ],
   "source": [
    "# Filter out irrelevant bioprojects\n",
    "SRdb = SRdb[~SRdb['BioProject'].isin(BioMeta[BioMeta['Method'] == 'IRRELEVANT']['BioProject'])]\n",
    "NEW = set(SRdb['Run'].tolist())\n",
    "print(\"{0} new samples (and {1} already processed = {2} total (including irrelevant already processed))\".format(len(NEW - VALIDATED_RUNS), len(VALIDATED_RUNS), len(SRdb)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PAIRED    453\n",
      "SINGLE      6\n",
      "Name: TrueLibraryLayout, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(SRdb['TrueLibraryLayout'].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# START PROCESSING!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save preliminary list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SRdb.to_csv('/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA_filtered.csv'.format(d4), index=False)\n",
    "d4 = '04202020'\n",
    "SRdb = pd.read_csv('/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA_filtered.csv'.format(d4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download samples that need downloading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "def check_sra_file(file):\n",
    "    cmd = \"/home/mattolm/miniconda3/bin/vdb-validate {0}\".format(file)\n",
    "    try:\n",
    "        output = subprocess.check_output(\n",
    "            cmd, stderr=subprocess.STDOUT, shell=True,\n",
    "            universal_newlines=True)\n",
    "        final_out = output.split('\\n')[-2]\n",
    "        return True\n",
    "    except Exception as exc:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_loc = '/home/mattolm/user_data/Covid_19/reads/'\n",
    "odir = '/home/mattolm/user_data/Covid_19/Pipeline/test/'\n",
    "\n",
    "SRdb['sra_file'] = [base_loc + x for x in SRdb['Run']]\n",
    "SRdb['sra_file_consistant'] = [(R in VALIDATED_RUNS) for R in SRdb['Run']]\n",
    "SRdb['sra_file_consistant'] = [check_sra_file(x) if y == False else y for x, y in zip(SRdb['sra_file'], SRdb['sra_file_consistant'])]\n",
    "\n",
    "TOTAL_LOOPS = 50\n",
    "loop = 0\n",
    "while len(SRdb[SRdb['sra_file_consistant'] == False]) > 0:\n",
    "    print('Running loop {0}; downloading {1} new files'.format(loop, len(SRdb[SRdb['sra_file_consistant'] == False])))\n",
    "    cmd_loc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/cmds/{0}_download.txt'.format(d4)\n",
    "    with open(cmd_loc, 'w') as o:\n",
    "        for i, row in SRdb[SRdb['sra_file_consistant'] == False].iterrows():\n",
    "            o.write(\"wget -c {0} -O {1}\\n\".format(row['sra_url'], row['sra_file']))\n",
    "    cmd = \"cat {0} | parallel -j {1}\".format(cmd_loc, THREADS)\n",
    "    call(cmd, shell=True)\n",
    "            \n",
    "    SRdb['sra_file_consistant'] = [check_sra_file(x) if y == False else y for x, y in zip(SRdb['sra_file'], SRdb['sra_file_consistant'])]\n",
    "    loop += 1\n",
    "    if loop == TOTAL_LOOPS:\n",
    "        break\n",
    "        \n",
    "if len(SRdb[SRdb['sra_file_consistant'] == False]) > 0:\n",
    "    print(\"THE FOLLOWING HAVE PROBLEMS:\")\n",
    "    for i, row in SRdb[SRdb['sra_file_consistant'] == False].iterrows():\n",
    "        print(\"vdb-validate {0}\".format(row['sra_file']))\n",
    "        print(\"wget -c {0} -O {1}\\n\".format(row['sra_url'], row['sra_file']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All are good!\n"
     ]
    }
   ],
   "source": [
    "if len(SRdb[SRdb['sra_file_consistant'] == False]) > 0:\n",
    "    print(\"THE FOLLOWING {0} HAVE PROBLEMS:\".format(len(SRdb[SRdb['sra_file_consistant'] == False])))\n",
    "    for i, row in SRdb[SRdb['sra_file_consistant'] == False].iterrows():\n",
    "        print(\"/usr/bin/vdb-validate {0}\".format(row['sra_file']))\n",
    "        print(\"wget -c {0} -O {1}\\n\".format(row['sra_url'], row['sra_file']))\n",
    "else:\n",
    "    print(\"All are good!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save new validations\n",
    "VALIDATED_RUNS = VALIDATED_RUNS.union(set(SRdb[SRdb['sra_file_consistant'] == True]['Run'].tolist()))\n",
    "with open('/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/validated_downloads.txt', 'w') as o:\n",
    "    for v in VALIDATED_RUNS:\n",
    "        o.write(v + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "SRdb = SRdb[SRdb['sra_file_consistant'] == True]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FastqQ Dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loop 0 - 7 samples are missing their fastq dumps\n",
      "Loop 1 - 7 samples are missing their fastq dumps\n",
      "Loop 2 - 7 samples are missing their fastq dumps\n",
      "7 samples are still missing their fastq dumps. They will be marked as single\n",
      "fastq-dump /home/mattolm/user_data/Covid_19/reads/SRR11542288 --split-files -O /home/mattolm/user_data/Covid_19/reads/\n",
      "\n",
      "fastq-dump /home/mattolm/user_data/Covid_19/reads/SRR11542289 --split-files -O /home/mattolm/user_data/Covid_19/reads/\n",
      "\n",
      "fastq-dump /home/mattolm/user_data/Covid_19/reads/SRR11513114 --split-files -O /home/mattolm/user_data/Covid_19/reads/\n",
      "\n",
      "fastq-dump /home/mattolm/user_data/Covid_19/reads/SRR11513118 --split-files -O /home/mattolm/user_data/Covid_19/reads/\n",
      "\n",
      "fastq-dump /home/mattolm/user_data/Covid_19/reads/SRR11479033 --split-files -O /home/mattolm/user_data/Covid_19/reads/\n",
      "\n",
      "fastq-dump /home/mattolm/user_data/Covid_19/reads/SRR11479040 --split-files -O /home/mattolm/user_data/Covid_19/reads/\n",
      "\n",
      "fastq-dump /home/mattolm/user_data/Covid_19/reads/SRR11479042 --split-files -O /home/mattolm/user_data/Covid_19/reads/\n",
      "\n"
     ]
    }
   ],
   "source": [
    "base_loc = '/home/mattolm/user_data/Covid_19/reads/'\n",
    "\n",
    "SRdb['fastq1'] = [x + '_1.fastq' for x in SRdb['sra_file']]\n",
    "SRdb['fastq2'] = [x + '_2.fastq' for x in SRdb['sra_file']]\n",
    "\n",
    "SRdb['fastq1_exists'] = [os.path.exists(x) for x in SRdb['fastq1']]\n",
    "SRdb['fastq2_exists'] = [os.path.exists(x) for x in SRdb['fastq2']]\n",
    "\n",
    "loop = 0\n",
    "MISSING_FASTA = SRdb[[(row['fastq1_exists'] == False) if row['TrueLibraryLayout'] == 'SINGLE' else (row['fastq2_exists'] == False) for i, row in SRdb.iterrows()]]\n",
    "while len(MISSING_FASTA) > 0:\n",
    "    print(\"Loop {0} - {1} samples are missing their fastq dumps\".format(loop, len(MISSING_FASTA)))\n",
    "    \n",
    "    cmd_loc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/cmds/{0}_dump.txt'.format(d4)\n",
    "    with open(cmd_loc, 'w') as o:\n",
    "        for i, row in MISSING_FASTA.iterrows():\n",
    "            o.write(\"fastq-dump {0} --split-files -O {1}\".format(row['sra_file'], base_loc) + '\\n')\n",
    "    cmd = \"cat {0} | parallel -j {1}\".format(cmd_loc, THREADS)\n",
    "    call(cmd, shell=True)\n",
    "\n",
    "    SRdb['fastq1_exists'] = [os.path.exists(x) for x in SRdb['fastq1']]\n",
    "    MISSING_FASTA = SRdb[[(row['fastq1_exists'] == False) if row['TrueLibraryLayout'] == 'SINGLE' else (row['fastq2_exists'] == False) for i, row in SRdb.iterrows()]]\n",
    "    loop += 1\n",
    "    \n",
    "    if loop == 3:\n",
    "        break\n",
    "        \n",
    "if len(MISSING_FASTA) > 0:\n",
    "    print(\"{0} samples are still missing their fastq dumps. They will be marked as single\".format(len(MISSING_FASTA)))\n",
    "    for i, row in MISSING_FASTA.iterrows():\n",
    "        print(\"fastq-dump {0} --split-files -O {1}\".format(row['sra_file'], base_loc) + '\\n')\n",
    "    \n",
    "SRdb['LibraryLayout'] = ['PAIRED' if t else 'SINGLE' for t in SRdb['fastq2_exists']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PAIRED    449\n",
       "SINGLE     10\n",
       "Name: LibraryLayout, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SRdb['LibraryLayout'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PAIRED\n",
      "PAIRED    446\n",
      "SINGLE      7\n",
      "Name: LibraryLayout, dtype: int64\n",
      "SINGLE\n",
      "SINGLE    3\n",
      "PAIRED    3\n",
      "Name: LibraryLayout, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "for x, db in SRdb.groupby('TrueLibraryLayout'):\n",
    "    print(x)\n",
    "    print(db['LibraryLayout'].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "SRdb['filtered_fastq1'] = ['/home/mattolm/user_data/Covid_19/reads/filtered/' + os.path.basename(rp).replace('_1.fastq', '_rep1.fastq').replace('_rep1.fastq', '_bbduk_1.fastq') for rp in SRdb['fastq1']]\n",
    "SRdb['filtered_fastq2'] = ['/home/mattolm/user_data/Covid_19/reads/filtered/' + os.path.basename(rp).replace('_2.fastq', '_rep2.fastq').replace('_rep2.fastq', '_bbduk_2.fastq') for rp in SRdb['fastq2']]\n",
    "SRdb['filtered_fastqS'] = ['/home/mattolm/user_data/Covid_19/reads/filtered/' + os.path.basename(rp).replace('_1.fastq', '_rep1.fastq').replace('_rep1.fastq', '_bbduk.fastq') for rp in SRdb['fastq1']]\n",
    "\n",
    "SRdb['filtered_fastq1_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastq1']]\n",
    "SRdb['filtered_fastq2_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastq2']]\n",
    "SRdb['filtered_fastqS_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastqS']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from subprocess import call\n",
    "def process_reads(r1, r2, outfolder):\n",
    "    # repair\n",
    "    rr1 = outfolder + os.path.basename(r1).replace('_1.fastq', '_rep1.fastq')\n",
    "    rr2 = outfolder + os.path.basename(r2).replace('_2.fastq', '_rep2.fastq')\n",
    "    cmd = \"repair.sh in={0} in2={1} out={2} out2={3}\".format(r1, r2, rr1, rr2)\n",
    "    print(cmd)\n",
    "    call(cmd, shell=True)\n",
    "    \n",
    "    # bbduk\n",
    "    rb1 = rr1.replace('_rep1.fastq', '_bbduk_1.fastq')\n",
    "    rb2 = rr2.replace('_rep2.fastq', '_bbduk_2.fastq')\n",
    "    cmd = \"bbduk.sh in={0} in2={1} out={2} out2={3} threads={4}\".format(rr1, rr2, rb1, rb2, THREADS)\n",
    "    print(cmd)\n",
    "    call(cmd, shell=True)\n",
    "    \n",
    "    return rb1, rb2\n",
    "\n",
    "def process_reads_s(r1, outfolder):\n",
    "    # bbduk\n",
    "    rr1 = outfolder + os.path.basename(r1).replace('_1.fastq', '_rep1.fastq')\n",
    "    rb1 = rr1.replace('_rep1.fastq', '_bbduk.fastq')\n",
    "    cmd = \"bbduk.sh in={0} out={1} threads={2}\".format(r1, rb1, THREADS)\n",
    "    print(cmd)\n",
    "    call(cmd, shell=True)\n",
    "    \n",
    "    return rb1 \n",
    "\n",
    "SRdb['filtered_fastq1_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastq1']]\n",
    "SRdb['filtered_fastq2_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastq2']]\n",
    "SRdb['filtered_fastqS_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastqS']]\n",
    "\n",
    "# Process paired\n",
    "for i, row in SRdb[(SRdb['LibraryLayout'] == 'PAIRED') & ((SRdb['filtered_fastq1_exists'] == False) | (SRdb['filtered_fastq2_exists'] == False))].iterrows():\n",
    "    rb1, rb2 = process_reads(row['fastq1'], row['fastq2'], '/home/mattolm/user_data/Covid_19/reads/filtered/')\n",
    "    \n",
    "# Process unpaired\n",
    "for i, row in SRdb[(SRdb['LibraryLayout'] == 'SINGLE') & (SRdb['filtered_fastqS_exists'] == False)].iterrows():\n",
    "    b1 = process_reads_s(row['fastq1'], '/home/mattolm/user_data/Covid_19/reads/filtered/')\n",
    "    \n",
    "SRdb['filtered_fastq1_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastq1']]\n",
    "SRdb['filtered_fastq2_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastq2']]\n",
    "SRdb['filtered_fastqS_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastqS']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, row in SRdb[(SRdb['LibraryLayout'] == 'PAIRED') & ((SRdb['filtered_fastq1_exists'] == False) | (SRdb['filtered_fastq2_exists'] == False))].iterrows():\n",
    "    print(\"{0} still has a problem\".format(row['Run']))\n",
    "    \n",
    "for i, row in SRdb[(SRdb['LibraryLayout'] == 'SINGLE') & (SRdb['filtered_fastqS_exists'] == False)].iterrows():\n",
    "    print(\"{0} still has a problem\".format(row['Run']))\n",
    "#     rb1, rb2 = process_reads(row['fastq1'], row['fastq2'], '/home/mattolm/user_data/Covid_19/reads/filtered/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Need to run 0 mapping jobs\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True    459\n",
       "Name: mapping_done, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BTL = '/home/mattolm/user_data/Covid_19/genomes/NC_045512.2.fasta.bt2'\n",
    "ML = '/home/mattolm/user_data/Covid_19/inStrain/mapping_files/'\n",
    "s2l = SRdb.set_index('Run')['LibraryLayout'].to_dict()\n",
    "\n",
    "SRdb['sam_file'] = [\"{0}{1}-vs-{2}.sam\".format(ML, os.path.basename(BTL), r) for r in SRdb['Run']]\n",
    "SRdb['mapping_done'] = [os.path.isfile(s) for s in SRdb['sam_file']]\n",
    "\n",
    "cmd_loc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/cmds/{0}_mapping.txt'.format(d4)\n",
    "re = open(cmd_loc, 'w')\n",
    "\n",
    "print(\"Need to run {0} mapping jobs\".format(len(SRdb[SRdb['mapping_done'] == False])))\n",
    "\n",
    "for i, row in SRdb[SRdb['mapping_done'] == False].iterrows():\n",
    "# for i, row in SRdb.iterrows():\n",
    "    sam_loc = row['sam_file']\n",
    "    if s2l[row['Run']] == 'PAIRED':\n",
    "        cmd = \"bowtie2 -x {3} -1 {0} -2 {1} --no-unal -S {2} -p {4} 2> {2}.log\".format(\n",
    "                row['filtered_fastq1'], row['filtered_fastq2'], sam_loc, BTL, THREADS)\n",
    "        re.write(cmd + '\\n')\n",
    "    else:\n",
    "        cmd = \"bowtie2 -x {3} -U {0} --no-unal -S {2} -p {4} 2> {2}.log\".format(\n",
    "                row['filtered_fastqS'], '', sam_loc, BTL, THREADS)\n",
    "        re.write(cmd + '\\n')\n",
    "re.close()\n",
    "\n",
    "cmd = \"cat {0} | bash\".format(cmd_loc)\n",
    "call(cmd, shell=True)\n",
    "\n",
    "SRdb['mapping_done'] = [os.path.isfile(s) for s in SRdb['sam_file']]\n",
    "SRdb['mapping_done'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse bowtie2 logs for information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_bt2(log):\n",
    "    table = defaultdict(list)\n",
    "    with open(log, 'r') as o:\n",
    "        for line in o.readlines():\n",
    "            line = line.strip()\n",
    "            if 'reads;' in line:\n",
    "                table['total_reads'].append(int(line.split()[0]))\n",
    "            if 'paired;' in line:\n",
    "                table['paired_reads'].append(int(line.split()[0]))\n",
    "            if 'overall alignment rate' in line:\n",
    "                table['percent_aligned'].append(float(line.split()[0][:-1]))\n",
    "    db = pd.DataFrame(table)\n",
    "    db['aligned_reads'] = db['total_reads'] * (db['percent_aligned'] / 100)\n",
    "    db['aligned_reads'] = db['aligned_reads'].astype(int)\n",
    "    return db\n",
    "\n",
    "dbs = []\n",
    "SRdb['mapping_log'] = SRdb['sam_file'] + '.log'\n",
    "SRdb['mapping_log_exists'] = [os.path.exists(x) for x in SRdb['mapping_log']]\n",
    "for i, row in SRdb.iterrows():\n",
    "    db = parse_bt2(row['mapping_log'])\n",
    "    db['Run'] = row['Run']\n",
    "    dbs.append(db)\n",
    "Mdb = pd.concat(dbs).reset_index(drop=True)\n",
    "for col in Mdb.columns:\n",
    "    if col in ['Run']:\n",
    "        continue\n",
    "    r2c = Mdb.set_index('Run')[col].to_dict()\n",
    "    SRdb[col] = SRdb['Run'].map(r2c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run inStrain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    459\n",
       "Name: inStrain_succeeded, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SRdb['inStrain_coverage'] = [\"/home/mattolm/user_data/Covid_19/inStrain/profiles_v2/{0}.IS/output/{0}.IS_scaffold_info.tsv\".format(os.path.basename(r)) for r in SRdb['sam_file']]\n",
    "SRdb['inStrain_succeeded'] = [os.path.isfile(s) for s in SRdb['inStrain_coverage']]\n",
    "SRdb['inStrain_succeeded'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cat /home/mattolm/user_data/Covid_19/Pipeline/Jupyter/cmds/04202020_instrain.txt | parallel -j 6\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False    459\n",
       "Name: inStrain_succeeded, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmd_loc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/cmds/{0}_instrain.txt'.format(d4)\n",
    "fasta = '/home/mattolm/user_data/Covid_19/genomes/NC_045512.2.fasta'\n",
    "genes ='/home/mattolm/user_data/Covid_19/genomes/NC_045512.2.gb'\n",
    "s2l = SRdb.set_index('Run')['LibraryLayout'].to_dict()\n",
    "\n",
    "re = open(cmd_loc, 'w')\n",
    "#for i, row in SRdb[SRdb['inStrain_succeeded'] == False].iterrows():\n",
    "for i, row in SRdb.iterrows():\n",
    "    if s2l[row['Run']] == 'PAIRED': \n",
    "        cmd = \"inStrain profile {0} {2} -o /home/mattolm/user_data/Covid_19/inStrain/profiles_v2/{1}.IS -p 1 -g {3} --skip_mm_profiling\".format(row['sam_file'], os.path.basename(row['sam_file']), fasta, genes)\n",
    "        re.write(cmd + '\\n')\n",
    "    else:\n",
    "        cmd = \"inStrain profile {0} {2} -o /home/mattolm/user_data/Covid_19/inStrain/profiles_v2/{1}.IS -p 1 --pairing_filter non_discordant -g {3} --skip_mm_profiling\".format(row['sam_file'], os.path.basename(row['sam_file']), fasta, genes)\n",
    "        re.write(cmd + '\\n')\n",
    "re.close()\n",
    "    \n",
    "cmd = \"cat {0} | parallel -j {1}\".format(cmd_loc, THREADS)\n",
    "print(cmd)\n",
    "#call(cmd, shell=True)\n",
    "    \n",
    "SRdb['inStrain_succeeded'] = [os.path.isfile(s) for s in SRdb['inStrain_coverage']]\n",
    "SRdb['inStrain_succeeded'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     458\n",
       "False      1\n",
       "Name: inStrain_succeeded, dtype: int64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SRdb['inStrain_succeeded'] = [os.path.isfile(s) for s in SRdb['inStrain_coverage']]\n",
    "SRdb['inStrain_succeeded'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 samples failed inStrain that should have passed\n"
     ]
    }
   ],
   "source": [
    "print(\"{0} samples failed inStrain that should have passed\".format(len(SRdb[(SRdb['inStrain_succeeded'] == False) & (SRdb['aligned_reads'] > 0)])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/home/mattolm/user_data/Covid_19/inStrain/profiles_v2/NC_045512.2.fasta.bt2-vs-SRR11059947.sam.IS/log/log.log']"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "['/'.join(x.split('/')[:-2]) + '/log/log.log' for x in SRdb[(SRdb['inStrain_succeeded'] == False) & (SRdb['aligned_reads'] > 0)]['inStrain_coverage'].tolist()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add sequence release date\n",
    "SRdb['release_date'] = [os.path.getctime(s) for s in SRdb['sra_file']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>experiment_desc</th>\n",
       "      <th>organism_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>sample_title_x</th>\n",
       "      <th>instrument</th>\n",
       "      <th>total_spots</th>\n",
       "      <th>total_size</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>run_total_spots</th>\n",
       "      <th>run_total_bases</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>sra_url_alt</th>\n",
       "      <th>sra_url</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>isolate</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>host</th>\n",
       "      <th>host_disease</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>lat_lon</th>\n",
       "      <th>BioSampleModel</th>\n",
       "      <th>passage_history</th>\n",
       "      <th>strain</th>\n",
       "      <th>source_name</th>\n",
       "      <th>subject status</th>\n",
       "      <th>tissue/cell type</th>\n",
       "      <th>treatment</th>\n",
       "      <th>time after treatment</th>\n",
       "      <th>sub_species</th>\n",
       "      <th>cell line</th>\n",
       "      <th>description</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>host_taxid</th>\n",
       "      <th>isol_growth_condt</th>\n",
       "      <th>propagation</th>\n",
       "      <th>sample_title_y</th>\n",
       "      <th>seq_methods</th>\n",
       "      <th>source_uvig</th>\n",
       "      <th>virus_enrich_appr</th>\n",
       "      <th>...</th>\n",
       "      <th>host_health_state</th>\n",
       "      <th>host_sex</th>\n",
       "      <th>host_subject_id</th>\n",
       "      <th>host_tissue_sampled</th>\n",
       "      <th>pathotype</th>\n",
       "      <th>serotype</th>\n",
       "      <th>serovar</th>\n",
       "      <th>specimen_voucher</th>\n",
       "      <th>subgroup</th>\n",
       "      <th>subtype</th>\n",
       "      <th>cell type</th>\n",
       "      <th>time point</th>\n",
       "      <th>identification_method</th>\n",
       "      <th>sample type</th>\n",
       "      <th>Laboratory Host</th>\n",
       "      <th>Extraction Method</th>\n",
       "      <th>ref_biomaterial</th>\n",
       "      <th>link_addit_analys</th>\n",
       "      <th>country</th>\n",
       "      <th>sra_url_alt1</th>\n",
       "      <th>sra_url_alt2</th>\n",
       "      <th>ena_fastq_url</th>\n",
       "      <th>ena_fastq_ftp</th>\n",
       "      <th>Run</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>TrueLibraryLayout</th>\n",
       "      <th>sra_file</th>\n",
       "      <th>sra_file_consistant</th>\n",
       "      <th>fastq1</th>\n",
       "      <th>fastq2</th>\n",
       "      <th>fastq1_exists</th>\n",
       "      <th>fastq2_exists</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>filtered_fastq1</th>\n",
       "      <th>filtered_fastq2</th>\n",
       "      <th>filtered_fastqS</th>\n",
       "      <th>filtered_fastq1_exists</th>\n",
       "      <th>filtered_fastq2_exists</th>\n",
       "      <th>filtered_fastqS_exists</th>\n",
       "      <th>sam_file</th>\n",
       "      <th>mapping_done</th>\n",
       "      <th>mapping_log</th>\n",
       "      <th>mapping_log_exists</th>\n",
       "      <th>total_reads</th>\n",
       "      <th>paired_reads</th>\n",
       "      <th>percent_aligned</th>\n",
       "      <th>aligned_reads</th>\n",
       "      <th>inStrain_coverage</th>\n",
       "      <th>inStrain_succeeded</th>\n",
       "      <th>release_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRP250294</td>\n",
       "      <td>SRX8112385</td>\n",
       "      <td>SISPA of SARS-CoV-2 from cell culture</td>\n",
       "      <td>SISPA of SARS-CoV-2 from cell culture</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>WGS</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>SRS6187447</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>128808.0</td>\n",
       "      <td>1.286284e+07</td>\n",
       "      <td>SRR11542288</td>\n",
       "      <td>128808.0</td>\n",
       "      <td>3.035218e+07</td>\n",
       "      <td>p2b_vero76.fastq</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR11542288</td>\n",
       "      <td>PRJNA607948</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1154...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1154...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1154...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>128808</td>\n",
       "      <td>128808</td>\n",
       "      <td>98.60</td>\n",
       "      <td>127004</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/prof...</td>\n",
       "      <td>True</td>\n",
       "      <td>1.587408e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRP250294</td>\n",
       "      <td>SRX8112384</td>\n",
       "      <td>SISPA of SARS-CoV-2 from cell culture</td>\n",
       "      <td>SISPA of SARS-CoV-2 from cell culture</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>WGS</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>SRS6187448</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>187666.0</td>\n",
       "      <td>1.795959e+07</td>\n",
       "      <td>SRR11542289</td>\n",
       "      <td>187666.0</td>\n",
       "      <td>4.271257e+07</td>\n",
       "      <td>p2a_vero76.fastq</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR11542289</td>\n",
       "      <td>PRJNA607948</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1154...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1154...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1154...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>187666</td>\n",
       "      <td>187666</td>\n",
       "      <td>98.98</td>\n",
       "      <td>185751</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/prof...</td>\n",
       "      <td>True</td>\n",
       "      <td>1.587408e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRP254488</td>\n",
       "      <td>SRX8112342</td>\n",
       "      <td>SARS-CoV-2/190300/human/2020/Malaysia _EPI_ISL...</td>\n",
       "      <td>SARS-CoV-2/190300/human/2020/Malaysia _EPI_ISL...</td>\n",
       "      <td>2697049.0</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PCR</td>\n",
       "      <td>SRS6477428</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina iSeq 100</td>\n",
       "      <td>898380.0</td>\n",
       "      <td>5.460871e+07</td>\n",
       "      <td>SRR11542243</td>\n",
       "      <td>898380.0</td>\n",
       "      <td>1.538686e+08</td>\n",
       "      <td>EPI_ISL_417920_S3_L001_R1_001.fastq.gz</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV-2/190300/human/2020/Malaysia</td>\n",
       "      <td>Universiti Malaya COVID Research group</td>\n",
       "      <td>22-Mar-2020</td>\n",
       "      <td>Malaysia</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Nasopharyngeal/throat swab</td>\n",
       "      <td>3.1390 N 101.6869 E</td>\n",
       "      <td>Pathogen.cl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR11542243</td>\n",
       "      <td>PRJNA616147</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1154...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1154...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1154...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>898380</td>\n",
       "      <td>898380</td>\n",
       "      <td>91.41</td>\n",
       "      <td>821209</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/prof...</td>\n",
       "      <td>True</td>\n",
       "      <td>1.587408e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRP254488</td>\n",
       "      <td>SRX8112341</td>\n",
       "      <td>SARS-CoV-2/186197/human/2020/Malaysia_EPI_ISL_...</td>\n",
       "      <td>SARS-CoV-2/186197/human/2020/Malaysia_EPI_ISL_...</td>\n",
       "      <td>2697049.0</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PCR</td>\n",
       "      <td>SRS6477427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina iSeq 100</td>\n",
       "      <td>733611.0</td>\n",
       "      <td>4.223798e+07</td>\n",
       "      <td>SRR11542244</td>\n",
       "      <td>733611.0</td>\n",
       "      <td>1.184563e+08</td>\n",
       "      <td>EPI_ISL_417919_S4_L001_R1_001.fastq.gz</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV-2/186197/human/2020/Malaysia</td>\n",
       "      <td>Universiti Malaya COVID Research group</td>\n",
       "      <td>14-Mar-2020</td>\n",
       "      <td>Malaysia</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Nasopharyngeal/throat swab</td>\n",
       "      <td>3.1390 N 101.6869 E</td>\n",
       "      <td>Pathogen.cl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR11542244</td>\n",
       "      <td>PRJNA616147</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1154...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1154...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1154...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>733611</td>\n",
       "      <td>733611</td>\n",
       "      <td>19.18</td>\n",
       "      <td>140706</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/prof...</td>\n",
       "      <td>True</td>\n",
       "      <td>1.587408e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRP255993</td>\n",
       "      <td>SRX8095881</td>\n",
       "      <td>WGS of SARS-CoV-2 R03006_2020_2</td>\n",
       "      <td>WGS of SARS-CoV-2 R03006_2020_2</td>\n",
       "      <td>2697049.0</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS6462475</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2971431.0</td>\n",
       "      <td>1.079579e+09</td>\n",
       "      <td>SRR11524818</td>\n",
       "      <td>2971431.0</td>\n",
       "      <td>1.787565e+09</td>\n",
       "      <td>R03006_2_2020_S2_L001_R1_001.fastq.gz</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>National Institute for Communicable Diseases, ...</td>\n",
       "      <td>2020-03-07</td>\n",
       "      <td>South Africa: KwaZulu-Natal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Combined nasopharyngeal and oropharyngeal swab</td>\n",
       "      <td>28.5306 S 30.8958 E</td>\n",
       "      <td>Pathogen.cl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R03006_2020_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR11524818</td>\n",
       "      <td>PRJNA624358</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1152...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1152...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1152...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>38800</td>\n",
       "      <td>38800</td>\n",
       "      <td>0.01</td>\n",
       "      <td>3</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/prof...</td>\n",
       "      <td>True</td>\n",
       "      <td>1.587408e+09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 106 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession experiment_accession  \\\n",
       "0       SRP250294           SRX8112385   \n",
       "1       SRP250294           SRX8112384   \n",
       "2       SRP254488           SRX8112342   \n",
       "3       SRP254488           SRX8112341   \n",
       "4       SRP255993           SRX8095881   \n",
       "\n",
       "                                    experiment_title  \\\n",
       "0              SISPA of SARS-CoV-2 from cell culture   \n",
       "1              SISPA of SARS-CoV-2 from cell culture   \n",
       "2  SARS-CoV-2/190300/human/2020/Malaysia _EPI_ISL...   \n",
       "3  SARS-CoV-2/186197/human/2020/Malaysia_EPI_ISL_...   \n",
       "4                    WGS of SARS-CoV-2 R03006_2020_2   \n",
       "\n",
       "                                     experiment_desc  organism_taxid   \\\n",
       "0              SISPA of SARS-CoV-2 from cell culture           9606.0   \n",
       "1              SISPA of SARS-CoV-2 from cell culture           9606.0   \n",
       "2  SARS-CoV-2/190300/human/2020/Malaysia _EPI_ISL...        2697049.0   \n",
       "3  SARS-CoV-2/186197/human/2020/Malaysia_EPI_ISL_...        2697049.0   \n",
       "4                    WGS of SARS-CoV-2 R03006_2020_2        2697049.0   \n",
       "\n",
       "                                     organism_name library_strategy  \\\n",
       "0                                     Homo sapiens              WGS   \n",
       "1                                     Homo sapiens              WGS   \n",
       "2  Severe acute respiratory syndrome coronavirus 2         AMPLICON   \n",
       "3  Severe acute respiratory syndrome coronavirus 2         AMPLICON   \n",
       "4  Severe acute respiratory syndrome coronavirus 2              WGS   \n",
       "\n",
       "  library_source library_selection sample_accession  sample_title_x  \\\n",
       "0    METAGENOMIC        RANDOM PCR       SRS6187447             NaN   \n",
       "1    METAGENOMIC        RANDOM PCR       SRS6187448             NaN   \n",
       "2      VIRAL RNA               PCR       SRS6477428             NaN   \n",
       "3      VIRAL RNA               PCR       SRS6477427             NaN   \n",
       "4        GENOMIC            RANDOM       SRS6462475             NaN   \n",
       "\n",
       "          instrument  total_spots    total_size run_accession  \\\n",
       "0     Illumina MiSeq     128808.0  1.286284e+07   SRR11542288   \n",
       "1     Illumina MiSeq     187666.0  1.795959e+07   SRR11542289   \n",
       "2  Illumina iSeq 100     898380.0  5.460871e+07   SRR11542243   \n",
       "3  Illumina iSeq 100     733611.0  4.223798e+07   SRR11542244   \n",
       "4     Illumina MiSeq    2971431.0  1.079579e+09   SRR11524818   \n",
       "\n",
       "   run_total_spots  run_total_bases                               run_alias  \\\n",
       "0         128808.0     3.035218e+07                        p2b_vero76.fastq   \n",
       "1         187666.0     4.271257e+07                        p2a_vero76.fastq   \n",
       "2         898380.0     1.538686e+08  EPI_ISL_417920_S3_L001_R1_001.fastq.gz   \n",
       "3         733611.0     1.184563e+08  EPI_ISL_417919_S4_L001_R1_001.fastq.gz   \n",
       "4        2971431.0     1.787565e+09   R03006_2_2020_S2_L001_R1_001.fastq.gz   \n",
       "\n",
       "                                         sra_url_alt  \\\n",
       "0  https://sra-download.ncbi.nlm.nih.gov/traces/s...   \n",
       "1  https://sra-download.ncbi.nlm.nih.gov/traces/s...   \n",
       "2  https://sra-download.ncbi.nlm.nih.gov/traces/s...   \n",
       "3  https://sra-download.ncbi.nlm.nih.gov/traces/s...   \n",
       "4  https://sra-download.ncbi.nlm.nih.gov/traces/s...   \n",
       "\n",
       "                                             sra_url experiment_alias  \\\n",
       "0  https://sra-download.ncbi.nlm.nih.gov/traces/s...              NaN   \n",
       "1  https://sra-download.ncbi.nlm.nih.gov/traces/s...              NaN   \n",
       "2  https://sra-download.ncbi.nlm.nih.gov/traces/s...              NaN   \n",
       "3  https://sra-download.ncbi.nlm.nih.gov/traces/s...              NaN   \n",
       "4  https://sra-download.ncbi.nlm.nih.gov/traces/s...              NaN   \n",
       "\n",
       "                                 isolate  \\\n",
       "0                                    NaN   \n",
       "1                                    NaN   \n",
       "2  SARS-CoV-2/190300/human/2020/Malaysia   \n",
       "3  SARS-CoV-2/186197/human/2020/Malaysia   \n",
       "4                                    NaN   \n",
       "\n",
       "                                        collected_by collection_date  \\\n",
       "0                                                NaN             NaN   \n",
       "1                                                NaN             NaN   \n",
       "2             Universiti Malaya COVID Research group     22-Mar-2020   \n",
       "3             Universiti Malaya COVID Research group     14-Mar-2020   \n",
       "4  National Institute for Communicable Diseases, ...      2020-03-07   \n",
       "\n",
       "                  geo_loc_name          host host_disease  \\\n",
       "0                          NaN           NaN          NaN   \n",
       "1                          NaN           NaN          NaN   \n",
       "2                     Malaysia  Homo sapiens     COVID-19   \n",
       "3                     Malaysia  Homo sapiens     COVID-19   \n",
       "4  South Africa: KwaZulu-Natal  Homo sapiens     COVID-19   \n",
       "\n",
       "                                 isolation_source              lat_lon  \\\n",
       "0                                             NaN                  NaN   \n",
       "1                                             NaN                  NaN   \n",
       "2                      Nasopharyngeal/throat swab  3.1390 N 101.6869 E   \n",
       "3                      Nasopharyngeal/throat swab  3.1390 N 101.6869 E   \n",
       "4  Combined nasopharyngeal and oropharyngeal swab  28.5306 S 30.8958 E   \n",
       "\n",
       "  BioSampleModel passage_history         strain  source_name  subject status  \\\n",
       "0            NaN             NaN            NaN          NaN             NaN   \n",
       "1            NaN             NaN            NaN          NaN             NaN   \n",
       "2    Pathogen.cl             NaN            NaN          NaN             NaN   \n",
       "3    Pathogen.cl             NaN            NaN          NaN             NaN   \n",
       "4    Pathogen.cl             NaN  R03006_2020_2          NaN             NaN   \n",
       "\n",
       "   tissue/cell type  treatment  time after treatment  sub_species  cell line  \\\n",
       "0               NaN        NaN                   NaN          NaN        NaN   \n",
       "1               NaN        NaN                   NaN          NaN        NaN   \n",
       "2               NaN        NaN                   NaN          NaN        NaN   \n",
       "3               NaN        NaN                   NaN          NaN        NaN   \n",
       "4               NaN        NaN                   NaN          NaN        NaN   \n",
       "\n",
       "  description env_broad_scale env_local_scale env_medium  host_taxid  \\\n",
       "0         NaN             NaN             NaN        NaN         NaN   \n",
       "1         NaN             NaN             NaN        NaN         NaN   \n",
       "2         NaN             NaN             NaN        NaN         NaN   \n",
       "3         NaN             NaN             NaN        NaN         NaN   \n",
       "4         NaN             NaN             NaN        NaN         NaN   \n",
       "\n",
       "  isol_growth_condt propagation sample_title_y seq_methods source_uvig  \\\n",
       "0               NaN         NaN            NaN         NaN         NaN   \n",
       "1               NaN         NaN            NaN         NaN         NaN   \n",
       "2               NaN         NaN            NaN         NaN         NaN   \n",
       "3               NaN         NaN            NaN         NaN         NaN   \n",
       "4               NaN         NaN            NaN         NaN         NaN   \n",
       "\n",
       "  virus_enrich_appr  ... host_health_state host_sex host_subject_id  \\\n",
       "0               NaN  ...               NaN      NaN             NaN   \n",
       "1               NaN  ...               NaN      NaN             NaN   \n",
       "2               NaN  ...               NaN      NaN             NaN   \n",
       "3               NaN  ...               NaN      NaN             NaN   \n",
       "4               NaN  ...               NaN      NaN             NaN   \n",
       "\n",
       "  host_tissue_sampled pathotype serotype serovar specimen_voucher subgroup  \\\n",
       "0                 NaN       NaN      NaN     NaN              NaN      NaN   \n",
       "1                 NaN       NaN      NaN     NaN              NaN      NaN   \n",
       "2                 NaN       NaN      NaN     NaN              NaN      NaN   \n",
       "3                 NaN       NaN      NaN     NaN              NaN      NaN   \n",
       "4                 NaN       NaN      NaN     NaN              NaN      NaN   \n",
       "\n",
       "  subtype cell type time point identification_method sample type  \\\n",
       "0     NaN       NaN        NaN                   NaN         NaN   \n",
       "1     NaN       NaN        NaN                   NaN         NaN   \n",
       "2     NaN       NaN        NaN                   NaN         NaN   \n",
       "3     NaN       NaN        NaN                   NaN         NaN   \n",
       "4     NaN       NaN        NaN                   NaN         NaN   \n",
       "\n",
       "  Laboratory Host Extraction Method  ref_biomaterial  link_addit_analys  \\\n",
       "0             NaN               NaN              NaN                NaN   \n",
       "1             NaN               NaN              NaN                NaN   \n",
       "2             NaN               NaN              NaN                NaN   \n",
       "3             NaN               NaN              NaN                NaN   \n",
       "4             NaN               NaN              NaN                NaN   \n",
       "\n",
       "  country sra_url_alt1 sra_url_alt2 ena_fastq_url ena_fastq_ftp          Run  \\\n",
       "0     NaN          NaN          NaN           NaN           NaN  SRR11542288   \n",
       "1     NaN          NaN          NaN           NaN           NaN  SRR11542289   \n",
       "2     NaN          NaN          NaN           NaN           NaN  SRR11542243   \n",
       "3     NaN          NaN          NaN           NaN           NaN  SRR11542244   \n",
       "4     NaN          NaN          NaN           NaN           NaN  SRR11524818   \n",
       "\n",
       "    BioProject TrueLibraryLayout  \\\n",
       "0  PRJNA607948            PAIRED   \n",
       "1  PRJNA607948            PAIRED   \n",
       "2  PRJNA616147            PAIRED   \n",
       "3  PRJNA616147            PAIRED   \n",
       "4  PRJNA624358            PAIRED   \n",
       "\n",
       "                                            sra_file  sra_file_consistant  \\\n",
       "0  /home/mattolm/user_data/Covid_19/reads/SRR1154...                 True   \n",
       "1  /home/mattolm/user_data/Covid_19/reads/SRR1154...                 True   \n",
       "2  /home/mattolm/user_data/Covid_19/reads/SRR1154...                 True   \n",
       "3  /home/mattolm/user_data/Covid_19/reads/SRR1154...                 True   \n",
       "4  /home/mattolm/user_data/Covid_19/reads/SRR1152...                 True   \n",
       "\n",
       "                                              fastq1  \\\n",
       "0  /home/mattolm/user_data/Covid_19/reads/SRR1154...   \n",
       "1  /home/mattolm/user_data/Covid_19/reads/SRR1154...   \n",
       "2  /home/mattolm/user_data/Covid_19/reads/SRR1154...   \n",
       "3  /home/mattolm/user_data/Covid_19/reads/SRR1154...   \n",
       "4  /home/mattolm/user_data/Covid_19/reads/SRR1152...   \n",
       "\n",
       "                                              fastq2 fastq1_exists  \\\n",
       "0  /home/mattolm/user_data/Covid_19/reads/SRR1154...          True   \n",
       "1  /home/mattolm/user_data/Covid_19/reads/SRR1154...          True   \n",
       "2  /home/mattolm/user_data/Covid_19/reads/SRR1154...          True   \n",
       "3  /home/mattolm/user_data/Covid_19/reads/SRR1154...          True   \n",
       "4  /home/mattolm/user_data/Covid_19/reads/SRR1152...          True   \n",
       "\n",
       "  fastq2_exists LibraryLayout  \\\n",
       "0         False        SINGLE   \n",
       "1         False        SINGLE   \n",
       "2          True        PAIRED   \n",
       "3          True        PAIRED   \n",
       "4          True        PAIRED   \n",
       "\n",
       "                                     filtered_fastq1  \\\n",
       "0  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "1  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "2  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "3  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "4  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "\n",
       "                                     filtered_fastq2  \\\n",
       "0  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "1  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "2  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "3  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "4  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "\n",
       "                                     filtered_fastqS  filtered_fastq1_exists  \\\n",
       "0  /home/mattolm/user_data/Covid_19/reads/filtere...                   False   \n",
       "1  /home/mattolm/user_data/Covid_19/reads/filtere...                   False   \n",
       "2  /home/mattolm/user_data/Covid_19/reads/filtere...                    True   \n",
       "3  /home/mattolm/user_data/Covid_19/reads/filtere...                    True   \n",
       "4  /home/mattolm/user_data/Covid_19/reads/filtere...                    True   \n",
       "\n",
       "   filtered_fastq2_exists filtered_fastqS_exists  \\\n",
       "0                   False                   True   \n",
       "1                   False                   True   \n",
       "2                    True                   True   \n",
       "3                    True                   True   \n",
       "4                    True                   True   \n",
       "\n",
       "                                            sam_file mapping_done  \\\n",
       "0  /home/mattolm/user_data/Covid_19/inStrain/mapp...         True   \n",
       "1  /home/mattolm/user_data/Covid_19/inStrain/mapp...         True   \n",
       "2  /home/mattolm/user_data/Covid_19/inStrain/mapp...         True   \n",
       "3  /home/mattolm/user_data/Covid_19/inStrain/mapp...         True   \n",
       "4  /home/mattolm/user_data/Covid_19/inStrain/mapp...         True   \n",
       "\n",
       "                                         mapping_log  mapping_log_exists  \\\n",
       "0  /home/mattolm/user_data/Covid_19/inStrain/mapp...                True   \n",
       "1  /home/mattolm/user_data/Covid_19/inStrain/mapp...                True   \n",
       "2  /home/mattolm/user_data/Covid_19/inStrain/mapp...                True   \n",
       "3  /home/mattolm/user_data/Covid_19/inStrain/mapp...                True   \n",
       "4  /home/mattolm/user_data/Covid_19/inStrain/mapp...                True   \n",
       "\n",
       "   total_reads  paired_reads percent_aligned  aligned_reads  \\\n",
       "0       128808        128808           98.60         127004   \n",
       "1       187666        187666           98.98         185751   \n",
       "2       898380        898380           91.41         821209   \n",
       "3       733611        733611           19.18         140706   \n",
       "4        38800         38800            0.01              3   \n",
       "\n",
       "                                   inStrain_coverage  inStrain_succeeded  \\\n",
       "0  /home/mattolm/user_data/Covid_19/inStrain/prof...                True   \n",
       "1  /home/mattolm/user_data/Covid_19/inStrain/prof...                True   \n",
       "2  /home/mattolm/user_data/Covid_19/inStrain/prof...                True   \n",
       "3  /home/mattolm/user_data/Covid_19/inStrain/prof...                True   \n",
       "4  /home/mattolm/user_data/Covid_19/inStrain/prof...                True   \n",
       "\n",
       "   release_date  \n",
       "0  1.587408e+09  \n",
       "1  1.587408e+09  \n",
       "2  1.587408e+09  \n",
       "3  1.587408e+09  \n",
       "4  1.587408e+09  \n",
       "\n",
       "[5 rows x 106 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Full\n",
    "SRdb.to_csv('/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA_full_info.csv'.format(d4), index=False)\n",
    "SRdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>experiment_desc</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>instrument</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>total_reads</th>\n",
       "      <th>paired_reads</th>\n",
       "      <th>percent_aligned</th>\n",
       "      <th>aligned_reads</th>\n",
       "      <th>release_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11542288</td>\n",
       "      <td>SISPA of SARS-CoV-2 from cell culture</td>\n",
       "      <td>SISPA of SARS-CoV-2 from cell culture</td>\n",
       "      <td>SRS6187447</td>\n",
       "      <td>WGS</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>128808</td>\n",
       "      <td>128808</td>\n",
       "      <td>98.60</td>\n",
       "      <td>127004</td>\n",
       "      <td>1.587408e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11542289</td>\n",
       "      <td>SISPA of SARS-CoV-2 from cell culture</td>\n",
       "      <td>SISPA of SARS-CoV-2 from cell culture</td>\n",
       "      <td>SRS6187448</td>\n",
       "      <td>WGS</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>187666</td>\n",
       "      <td>187666</td>\n",
       "      <td>98.98</td>\n",
       "      <td>185751</td>\n",
       "      <td>1.587408e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11542243</td>\n",
       "      <td>SARS-CoV-2/190300/human/2020/Malaysia _EPI_ISL...</td>\n",
       "      <td>SARS-CoV-2/190300/human/2020/Malaysia _EPI_ISL...</td>\n",
       "      <td>SRS6477428</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PCR</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina iSeq 100</td>\n",
       "      <td>Universiti Malaya COVID Research group</td>\n",
       "      <td>22-Mar-2020</td>\n",
       "      <td>Nasopharyngeal/throat swab</td>\n",
       "      <td>898380</td>\n",
       "      <td>898380</td>\n",
       "      <td>91.41</td>\n",
       "      <td>821209</td>\n",
       "      <td>1.587408e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11542244</td>\n",
       "      <td>SARS-CoV-2/186197/human/2020/Malaysia_EPI_ISL_...</td>\n",
       "      <td>SARS-CoV-2/186197/human/2020/Malaysia_EPI_ISL_...</td>\n",
       "      <td>SRS6477427</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PCR</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina iSeq 100</td>\n",
       "      <td>Universiti Malaya COVID Research group</td>\n",
       "      <td>14-Mar-2020</td>\n",
       "      <td>Nasopharyngeal/throat swab</td>\n",
       "      <td>733611</td>\n",
       "      <td>733611</td>\n",
       "      <td>19.18</td>\n",
       "      <td>140706</td>\n",
       "      <td>1.587408e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR11524818</td>\n",
       "      <td>WGS of SARS-CoV-2 R03006_2020_2</td>\n",
       "      <td>WGS of SARS-CoV-2 R03006_2020_2</td>\n",
       "      <td>SRS6462475</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>National Institute for Communicable Diseases, ...</td>\n",
       "      <td>2020-03-07</td>\n",
       "      <td>Combined nasopharyngeal and oropharyngeal swab</td>\n",
       "      <td>38800</td>\n",
       "      <td>38800</td>\n",
       "      <td>0.01</td>\n",
       "      <td>3</td>\n",
       "      <td>1.587408e+09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Run                                   experiment_title  \\\n",
       "0  SRR11542288              SISPA of SARS-CoV-2 from cell culture   \n",
       "1  SRR11542289              SISPA of SARS-CoV-2 from cell culture   \n",
       "2  SRR11542243  SARS-CoV-2/190300/human/2020/Malaysia _EPI_ISL...   \n",
       "3  SRR11542244  SARS-CoV-2/186197/human/2020/Malaysia_EPI_ISL_...   \n",
       "4  SRR11524818                    WGS of SARS-CoV-2 R03006_2020_2   \n",
       "\n",
       "                                     experiment_desc sample_accession  \\\n",
       "0              SISPA of SARS-CoV-2 from cell culture       SRS6187447   \n",
       "1              SISPA of SARS-CoV-2 from cell culture       SRS6187448   \n",
       "2  SARS-CoV-2/190300/human/2020/Malaysia _EPI_ISL...       SRS6477428   \n",
       "3  SARS-CoV-2/186197/human/2020/Malaysia_EPI_ISL_...       SRS6477427   \n",
       "4                    WGS of SARS-CoV-2 R03006_2020_2       SRS6462475   \n",
       "\n",
       "  library_strategy library_source library_selection LibraryLayout  \\\n",
       "0              WGS    METAGENOMIC        RANDOM PCR        SINGLE   \n",
       "1              WGS    METAGENOMIC        RANDOM PCR        SINGLE   \n",
       "2         AMPLICON      VIRAL RNA               PCR        PAIRED   \n",
       "3         AMPLICON      VIRAL RNA               PCR        PAIRED   \n",
       "4              WGS        GENOMIC            RANDOM        PAIRED   \n",
       "\n",
       "          instrument                                       collected_by  \\\n",
       "0     Illumina MiSeq                                                NaN   \n",
       "1     Illumina MiSeq                                                NaN   \n",
       "2  Illumina iSeq 100             Universiti Malaya COVID Research group   \n",
       "3  Illumina iSeq 100             Universiti Malaya COVID Research group   \n",
       "4     Illumina MiSeq  National Institute for Communicable Diseases, ...   \n",
       "\n",
       "  collection_date                                isolation_source  \\\n",
       "0             NaN                                             NaN   \n",
       "1             NaN                                             NaN   \n",
       "2     22-Mar-2020                      Nasopharyngeal/throat swab   \n",
       "3     14-Mar-2020                      Nasopharyngeal/throat swab   \n",
       "4      2020-03-07  Combined nasopharyngeal and oropharyngeal swab   \n",
       "\n",
       "   total_reads  paired_reads  percent_aligned  aligned_reads  release_date  \n",
       "0       128808        128808            98.60         127004  1.587408e+09  \n",
       "1       187666        187666            98.98         185751  1.587408e+09  \n",
       "2       898380        898380            91.41         821209  1.587408e+09  \n",
       "3       733611        733611            19.18         140706  1.587408e+09  \n",
       "4        38800         38800             0.01              3  1.587408e+09  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Basic\n",
    "PMdb = SRdb[['Run', 'experiment_title', 'experiment_desc', 'sample_accession', \n",
    "             'library_strategy', 'library_source', 'library_selection', 'LibraryLayout', \n",
    "             'instrument', 'collected_by', 'collection_date', 'isolation_source',\n",
    "             'total_reads',\t'paired_reads', 'percent_aligned', 'aligned_reads', 'release_date']]#, 'BioSample']]\n",
    "PMdb.to_csv('/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA_basic_info.csv'.format(d4), index=False)\n",
    "PMdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "425 of 453 samples have >=50 coverage\n",
      "413 of 453 samples have >=0.9 breadth\n",
      "401 of 453 samples have >=50 coverage and >=0.9 breadth\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Center_ID</th>\n",
       "      <th>Group</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Method</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>instrument</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>total_reads</th>\n",
       "      <th>percent_aligned</th>\n",
       "      <th>coverage</th>\n",
       "      <th>breadth</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_desc</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>release_date</th>\n",
       "      <th>Sufficient_cov</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11542243</td>\n",
       "      <td>UM_Malaysia</td>\n",
       "      <td>University Malaya</td>\n",
       "      <td>PRJNA616147</td>\n",
       "      <td>PCR_ARCTIC</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>Illumina iSeq 100</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>898380</td>\n",
       "      <td>91.41</td>\n",
       "      <td>2968.623549</td>\n",
       "      <td>0.989265</td>\n",
       "      <td>SRS6477428</td>\n",
       "      <td>SARS-CoV-2/190300/human/2020/Malaysia _EPI_ISL...</td>\n",
       "      <td>22-Mar-2020</td>\n",
       "      <td>Nasopharyngeal/throat swab</td>\n",
       "      <td>1.587408e+09</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11542244</td>\n",
       "      <td>UM_Malaysia</td>\n",
       "      <td>University Malaya</td>\n",
       "      <td>PRJNA616147</td>\n",
       "      <td>PCR_ARCTIC</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>Illumina iSeq 100</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>733611</td>\n",
       "      <td>19.18</td>\n",
       "      <td>436.001371</td>\n",
       "      <td>0.790623</td>\n",
       "      <td>SRS6477427</td>\n",
       "      <td>SARS-CoV-2/186197/human/2020/Malaysia_EPI_ISL_...</td>\n",
       "      <td>14-Mar-2020</td>\n",
       "      <td>Nasopharyngeal/throat swab</td>\n",
       "      <td>1.587408e+09</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11524818</td>\n",
       "      <td>NICD_SouthAfrica</td>\n",
       "      <td>National Institute for Communicable Diseases</td>\n",
       "      <td>PRJNA624358</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>38800</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.034813</td>\n",
       "      <td>0.034813</td>\n",
       "      <td>SRS6462475</td>\n",
       "      <td>WGS of SARS-CoV-2 R03006_2020_2</td>\n",
       "      <td>2020-03-07</td>\n",
       "      <td>Combined nasopharyngeal and oropharyngeal swab</td>\n",
       "      <td>1.587408e+09</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11514749</td>\n",
       "      <td>UM_Malaysia</td>\n",
       "      <td>University Malaya</td>\n",
       "      <td>PRJNA616147</td>\n",
       "      <td>PCR_ARCTIC</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>Illumina iSeq 100</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>1028510</td>\n",
       "      <td>94.68</td>\n",
       "      <td>3470.610942</td>\n",
       "      <td>0.997726</td>\n",
       "      <td>SRS6395996</td>\n",
       "      <td>SARS-CoV-2/188407/human/2020/Malaysia_EPI_ISL_...</td>\n",
       "      <td>18-Mar-2020</td>\n",
       "      <td>Nasopharyngeal/throat swab</td>\n",
       "      <td>1.587408e+09</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR11514750</td>\n",
       "      <td>UM_Malaysia</td>\n",
       "      <td>University Malaya</td>\n",
       "      <td>PRJNA616147</td>\n",
       "      <td>PCR_ARCTIC</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>Illumina iSeq 100</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>947580</td>\n",
       "      <td>95.87</td>\n",
       "      <td>3274.178276</td>\n",
       "      <td>0.997392</td>\n",
       "      <td>SRS6395995</td>\n",
       "      <td>SARS-CoV-2/189332/human/2020/Malaysia_EPI_ISL_...</td>\n",
       "      <td>20-Mar-2020</td>\n",
       "      <td>Nasopharyngeal/throat swab</td>\n",
       "      <td>1.587408e+09</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Run         Center_ID  \\\n",
       "0  SRR11542243       UM_Malaysia   \n",
       "1  SRR11542244       UM_Malaysia   \n",
       "2  SRR11524818  NICD_SouthAfrica   \n",
       "3  SRR11514749       UM_Malaysia   \n",
       "4  SRR11514750       UM_Malaysia   \n",
       "\n",
       "                                          Group   BioProject      Method  \\\n",
       "0                             University Malaya  PRJNA616147  PCR_ARCTIC   \n",
       "1                             University Malaya  PRJNA616147  PCR_ARCTIC   \n",
       "2  National Institute for Communicable Diseases  PRJNA624358     RNA-seq   \n",
       "3                             University Malaya  PRJNA616147  PCR_ARCTIC   \n",
       "4                             University Malaya  PRJNA616147  PCR_ARCTIC   \n",
       "\n",
       "  library_strategy library_source         instrument LibraryLayout  \\\n",
       "0         AMPLICON      VIRAL RNA  Illumina iSeq 100        PAIRED   \n",
       "1         AMPLICON      VIRAL RNA  Illumina iSeq 100        PAIRED   \n",
       "2              WGS        GENOMIC     Illumina MiSeq        PAIRED   \n",
       "3         AMPLICON      VIRAL RNA  Illumina iSeq 100        PAIRED   \n",
       "4         AMPLICON      VIRAL RNA  Illumina iSeq 100        PAIRED   \n",
       "\n",
       "   total_reads  percent_aligned     coverage   breadth sample_accession  \\\n",
       "0       898380            91.41  2968.623549  0.989265       SRS6477428   \n",
       "1       733611            19.18   436.001371  0.790623       SRS6477427   \n",
       "2        38800             0.01     0.034813  0.034813       SRS6462475   \n",
       "3      1028510            94.68  3470.610942  0.997726       SRS6395996   \n",
       "4       947580            95.87  3274.178276  0.997392       SRS6395995   \n",
       "\n",
       "                                     experiment_desc collection_date  \\\n",
       "0  SARS-CoV-2/190300/human/2020/Malaysia _EPI_ISL...     22-Mar-2020   \n",
       "1  SARS-CoV-2/186197/human/2020/Malaysia_EPI_ISL_...     14-Mar-2020   \n",
       "2                    WGS of SARS-CoV-2 R03006_2020_2      2020-03-07   \n",
       "3  SARS-CoV-2/188407/human/2020/Malaysia_EPI_ISL_...     18-Mar-2020   \n",
       "4  SARS-CoV-2/189332/human/2020/Malaysia_EPI_ISL_...     20-Mar-2020   \n",
       "\n",
       "                                 isolation_source  release_date  \\\n",
       "0                      Nasopharyngeal/throat swab  1.587408e+09   \n",
       "1                      Nasopharyngeal/throat swab  1.587408e+09   \n",
       "2  Combined nasopharyngeal and oropharyngeal swab  1.587408e+09   \n",
       "3                      Nasopharyngeal/throat swab  1.587408e+09   \n",
       "4                      Nasopharyngeal/throat swab  1.587408e+09   \n",
       "\n",
       "   Sufficient_cov  \n",
       "0            True  \n",
       "1           False  \n",
       "2           False  \n",
       "3            True  \n",
       "4            True  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parsed\n",
    "\n",
    "PLdb = SRdb.copy()\n",
    "\n",
    "# Add from BioMeta\n",
    "PLdb = pd.merge(PLdb, BioMeta[['BioProject', 'Group', 'Method', 'Center_ID']], on='BioProject', how='left')\n",
    "\n",
    "VARIABLES = ['Center_ID', 'Group', 'BioProject', 'Method', 'library_strategy', 'library_source', 'instrument', 'LibraryLayout']\n",
    "for v in VARIABLES:\n",
    "    if len(PLdb[PLdb[v].isna()]) > 0:\n",
    "        print(\"{0} samples have no {1}\".format(len(PLdb[PLdb[v].isna()]), v))\n",
    "        \n",
    "# Remove irrelevant samples\n",
    "IRR = set(RunMeta[RunMeta['Status'] == 'NonPatient']['Run'].tolist())\n",
    "PLdb = PLdb[~PLdb['Run'].isin(IRR)]\n",
    "PLdb = PLdb[PLdb['Method'] != 'IRRELEVANT']\n",
    "\n",
    "# Adjust method\n",
    "r2m = RunMeta.set_index('Run')['Method'].to_dict()\n",
    "PLdb['Method'] = [r2m[r] if r in r2m else m for r, m in zip(PLdb['Run'], PLdb['Method'])]\n",
    "\n",
    "# Add basic instrain\n",
    "dbs = []\n",
    "for i, row in PLdb[PLdb['inStrain_succeeded'] == True].iterrows():\n",
    "    db = pd.read_csv(row['inStrain_coverage'], sep='\\t')\n",
    "    db['Run'] = row['Run']\n",
    "    dbs.append(db)\n",
    "COdb = pd.concat(dbs).reset_index(drop=True)\n",
    "\n",
    "PLdb = pd.merge(PLdb, COdb[['Run', 'coverage', 'breadth']], how='left', on='Run')\n",
    "assert len(PLdb[(PLdb['coverage'].isna()) & (PLdb['aligned_reads'] > 0)]) <= 1\n",
    "\n",
    "# Filter\n",
    "DESC_VARS = ['sample_accession', 'experiment_desc', 'collection_date', 'isolation_source', 'release_date']\n",
    "INFO_VARS = ['total_reads', 'percent_aligned', 'coverage', 'breadth']\n",
    "PLdb = PLdb[['Run'] + VARIABLES + INFO_VARS + DESC_VARS]\n",
    "\n",
    "# Print a little description\n",
    "COV_LIM = 50\n",
    "BRE_LIM = 0.9\n",
    "\n",
    "fdb = PLdb[(PLdb['coverage'] >= COV_LIM)]\n",
    "print(\"{0} of {1} samples have >={2} coverage\".format(len(fdb), len(PLdb), COV_LIM, BRE_LIM))\n",
    "\n",
    "fdb = PLdb[(PLdb['breadth'] >= BRE_LIM)]\n",
    "print(\"{0} of {1} samples have >={3} breadth\".format(len(fdb), len(PLdb), COV_LIM, BRE_LIM))\n",
    "\n",
    "fdb = PLdb[(PLdb['coverage'] >= COV_LIM) & (PLdb['breadth'] >= BRE_LIM)]\n",
    "print(\"{0} of {1} samples have >={2} coverage and >={3} breadth\".format(len(fdb), len(PLdb), COV_LIM, BRE_LIM))\n",
    "fdb['Center_ID'].value_counts()\n",
    "\n",
    "PLdb['Sufficient_cov'] = [((coverage >= COV_LIM) & (breadth >= BRE_LIM)) for coverage, breadth in zip(PLdb['coverage'], PLdb['breadth'])]\n",
    "PLdb.to_csv('/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA_parsed_info_v2.csv'.format(d4), index=False)\n",
    "PLdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Peter Doherty Institute for Infection and Immunity;319\n",
      "Utah Public Health Laboratory;55\n",
      "CDC Pathogen Discovery Team;9\n",
      "Hubei Provincial Center for Disease Control and Prevention;6\n",
      "University Malaya;3\n",
      "Beijing Institute of Genomics, Chinese Academy of Sciences;2\n",
      "University of Maryland Institute for Genome Sciences (UMIGS);2\n",
      "Peruvian National Institute of Health;1\n",
      "Wuhan Institute of Virology, Chinese Academy of Sciences;1\n",
      "The Scripps Research Institute;1\n",
      "Universidad Tecnologica de Pereira, Nepal;1\n",
      "University of Wisconsin - Madison;1\n"
     ]
    }
   ],
   "source": [
    "for m, c in PLdb[PLdb['Sufficient_cov'] == True]['Group'].value_counts().to_dict().items():\n",
    "    print(\"{0};{1}\".format(m, c))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PCR_ARCTIC;378\n",
      "RNA-seq;12\n",
      "PCR_CUSTOM;10\n",
      "Capture_Twist;1\n"
     ]
    }
   ],
   "source": [
    "for m, c in PLdb[PLdb['Sufficient_cov'] == True]['Method'].value_counts().to_dict().items():\n",
    "    print(\"{0};{1}\".format(m, c))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
