{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import scipy\n",
    "import sklearn\n",
    "import matplotlib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from collections import defaultdict\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set_style('whitegrid')\n",
    "pd.set_option('display.max_rows', 100)\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "pd.set_option('display.max_columns', 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [],
   "source": [
    "PMdb = pd.read_csv('/home/mattolm/user_data/Covid_19/datatables/SRA_metadata_v1.csv')\n",
    "PLdb = pd.read_csv('/home/mattolm/user_data/Covid_19/datatables/COVID_local_locations_v2.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True    36\n",
       "Name: inStrain_succeeded, dtype: int64"
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PLdb['inStrain_succeeded'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and save inStrain genome coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome</th>\n",
       "      <th>true_length</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>BiAllelic_SNPs</th>\n",
       "      <th>MultiAllelic_SNPs</th>\n",
       "      <th>consensus_SNPs</th>\n",
       "      <th>population_SNPs</th>\n",
       "      <th>breadth</th>\n",
       "      <th>coverage</th>\n",
       "      <th>std_cov</th>\n",
       "      <th>mean_clonality</th>\n",
       "      <th>rarefied_mean_microdiversity</th>\n",
       "      <th>conANI</th>\n",
       "      <th>popANI</th>\n",
       "      <th>unmaskedBreadth</th>\n",
       "      <th>expected_breadth</th>\n",
       "      <th>Run</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT039887.1</td>\n",
       "      <td>29903</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.998060</td>\n",
       "      <td>51015.203658</td>\n",
       "      <td>16355.048169</td>\n",
       "      <td>0.994172</td>\n",
       "      <td>0.005656</td>\n",
       "      <td>0.999966</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.997392</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SRR11177792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT039887.1</td>\n",
       "      <td>29903</td>\n",
       "      <td>25</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>0.993613</td>\n",
       "      <td>242.371066</td>\n",
       "      <td>104.383155</td>\n",
       "      <td>0.998488</td>\n",
       "      <td>0.001427</td>\n",
       "      <td>0.999798</td>\n",
       "      <td>0.999832</td>\n",
       "      <td>0.992977</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SRR11278092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT039887.1</td>\n",
       "      <td>29903</td>\n",
       "      <td>19</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>0.999799</td>\n",
       "      <td>1430.516771</td>\n",
       "      <td>591.378209</td>\n",
       "      <td>0.998533</td>\n",
       "      <td>0.001413</td>\n",
       "      <td>0.999799</td>\n",
       "      <td>0.999833</td>\n",
       "      <td>0.999799</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SRR11278091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT039887.1</td>\n",
       "      <td>29903</td>\n",
       "      <td>111</td>\n",
       "      <td>106</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>0.979166</td>\n",
       "      <td>56.407819</td>\n",
       "      <td>32.761759</td>\n",
       "      <td>0.998343</td>\n",
       "      <td>0.001674</td>\n",
       "      <td>0.999761</td>\n",
       "      <td>0.999829</td>\n",
       "      <td>0.977661</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SRR11278090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT039887.1</td>\n",
       "      <td>29903</td>\n",
       "      <td>192</td>\n",
       "      <td>184</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.989232</td>\n",
       "      <td>80.702438</td>\n",
       "      <td>61.107891</td>\n",
       "      <td>0.997692</td>\n",
       "      <td>0.002322</td>\n",
       "      <td>0.999831</td>\n",
       "      <td>0.999831</td>\n",
       "      <td>0.988496</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SRR11278168</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       genome  true_length  SNPs  BiAllelic_SNPs  MultiAllelic_SNPs  \\\n",
       "0  MT039887.1        29903    23              22                  1   \n",
       "1  MT039887.1        29903    25              20                  0   \n",
       "2  MT039887.1        29903    19              14                  0   \n",
       "3  MT039887.1        29903   111             106                  0   \n",
       "4  MT039887.1        29903   192             184                  3   \n",
       "\n",
       "   consensus_SNPs  population_SNPs   breadth      coverage       std_cov  \\\n",
       "0               1                0  0.998060  51015.203658  16355.048169   \n",
       "1               6                5  0.993613    242.371066    104.383155   \n",
       "2               6                5  0.999799   1430.516771    591.378209   \n",
       "3               7                5  0.979166     56.407819     32.761759   \n",
       "4               5                5  0.989232     80.702438     61.107891   \n",
       "\n",
       "   mean_clonality  rarefied_mean_microdiversity    conANI    popANI  \\\n",
       "0        0.994172                      0.005656  0.999966  1.000000   \n",
       "1        0.998488                      0.001427  0.999798  0.999832   \n",
       "2        0.998533                      0.001413  0.999799  0.999833   \n",
       "3        0.998343                      0.001674  0.999761  0.999829   \n",
       "4        0.997692                      0.002322  0.999831  0.999831   \n",
       "\n",
       "   unmaskedBreadth  expected_breadth          Run  \n",
       "0         0.997392               1.0  SRR11177792  \n",
       "1         0.992977               1.0  SRR11278092  \n",
       "2         0.999799               1.0  SRR11278091  \n",
       "3         0.977661               1.0  SRR11278090  \n",
       "4         0.988496               1.0  SRR11278168  "
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import inStrain\n",
    "import inStrain.SNVprofile\n",
    "\n",
    "dbs = []\n",
    "for i, row in PLdb[PLdb['inStrain_succeeded'] == True].iterrows():\n",
    "    db = pd.read_csv(row['inStrain_coverage'], sep='\\t')\n",
    "    db['Run'] = row['Run']\n",
    "    dbs.append(db)\n",
    "\n",
    "\n",
    "COdb = pd.concat(dbs).reset_index(drop=True)\n",
    "COdb['genome'] = 'MT039887.1'\n",
    "for c in ['detected_scaffolds', 'true_scaffolds']:\n",
    "    del COdb[c]\n",
    "    \n",
    "COdb.to_csv('/home/mattolm/user_data/Covid_19/datatables/COVID_genome_coverage_v2.csv', index=False)\n",
    "COdb.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load inStrain positional coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>position</th>\n",
       "      <th>coverage</th>\n",
       "      <th>normalized_coverage</th>\n",
       "      <th>microdiversity</th>\n",
       "      <th>normalized_microdiversity</th>\n",
       "      <th>rarefied_microdiversity</th>\n",
       "      <th>normalized_rarefied_microdiversity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11177792</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3.238322e-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11177792</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3.238322e-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11177792</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3.238322e-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11177792</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3.238322e-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR11177792</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3.238322e-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Run  position  coverage  normalized_coverage  microdiversity  \\\n",
       "0  SRR11177792         0         3         3.238322e-07             NaN   \n",
       "1  SRR11177792         1         3         3.238322e-07             NaN   \n",
       "2  SRR11177792         2         3         3.238322e-07             NaN   \n",
       "3  SRR11177792         3         3         3.238322e-07             NaN   \n",
       "4  SRR11177792         4         3         3.238322e-07             NaN   \n",
       "\n",
       "   normalized_microdiversity  rarefied_microdiversity  \\\n",
       "0                        NaN                      NaN   \n",
       "1                        NaN                      NaN   \n",
       "2                        NaN                      NaN   \n",
       "3                        NaN                      NaN   \n",
       "4                        NaN                      NaN   \n",
       "\n",
       "   normalized_rarefied_microdiversity  \n",
       "0                                 NaN  \n",
       "1                                 NaN  \n",
       "2                                 NaN  \n",
       "3                                 NaN  \n",
       "4                                 NaN  "
      ]
     },
     "execution_count": 234,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TRUE_LENGTH = 29903\n",
    "SCAFFOLD = 'NC_045512.2'\n",
    "bldb = pd.DataFrame(list(range(0, TRUE_LENGTH))).rename(columns={0:'position'})\n",
    "\n",
    "dbs = []\n",
    "for i, row in PLdb[PLdb['inStrain_succeeded'] == True].iterrows():\n",
    "    IS_loc = '/'.join(row['inStrain_coverage'].split('/')[:-2])\n",
    "    IS = inStrain.SNVprofile.SNVprofile(IS_loc)\n",
    "    \n",
    "    # Load positional coverage\n",
    "    covT = IS.get('covT')\n",
    "    cov = inStrain.profileUtilities._mm_counts_to_counts_shrunk(covT[SCAFFOLD])\n",
    "    db = pd.DataFrame(cov).reset_index(drop=False).rename(columns={'index':'position', 0:'coverage'})\n",
    "    db['coverage'] = db['coverage'].astype(int)\n",
    "    \n",
    "    # Load positional microdiversity\n",
    "    covT = IS.get('clonT')\n",
    "    cov = inStrain.plottingUtilities._get_basewise_clons3(covT[SCAFFOLD], 100)\n",
    "    cdb = cov.reset_index(drop=False).rename(columns={'index':'position', 0:'clonality'})\n",
    "    cdb['clonality'] = cdb['clonality'].astype(float)\n",
    "    cdb['microdiversity'] = 1 - cdb['clonality']\n",
    "    cdb['normalized_microdiversity'] = sklearn.preprocessing.normalize(cdb['microdiversity'].values.reshape(-1, 1), axis=0) # Normalize clonality before NaN values are introduced\n",
    "    \n",
    "    # Load rarefied positional microdiversity\n",
    "    covTR = IS.get('clonTR')\n",
    "    cov = inStrain.plottingUtilities._get_basewise_clons3(covTR[SCAFFOLD], 100)\n",
    "    crdb = cov.reset_index(drop=False).rename(columns={'index':'position', 0:'rarefied_clonality'})\n",
    "    crdb['rarefied_clonality'] = crdb['rarefied_clonality'].astype(float)\n",
    "    crdb['rarefied_microdiversity'] = 1 - crdb['rarefied_clonality']\n",
    "    try:\n",
    "        crdb['normalized_rarefied_microdiversity'] = sklearn.preprocessing.normalize(crdb['rarefied_microdiversity'].values.reshape(-1, 1), axis=0) # Normalize clonality before NaN values are introduced\n",
    "    except ValueError:\n",
    "        crdb['normalized_rarefied_microdiversity'] = np.nan\n",
    "    \n",
    "    # Merge\n",
    "    db = pd.merge(bldb, db, on='position', how='outer')\n",
    "    db = pd.merge(db, cdb, on='position', how='outer')\n",
    "    db = pd.merge(db, crdb, on='position', how='outer')\n",
    "    db['Run'] = row['Run']\n",
    "    db = db.sort_values('position')\n",
    "    \n",
    "    # Backfill coverage with 0s and normalize\n",
    "    db['coverage'] = db['coverage'].fillna(0).astype(int)\n",
    "    db['normalized_coverage'] = sklearn.preprocessing.normalize(db['coverage'].values.reshape(-1, 1), axis=0)\n",
    "    \n",
    "    # Add mean and adjust\n",
    "#     db['mean_coverage'] = db['coverage'].mean()\n",
    "#     db['adjusted_coverage'] = db['coverage'] / db['mean_coverage']\n",
    "    \n",
    "#     db['mean_microdiversity'] = db['microdiversity'].mean()\n",
    "#     db['adjusted_microdiversity'] = db['coverage'] / db['mean_microdiversity']\n",
    "    \n",
    "#     db['mean_rarefied_microdiversity'] = db['rarefied_microdiversity'].mean()\n",
    "#     db['adjusted_rarefied_microdiversity'] = db['coverage'] / db['mean_rarefied_microdiversity']\n",
    "    \n",
    "    dbs.append(db)\n",
    "    \n",
    "# Re-order columns\n",
    "Odb = pd.concat(dbs).reset_index(drop=True)\n",
    "order = ['Run', 'position']\n",
    "for item in ['coverage', 'microdiversity', 'rarefied_microdiversity']:\n",
    "    order += [item, 'normalized_' + item]\n",
    "Odb = Odb[order]\n",
    "\n",
    "# Save\n",
    "Odb.to_csv('/home/mattolm/user_data/Covid_19/datatables/COVID_positional_coverage_v2.csv', index=False)\n",
    "Odb.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make new genes table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[265:805](+)\n",
      "[805:2719](+)\n",
      "[2719:8554](+)\n",
      "[8554:10054](+)\n",
      "[10054:10972](+)\n",
      "[10972:11842](+)\n",
      "[11842:12091](+)\n",
      "[12091:12685](+)\n",
      "[12685:13024](+)\n",
      "[13024:13441](+)\n",
      "join{[13441:13468](+), [13467:16236](+)}\n",
      "[16236:18039](+)\n",
      "[18039:19620](+)\n",
      "[19620:20658](+)\n",
      "[20658:21552](+)\n",
      "[13441:13480](+)\n"
     ]
    }
   ],
   "source": [
    "import inStrain\n",
    "import Bio\n",
    "import logging\n",
    "from Bio import SeqIO\n",
    "\n",
    "def parse_genbank_genes_COV(gene_file, gene_name='gene'):\n",
    "    table = defaultdict(list)\n",
    "    gene2sequence = {}\n",
    "    gene2aa = {}\n",
    "    for record in SeqIO.parse(gene_file, 'gb'):\n",
    "        scaffold = record.id\n",
    "        for feature in record.features:\n",
    "            if feature.type == 'CDS':\n",
    "                sequence = feature.location.extract(record).seq\n",
    "                \n",
    "                gene = feature.qualifiers[gene_name][0]\n",
    "                loc = feature.location\n",
    "                if type(loc) is Bio.SeqFeature.CompoundLocation:\n",
    "                    partial = 'compound'\n",
    "                else:\n",
    "                    partial = False\n",
    "\n",
    "                table['gene'].append(gene)\n",
    "                table['scaffold'].append(scaffold)\n",
    "                table['direction'].append(feature.location.strand)\n",
    "                table['translated'].append(True)\n",
    "\n",
    "                table['start'].append(loc.start)\n",
    "                table['end'].append(loc.end)\n",
    "                table['location'].append(feature.location)\n",
    "\n",
    "                #try:\n",
    "                gene2sequence[gene] = sequence\n",
    "                TransSeq = str(sequence.translate()[:-1])\n",
    "                assert TransSeq == str(feature.qualifiers['translation'][0]), [TransSeq, str(feature.qualifiers['translation'][0])]\n",
    "                gene2aa[gene] = TransSeq\n",
    "\n",
    "            elif ('UTR' in feature.type) | ('stem_loop' in feature.type):\n",
    "                loc = feature.location\n",
    "                gene = feature.type\n",
    "                \n",
    "                table['gene'].append(gene)\n",
    "                table['scaffold'].append(scaffold)\n",
    "                table['direction'].append(feature.location.strand)\n",
    "                table['translated'].append(False)\n",
    "\n",
    "                table['start'].append(loc.start)\n",
    "                table['end'].append(loc.end)\n",
    "                table['location'].append(feature.location)\n",
    "                \n",
    "            elif feature.type == 'mat_peptide':\n",
    "                loc = feature.location\n",
    "                gene = feature.qualifiers['protein_id'][0]\n",
    "                sequence = feature.location.extract(record).seq\n",
    "                \n",
    "                print(feature.location)\n",
    "                \n",
    "                table['gene'].append(gene)\n",
    "                table['scaffold'].append(scaffold)\n",
    "                table['direction'].append(feature.location.strand)\n",
    "                table['translated'].append(True)\n",
    "\n",
    "                table['start'].append(loc.start)\n",
    "                table['end'].append(loc.end)\n",
    "                table['location'].append(feature.location)\n",
    "                \n",
    "                gene2sequence[gene] = sequence\n",
    "                TransSeq = str(sequence.translate()[:-1])\n",
    "                #assert TransSeq == str(feature.qualifiers['translation'][0]), [TransSeq, str(feature.qualifiers['translation'][0])]\n",
    "                assert len(sequence) % 3 == 0\n",
    "                gene2aa[gene] = TransSeq\n",
    "                \n",
    "            elif feature.type in ['gene', 'source']:\n",
    "                pass\n",
    "\n",
    "            else:\n",
    "                print(feature.type)\n",
    "\n",
    "    Gdb = pd.DataFrame(table)\n",
    "\n",
    "    return Gdb, gene2sequence, gene2aa, record.seq\n",
    "\n",
    "Gdb, gene2sequence, gene2aa, source = parse_genbank_genes_COV('/home/mattolm/user_data/Covid_19/genomes/NC_045512.2.gb')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Convert to positional"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_positional_gene_table_COV(Gdb, gene2sequence, gene2aa, source):\n",
    "    # Make a table of the genome\n",
    "    table = defaultdict(list)\n",
    "    for i, char in enumerate(source):\n",
    "        table['position'].append(i)\n",
    "        table['ref_seq'].append(char)\n",
    "    Sdb = pd.DataFrame(table)\n",
    "    l2c = Sdb.set_index('position')['ref_seq'].to_dict()\n",
    "    \n",
    "    # Add translations for gene products\n",
    "    table = defaultdict(list)\n",
    "    for j, row in Gdb[Gdb['translated'] == True].iterrows():\n",
    "        \n",
    "        # Skip this  polyprotein; you handle it with the products\n",
    "        if row['gene'] in ['orf1ab']:\n",
    "            continue\n",
    "            \n",
    "        NSEQ = gene2sequence[row['gene']]\n",
    "        location = row['location']\n",
    "        ASEQ = gene2aa[row['gene']]\n",
    "        EXPA = []\n",
    "        for a in ASEQ:\n",
    "            EXPA += [a] * 3\n",
    "        EXPA += ['stop'] * 3\n",
    "        assert len(NSEQ) == len(location) == len(EXPA), [len(NSEQ), len(location), len(EXPA)]\n",
    "\n",
    "        pos = 0\n",
    "        for loc, char, a in zip(location, NSEQ, EXPA):\n",
    "            table['position'].append(loc)\n",
    "            table['protein'].append(row['gene'])\n",
    "            table['orf_position'].append((pos) % 3)\n",
    "            table['aa'].append(a)\n",
    "            table['aa_index'].append(int(pos/3))\n",
    "            assert l2c[loc] == char, [row['gene'],l2c[loc], char]\n",
    "            pos += 1\n",
    "        assert (pos) % 3 == 0\n",
    "    Tdb = pd.DataFrame(table)\n",
    "    #print(Tdb[Tdb.duplicated(subset=['location'], keep=False)].sort_values('location'))\n",
    "        \n",
    "    # Add genes\n",
    "    table = defaultdict(list)\n",
    "    for j, row in Gdb[Gdb['translated'] == False].iterrows():\n",
    "        location = row['location']\n",
    "        for loc in location:\n",
    "            table['position'].append(loc)\n",
    "            table['gene'].append(row['gene'])\n",
    "    gdb = pd.DataFrame(table)\n",
    "    \n",
    "    # Merge\n",
    "    db = pd.merge(Sdb, Tdb, on='position', how='left').merge(gdb, on='position', how='left')\n",
    "    return db\n",
    "        \n",
    "POdb = make_positional_gene_table_COV(Gdb, gene2sequence, gene2aa, source)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save genes stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>direction</th>\n",
       "      <th>translated</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5'UTR</td>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>265</td>\n",
       "      <td>[0:265](+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>orf1ab</td>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>265</td>\n",
       "      <td>21555</td>\n",
       "      <td>join{[265:13468](+), [13467:21555](+)}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YP_009725297.1</td>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>265</td>\n",
       "      <td>805</td>\n",
       "      <td>[265:805](+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YP_009725298.1</td>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>805</td>\n",
       "      <td>2719</td>\n",
       "      <td>[805:2719](+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YP_009725299.1</td>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>2719</td>\n",
       "      <td>8554</td>\n",
       "      <td>[2719:8554](+)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene     scaffold  direction  translated  start    end  \\\n",
       "0           5'UTR  NC_045512.2          1       False      0    265   \n",
       "1          orf1ab  NC_045512.2          1        True    265  21555   \n",
       "2  YP_009725297.1  NC_045512.2          1        True    265    805   \n",
       "3  YP_009725298.1  NC_045512.2          1        True    805   2719   \n",
       "4  YP_009725299.1  NC_045512.2          1        True   2719   8554   \n",
       "\n",
       "                                 location  \n",
       "0                              [0:265](+)  \n",
       "1  join{[265:13468](+), [13467:21555](+)}  \n",
       "2                            [265:805](+)  \n",
       "3                           [805:2719](+)  \n",
       "4                          [2719:8554](+)  "
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GOdb = Gdb.copy()\n",
    "GOdb['location'] = [str(x) for x in GOdb['location']]\n",
    "GOdb.to_csv('/home/mattolm/user_data/Covid_19/datatables/COVID_genes_table_v2.csv', index=False)\n",
    "GOdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>ref_seq</th>\n",
       "      <th>protein</th>\n",
       "      <th>orf_position</th>\n",
       "      <th>aa</th>\n",
       "      <th>aa_index</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   position ref_seq protein  orf_position   aa  aa_index   gene\n",
       "0         0       A     NaN           NaN  NaN       NaN  5'UTR\n",
       "1         1       T     NaN           NaN  NaN       NaN  5'UTR\n",
       "2         2       T     NaN           NaN  NaN       NaN  5'UTR\n",
       "3         3       A     NaN           NaN  NaN       NaN  5'UTR\n",
       "4         4       A     NaN           NaN  NaN       NaN  5'UTR"
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "POdb.to_csv('/home/mattolm/user_data/Covid_19/datatables/COVID_genes_positional_v2.csv', index=False)\n",
    "POdb.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save version 2.2 with CCs names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>direction</th>\n",
       "      <th>translated</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>location</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5'UTR</td>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>265</td>\n",
       "      <td>[0:265](+)</td>\n",
       "      <td>5'UTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>orf1ab</td>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>265</td>\n",
       "      <td>21555</td>\n",
       "      <td>join{[265:13468](+), [13467:21555](+)}</td>\n",
       "      <td>orf1ab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YP_009725297.1</td>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>265</td>\n",
       "      <td>805</td>\n",
       "      <td>[265:805](+)</td>\n",
       "      <td>leader protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YP_009725298.1</td>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>805</td>\n",
       "      <td>2719</td>\n",
       "      <td>[805:2719](+)</td>\n",
       "      <td>nsp2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YP_009725299.1</td>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>2719</td>\n",
       "      <td>8554</td>\n",
       "      <td>[2719:8554](+)</td>\n",
       "      <td>nsp3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene     scaffold  direction  translated  start    end  \\\n",
       "0           5'UTR  NC_045512.2          1       False      0    265   \n",
       "1          orf1ab  NC_045512.2          1        True    265  21555   \n",
       "2  YP_009725297.1  NC_045512.2          1        True    265    805   \n",
       "3  YP_009725298.1  NC_045512.2          1        True    805   2719   \n",
       "4  YP_009725299.1  NC_045512.2          1        True   2719   8554   \n",
       "\n",
       "                                 location            name  \n",
       "0                              [0:265](+)           5'UTR  \n",
       "1  join{[265:13468](+), [13467:21555](+)}          orf1ab  \n",
       "2                            [265:805](+)  leader protein  \n",
       "3                           [805:2719](+)            nsp2  \n",
       "4                          [2719:8554](+)            nsp3  "
      ]
     },
     "execution_count": 276,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Do it with CCs names\n",
    "Ndb = pd.read_csv('/home/mattolm/user_data/Covid_19/datatables/COVID_genes_table_v2.tsv', sep='\\t')\n",
    "\n",
    "GOdb = Gdb.copy()\n",
    "GOdb['name'] = GOdb['gene'].map(Ndb.set_index('gene')['name'].to_dict())\n",
    "GOdb.at[17, 'name'] = 'orf1a'\n",
    "\n",
    "GOdb['location'] = [str(x) for x in GOdb['location']]\n",
    "GOdb.to_csv('/home/mattolm/user_data/Covid_19/datatables/COVID_genes_table_v2.2.csv', index=False)\n",
    "GOdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>ref_seq</th>\n",
       "      <th>protein</th>\n",
       "      <th>orf_position</th>\n",
       "      <th>aa</th>\n",
       "      <th>aa_index</th>\n",
       "      <th>gene</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5'UTR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   position ref_seq protein  orf_position   aa  aa_index   gene protein_name  \\\n",
       "0         0       A     NaN           NaN  NaN       NaN  5'UTR          NaN   \n",
       "1         1       T     NaN           NaN  NaN       NaN  5'UTR          NaN   \n",
       "2         2       T     NaN           NaN  NaN       NaN  5'UTR          NaN   \n",
       "3         3       A     NaN           NaN  NaN       NaN  5'UTR          NaN   \n",
       "4         4       A     NaN           NaN  NaN       NaN  5'UTR          NaN   \n",
       "\n",
       "  gene_name  \n",
       "0     5'UTR  \n",
       "1     5'UTR  \n",
       "2     5'UTR  \n",
       "3     5'UTR  \n",
       "4     5'UTR  "
      ]
     },
     "execution_count": 282,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "POdb2 = POdb.copy()\n",
    "POdb2['protein_name'] = POdb2['protein'].map(GOdb.set_index('gene')['name'].to_dict())\n",
    "POdb2['gene_name'] = POdb2['gene'].map(GOdb.set_index('gene')['name'].to_dict())\n",
    "\n",
    "POdb2.to_csv('/home/mattolm/user_data/Covid_19/datatables/COVID_genes_positional_v2.2.csv', index=False)\n",
    "POdb2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
