{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import scipy\n",
    "import sklearn\n",
    "import matplotlib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from collections import defaultdict\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set_style('whitegrid')\n",
    "pd.set_option('display.max_rows', 100)\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "pd.set_option('display.max_columns', 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATE = '04202020'\n",
    "PLdb = pd.read_csv('/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA_full_info.csv'.format(DATE))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     449\n",
       "False      1\n",
       "Name: inStrain_succeeded, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PLdb = PLdb[PLdb['aligned_reads'] > 0]\n",
    "PLdb['inStrain_succeeded'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PAIRED    440\n",
       "SINGLE     10\n",
       "Name: LibraryLayout, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PLdb['LibraryLayout'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adjust metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "saveloc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_tables/'.format(DATE)\n",
    "if not os.path.isdir(saveloc):\n",
    "    os.mkdir(saveloc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and save coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>length</th>\n",
       "      <th>breadth</th>\n",
       "      <th>coverage</th>\n",
       "      <th>median_cov</th>\n",
       "      <th>std_cov</th>\n",
       "      <th>bases_w_0_coverage</th>\n",
       "      <th>mean_clonality</th>\n",
       "      <th>median_clonality</th>\n",
       "      <th>mean_microdiversity</th>\n",
       "      <th>median_microdiversity</th>\n",
       "      <th>rarefied_mean_microdiversity</th>\n",
       "      <th>rarefied_median_microdiversity</th>\n",
       "      <th>unmaskedBreadth</th>\n",
       "      <th>rarefied_breadth</th>\n",
       "      <th>expected_breadth</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>Reference_SNPs</th>\n",
       "      <th>BiAllelic_SNPs</th>\n",
       "      <th>MultiAllelic_SNPs</th>\n",
       "      <th>consensus_SNPs</th>\n",
       "      <th>population_SNPs</th>\n",
       "      <th>conANI</th>\n",
       "      <th>popANI</th>\n",
       "      <th>Run</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.999131</td>\n",
       "      <td>969.343812</td>\n",
       "      <td>812</td>\n",
       "      <td>807.972740</td>\n",
       "      <td>26</td>\n",
       "      <td>0.991916</td>\n",
       "      <td>0.995506</td>\n",
       "      <td>0.008084</td>\n",
       "      <td>0.004494</td>\n",
       "      <td>0.007682</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.997458</td>\n",
       "      <td>0.991974</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>64</td>\n",
       "      <td>2</td>\n",
       "      <td>50</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>0.999799</td>\n",
       "      <td>0.999899</td>\n",
       "      <td>SRR11542288</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.999064</td>\n",
       "      <td>1372.825770</td>\n",
       "      <td>1180</td>\n",
       "      <td>1062.479273</td>\n",
       "      <td>28</td>\n",
       "      <td>0.991800</td>\n",
       "      <td>0.995488</td>\n",
       "      <td>0.008200</td>\n",
       "      <td>0.004512</td>\n",
       "      <td>0.007680</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.998027</td>\n",
       "      <td>0.995519</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>91</td>\n",
       "      <td>2</td>\n",
       "      <td>78</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0.999866</td>\n",
       "      <td>0.999899</td>\n",
       "      <td>SRR11542289</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.989265</td>\n",
       "      <td>2968.623549</td>\n",
       "      <td>1286</td>\n",
       "      <td>3938.030320</td>\n",
       "      <td>321</td>\n",
       "      <td>0.998092</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.001908</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001493</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.954754</td>\n",
       "      <td>0.897669</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>59</td>\n",
       "      <td>6</td>\n",
       "      <td>51</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>0.999650</td>\n",
       "      <td>0.999790</td>\n",
       "      <td>SRR11542243</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.790623</td>\n",
       "      <td>436.001371</td>\n",
       "      <td>202</td>\n",
       "      <td>632.044924</td>\n",
       "      <td>6261</td>\n",
       "      <td>0.998721</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.001279</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001078</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.750694</td>\n",
       "      <td>0.654416</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>26</td>\n",
       "      <td>2</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0.999822</td>\n",
       "      <td>0.999911</td>\n",
       "      <td>SRR11542244</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.034813</td>\n",
       "      <td>0.034813</td>\n",
       "      <td>0</td>\n",
       "      <td>0.183305</td>\n",
       "      <td>28862</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030272</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>SRR11524818</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      scaffold  length   breadth     coverage  median_cov      std_cov  \\\n",
       "0  NC_045512.2   29903  0.999131   969.343812         812   807.972740   \n",
       "1  NC_045512.2   29903  0.999064  1372.825770        1180  1062.479273   \n",
       "2  NC_045512.2   29903  0.989265  2968.623549        1286  3938.030320   \n",
       "3  NC_045512.2   29903  0.790623   436.001371         202   632.044924   \n",
       "4  NC_045512.2   29903  0.034813     0.034813           0     0.183305   \n",
       "\n",
       "   bases_w_0_coverage  mean_clonality  median_clonality  mean_microdiversity  \\\n",
       "0                  26        0.991916          0.995506             0.008084   \n",
       "1                  28        0.991800          0.995488             0.008200   \n",
       "2                 321        0.998092          1.000000             0.001908   \n",
       "3                6261        0.998721          1.000000             0.001279   \n",
       "4               28862             NaN               NaN                  NaN   \n",
       "\n",
       "   median_microdiversity  rarefied_mean_microdiversity  \\\n",
       "0               0.004494                      0.007682   \n",
       "1               0.004512                      0.007680   \n",
       "2               0.000000                      0.001493   \n",
       "3               0.000000                      0.001078   \n",
       "4                    NaN                           NaN   \n",
       "\n",
       "   rarefied_median_microdiversity  unmaskedBreadth  rarefied_breadth  \\\n",
       "0                             0.0         0.997458          0.991974   \n",
       "1                             0.0         0.998027          0.995519   \n",
       "2                             0.0         0.954754          0.897669   \n",
       "3                             0.0         0.750694          0.654416   \n",
       "4                             NaN         0.000000          0.000000   \n",
       "\n",
       "   expected_breadth  SNPs  Reference_SNPs  BiAllelic_SNPs  MultiAllelic_SNPs  \\\n",
       "0          1.000000    64               2              50                 12   \n",
       "1          1.000000    91               2              78                 11   \n",
       "2          1.000000    59               6              51                  2   \n",
       "3          1.000000    26               2              24                  0   \n",
       "4          0.030272     0               0               0                  0   \n",
       "\n",
       "   consensus_SNPs  population_SNPs    conANI    popANI          Run  \\\n",
       "0               6                3  0.999799  0.999899  SRR11542288   \n",
       "1               4                3  0.999866  0.999899  SRR11542289   \n",
       "2              10                6  0.999650  0.999790  SRR11542243   \n",
       "3               4                2  0.999822  0.999911  SRR11542244   \n",
       "4               0                0  0.000000  0.000000  SRR11524818   \n",
       "\n",
       "       genome  \n",
       "0  MT039887.1  \n",
       "1  MT039887.1  \n",
       "2  MT039887.1  \n",
       "3  MT039887.1  \n",
       "4  MT039887.1  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import inStrain\n",
    "import inStrain.SNVprofile\n",
    "\n",
    "dbs = []\n",
    "for i, row in PLdb[PLdb['inStrain_succeeded'] == True].iterrows():\n",
    "    db = pd.read_csv(row['inStrain_coverage'], sep='\\t')\n",
    "    db['Run'] = row['Run']\n",
    "    dbs.append(db)\n",
    "\n",
    "COdb = pd.concat(dbs).reset_index(drop=True)\n",
    "COdb['genome'] = 'MT039887.1'\n",
    "# for c in ['detected_scaffolds', 'true_scaffolds']:\n",
    "#     del COdb[c]\n",
    "    \n",
    "COdb.to_csv(saveloc + 'COVID_genome_coverage.csv', index=False)\n",
    "COdb.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>length</th>\n",
       "      <th>breadth</th>\n",
       "      <th>coverage</th>\n",
       "      <th>median_cov</th>\n",
       "      <th>std_cov</th>\n",
       "      <th>bases_w_0_coverage</th>\n",
       "      <th>mean_clonality</th>\n",
       "      <th>median_clonality</th>\n",
       "      <th>mean_microdiversity</th>\n",
       "      <th>median_microdiversity</th>\n",
       "      <th>rarefied_mean_microdiversity</th>\n",
       "      <th>rarefied_median_microdiversity</th>\n",
       "      <th>unmaskedBreadth</th>\n",
       "      <th>rarefied_breadth</th>\n",
       "      <th>expected_breadth</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>Reference_SNPs</th>\n",
       "      <th>BiAllelic_SNPs</th>\n",
       "      <th>MultiAllelic_SNPs</th>\n",
       "      <th>consensus_SNPs</th>\n",
       "      <th>population_SNPs</th>\n",
       "      <th>conANI</th>\n",
       "      <th>popANI</th>\n",
       "      <th>Run</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.999131</td>\n",
       "      <td>969.343812</td>\n",
       "      <td>812</td>\n",
       "      <td>807.972740</td>\n",
       "      <td>26</td>\n",
       "      <td>0.991916</td>\n",
       "      <td>0.995506</td>\n",
       "      <td>0.008084</td>\n",
       "      <td>0.004494</td>\n",
       "      <td>0.007682</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.997458</td>\n",
       "      <td>0.991974</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>64</td>\n",
       "      <td>2</td>\n",
       "      <td>50</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>0.999799</td>\n",
       "      <td>0.999899</td>\n",
       "      <td>SRR11542288</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.999064</td>\n",
       "      <td>1372.825770</td>\n",
       "      <td>1180</td>\n",
       "      <td>1062.479273</td>\n",
       "      <td>28</td>\n",
       "      <td>0.991800</td>\n",
       "      <td>0.995488</td>\n",
       "      <td>0.008200</td>\n",
       "      <td>0.004512</td>\n",
       "      <td>0.007680</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.998027</td>\n",
       "      <td>0.995519</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>91</td>\n",
       "      <td>2</td>\n",
       "      <td>78</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0.999866</td>\n",
       "      <td>0.999899</td>\n",
       "      <td>SRR11542289</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.989265</td>\n",
       "      <td>2968.623549</td>\n",
       "      <td>1286</td>\n",
       "      <td>3938.030320</td>\n",
       "      <td>321</td>\n",
       "      <td>0.998092</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.001908</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001493</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.954754</td>\n",
       "      <td>0.897669</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>59</td>\n",
       "      <td>6</td>\n",
       "      <td>51</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>0.999650</td>\n",
       "      <td>0.999790</td>\n",
       "      <td>SRR11542243</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.790623</td>\n",
       "      <td>436.001371</td>\n",
       "      <td>202</td>\n",
       "      <td>632.044924</td>\n",
       "      <td>6261</td>\n",
       "      <td>0.998721</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.001279</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001078</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.750694</td>\n",
       "      <td>0.654416</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>26</td>\n",
       "      <td>2</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0.999822</td>\n",
       "      <td>0.999911</td>\n",
       "      <td>SRR11542244</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.034813</td>\n",
       "      <td>0.034813</td>\n",
       "      <td>0</td>\n",
       "      <td>0.183305</td>\n",
       "      <td>28862</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.030272</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>SRR11524818</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      scaffold  length   breadth     coverage  median_cov      std_cov  \\\n",
       "0  NC_045512.2   29903  0.999131   969.343812         812   807.972740   \n",
       "1  NC_045512.2   29903  0.999064  1372.825770        1180  1062.479273   \n",
       "2  NC_045512.2   29903  0.989265  2968.623549        1286  3938.030320   \n",
       "3  NC_045512.2   29903  0.790623   436.001371         202   632.044924   \n",
       "4  NC_045512.2   29903  0.034813     0.034813           0     0.183305   \n",
       "\n",
       "   bases_w_0_coverage  mean_clonality  median_clonality  mean_microdiversity  \\\n",
       "0                  26        0.991916          0.995506             0.008084   \n",
       "1                  28        0.991800          0.995488             0.008200   \n",
       "2                 321        0.998092          1.000000             0.001908   \n",
       "3                6261        0.998721          1.000000             0.001279   \n",
       "4               28862             NaN               NaN                  NaN   \n",
       "\n",
       "   median_microdiversity  rarefied_mean_microdiversity  \\\n",
       "0               0.004494                      0.007682   \n",
       "1               0.004512                      0.007680   \n",
       "2               0.000000                      0.001493   \n",
       "3               0.000000                      0.001078   \n",
       "4                    NaN                           NaN   \n",
       "\n",
       "   rarefied_median_microdiversity  unmaskedBreadth  rarefied_breadth  \\\n",
       "0                             0.0         0.997458          0.991974   \n",
       "1                             0.0         0.998027          0.995519   \n",
       "2                             0.0         0.954754          0.897669   \n",
       "3                             0.0         0.750694          0.654416   \n",
       "4                             NaN         0.000000          0.000000   \n",
       "\n",
       "   expected_breadth  SNPs  Reference_SNPs  BiAllelic_SNPs  MultiAllelic_SNPs  \\\n",
       "0          1.000000    64               2              50                 12   \n",
       "1          1.000000    91               2              78                 11   \n",
       "2          1.000000    59               6              51                  2   \n",
       "3          1.000000    26               2              24                  0   \n",
       "4          0.030272     0               0               0                  0   \n",
       "\n",
       "   consensus_SNPs  population_SNPs    conANI    popANI          Run  \\\n",
       "0               6                3  0.999799  0.999899  SRR11542288   \n",
       "1               4                3  0.999866  0.999899  SRR11542289   \n",
       "2              10                6  0.999650  0.999790  SRR11542243   \n",
       "3               4                2  0.999822  0.999911  SRR11542244   \n",
       "4               0                0  0.000000  0.000000  SRR11524818   \n",
       "\n",
       "       genome  \n",
       "0  MT039887.1  \n",
       "1  MT039887.1  \n",
       "2  MT039887.1  \n",
       "3  MT039887.1  \n",
       "4  MT039887.1  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import inStrain\n",
    "import inStrain.SNVprofile\n",
    "\n",
    "dbs = []\n",
    "for i, row in PLdb[PLdb['inStrain_succeeded'] == True].iterrows():\n",
    "    loc = row['inStrain_coverage'].replace('_genomeWide_scaffold_info.tsv', '_SNVs.tsv')\n",
    "    try:\n",
    "        db = pd.read_csv(loc, sep='\\t')\n",
    "        db['Run'] = row['Run']\n",
    "        dbs.append(db)\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "\n",
    "Sdb = pd.concat(dbs).reset_index(drop=True)\n",
    "Sdb['genome'] = 'MT039887.1'\n",
    "\n",
    "Sdb.to_csv(saveloc + 'COVID_raw_SNVs.csv', index=False)\n",
    "Sdb.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and save positional coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>position</th>\n",
       "      <th>coverage</th>\n",
       "      <th>normalized_coverage</th>\n",
       "      <th>microdiversity</th>\n",
       "      <th>normalized_microdiversity</th>\n",
       "      <th>rarefied_microdiversity</th>\n",
       "      <th>normalized_rarefied_microdiversity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11542288</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11542288</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11542288</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11542288</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR11542288</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Run  position  coverage  normalized_coverage  microdiversity  \\\n",
       "0  SRR11542288         0         3             0.000014             NaN   \n",
       "1  SRR11542288         1         3             0.000014             NaN   \n",
       "2  SRR11542288         2         3             0.000014             NaN   \n",
       "3  SRR11542288         3         3             0.000014             NaN   \n",
       "4  SRR11542288         4         3             0.000014             NaN   \n",
       "\n",
       "   normalized_microdiversity  rarefied_microdiversity  \\\n",
       "0                        NaN                      NaN   \n",
       "1                        NaN                      NaN   \n",
       "2                        NaN                      NaN   \n",
       "3                        NaN                      NaN   \n",
       "4                        NaN                      NaN   \n",
       "\n",
       "   normalized_rarefied_microdiversity  \n",
       "0                                 NaN  \n",
       "1                                 NaN  \n",
       "2                                 NaN  \n",
       "3                                 NaN  \n",
       "4                                 NaN  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TRUE_LENGTH = 29903\n",
    "SCAFFOLD = 'NC_045512.2'\n",
    "bldb = pd.DataFrame(list(range(0, TRUE_LENGTH))).rename(columns={0:'position'})\n",
    "\n",
    "dbs = []\n",
    "for i, row in PLdb[PLdb['inStrain_succeeded'] == True].iterrows():\n",
    "    IS_loc = '/'.join(row['inStrain_coverage'].split('/')[:-2])\n",
    "    IS = inStrain.SNVprofile.SNVprofile(IS_loc)\n",
    "    \n",
    "    # Load positional coverage\n",
    "    covT = IS.get('covT')\n",
    "    cov = inStrain.profileUtilities._mm_counts_to_counts_shrunk(covT[SCAFFOLD])\n",
    "    db = pd.DataFrame(cov).reset_index(drop=False).rename(columns={'index':'position', 0:'coverage'})\n",
    "    db['coverage'] = db['coverage'].astype(int)\n",
    "    \n",
    "    # Load positional microdiversity\n",
    "    covT = IS.get('clonT')\n",
    "    cov = inStrain.plottingUtilities._get_basewise_clons3(covT[SCAFFOLD], 100)\n",
    "    cdb = cov.reset_index(drop=False).rename(columns={'index':'position', 0:'clonality'})\n",
    "    cdb['clonality'] = cdb['clonality'].astype(float)\n",
    "    cdb['microdiversity'] = 1 - cdb['clonality']\n",
    "    if len(cdb) != 0:\n",
    "        cdb['normalized_microdiversity'] = sklearn.preprocessing.normalize(cdb['microdiversity'].values.reshape(-1, 1), axis=0) # Normalize clonality before NaN values are introduced\n",
    "    else:\n",
    "        cdb['normalized_microdiversity'] = np.nan\n",
    "    \n",
    "    # Load rarefied positional microdiversity\n",
    "    covTR = IS.get('clonTR')\n",
    "    cov = inStrain.plottingUtilities._get_basewise_clons3(covTR[SCAFFOLD], 100)\n",
    "    crdb = cov.reset_index(drop=False).rename(columns={'index':'position', 0:'rarefied_clonality'})\n",
    "    crdb['rarefied_clonality'] = crdb['rarefied_clonality'].astype(float)\n",
    "    crdb['rarefied_microdiversity'] = 1 - crdb['rarefied_clonality']\n",
    "    try:\n",
    "        crdb['normalized_rarefied_microdiversity'] = sklearn.preprocessing.normalize(crdb['rarefied_microdiversity'].values.reshape(-1, 1), axis=0) # Normalize clonality before NaN values are introduced\n",
    "    except ValueError:\n",
    "        crdb['normalized_rarefied_microdiversity'] = np.nan\n",
    "    \n",
    "    # Merge\n",
    "    db = pd.merge(bldb, db, on='position', how='outer')\n",
    "    db = pd.merge(db, cdb, on='position', how='outer')\n",
    "    db = pd.merge(db, crdb, on='position', how='outer')\n",
    "    db['Run'] = row['Run']\n",
    "    db = db.sort_values('position')\n",
    "    \n",
    "    # Backfill coverage with 0s and normalize\n",
    "    db['coverage'] = db['coverage'].fillna(0).astype(int)\n",
    "    db['normalized_coverage'] = sklearn.preprocessing.normalize(db['coverage'].values.reshape(-1, 1), axis=0)\n",
    "    \n",
    "    # Add mean and adjust\n",
    "#     db['mean_coverage'] = db['coverage'].mean()\n",
    "#     db['adjusted_coverage'] = db['coverage'] / db['mean_coverage']\n",
    "    \n",
    "#     db['mean_microdiversity'] = db['microdiversity'].mean()\n",
    "#     db['adjusted_microdiversity'] = db['coverage'] / db['mean_microdiversity']\n",
    "    \n",
    "#     db['mean_rarefied_microdiversity'] = db['rarefied_microdiversity'].mean()\n",
    "#     db['adjusted_rarefied_microdiversity'] = db['coverage'] / db['mean_rarefied_microdiversity']\n",
    "    \n",
    "    dbs.append(db)\n",
    "    \n",
    "# Re-order columns\n",
    "Odb = pd.concat(dbs).reset_index(drop=True)\n",
    "order = ['Run', 'position']\n",
    "for item in ['coverage', 'microdiversity', 'rarefied_microdiversity']:\n",
    "    order += [item, 'normalized_' + item]\n",
    "Odb = Odb[order]\n",
    "\n",
    "# Save\n",
    "Odb.to_csv(saveloc + 'COVID_positional_coverage_v2.csv.gz', index=False)\n",
    "Odb.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene-level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>gene</th>\n",
       "      <th>coverage</th>\n",
       "      <th>normalized_microdiversity</th>\n",
       "      <th>normalized_microdiversity_rank</th>\n",
       "      <th>coverage_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11059940</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>13.192140</td>\n",
       "      <td>0.001564</td>\n",
       "      <td>14.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11059940</td>\n",
       "      <td>5'UTR</td>\n",
       "      <td>1.607547</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11059940</td>\n",
       "      <td>GU280_gp01_stem_loop</td>\n",
       "      <td>16.060241</td>\n",
       "      <td>0.004236</td>\n",
       "      <td>30.0</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11059940</td>\n",
       "      <td>GU280_gp11_stem_loop</td>\n",
       "      <td>19.430769</td>\n",
       "      <td>0.004316</td>\n",
       "      <td>31.0</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR11059940</td>\n",
       "      <td>YP_009724389.1</td>\n",
       "      <td>12.438965</td>\n",
       "      <td>0.001565</td>\n",
       "      <td>15.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Run                  gene   coverage  normalized_microdiversity  \\\n",
       "0  SRR11059940                 3'UTR  13.192140                   0.001564   \n",
       "1  SRR11059940                 5'UTR   1.607547                        NaN   \n",
       "2  SRR11059940  GU280_gp01_stem_loop  16.060241                   0.004236   \n",
       "3  SRR11059940  GU280_gp11_stem_loop  19.430769                   0.004316   \n",
       "4  SRR11059940        YP_009724389.1  12.438965                   0.001565   \n",
       "\n",
       "   normalized_microdiversity_rank  coverage_rank  \n",
       "0                            14.0           15.0  \n",
       "1                             NaN            1.0  \n",
       "2                            30.0           22.0  \n",
       "3                            31.0           27.0  \n",
       "4                            15.0           13.0  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ON = 'normalized_microdiversity'\n",
    "MIN_COV = 0\n",
    "\n",
    "POdb = pd.read_csv('/home/mattolm/user_data/Covid_19/datatables/COVID_genes_positional_v4.csv')\n",
    "\n",
    "fdb = pd.merge(Odb, POdb, on='position', how='left').merge(COdb, how='left', on='Run')\n",
    "fdb = fdb[fdb['coverage_y'] >= MIN_COV]\n",
    "\n",
    "table = defaultdict(list)\n",
    "for run, rdb in fdb.groupby('Run'):\n",
    "    for gene, db in rdb.groupby('ID'):\n",
    "        table['Run'].append(run)\n",
    "        table['gene'].append(gene)\n",
    "        table['coverage'].append(db['coverage_x'].mean())\n",
    "        table[ON].append(db[ON].mean())\n",
    "\n",
    "run = 'all'\n",
    "rdb = fdb\n",
    "for gene, db in rdb.groupby('ID'):\n",
    "    table['Run'].append(run)\n",
    "    table['gene'].append(gene)\n",
    "    table['coverage'].append(db['coverage_x'].mean())\n",
    "    table[ON].append(db[ON].mean())\n",
    "\n",
    "MGdb = pd.DataFrame(table)\n",
    "\n",
    "# Add rank order\n",
    "dbs = defaultdict(list)\n",
    "ITEMS = [ON, 'coverage']\n",
    "for run, db in MGdb.groupby('Run'):\n",
    "    for item in ITEMS:\n",
    "        db['{0}_rank'.format(item)] = db[item].rank(method='min')#.astype(int)\n",
    "        dbs[item].append(db)\n",
    "for item in ITEMS:\n",
    "    db = pd.concat(dbs[item]).reset_index(drop=True)\n",
    "    MGdb = pd.merge(MGdb, db, how='left', on=list(MGdb.columns))\n",
    "    \n",
    "# Save\n",
    "MGdb.to_csv(saveloc + 'COVID_gene_coverage_v4.csv', index=False)\n",
    "MGdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
