{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import scipy\n",
    "import matplotlib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from collections import defaultdict\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "\n",
    "from subprocess import call\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set_style('whitegrid')\n",
    "pd.set_option('display.max_rows', 100)\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "pd.set_option('display.max_columns', 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def SRA_to_bioproject(accession):\n",
    "    '''\n",
    "    From an SRA (like SRR1883283) return a biosample\n",
    "    '''\n",
    "    from Bio import Entrez\n",
    "    Entrez.email = \"mattolm@gmail.com\"\n",
    "\n",
    "    # GO FROM SRA TO ID\n",
    "    #accession = 'SRX1883283'\n",
    "    handle = Entrez.esearch(db='sra', term=accession, retmode='text')\n",
    "    record = Entrez.read(handle)\n",
    "    handle.close()\n",
    "    ID = record['IdList'][0]\n",
    "    #return ID\n",
    "    \n",
    "    # ELINK FROM SRA ID TO BIOSAMPLE ID\n",
    "    accession = ID\n",
    "    handle = Entrez.elink(dbfrom=\"sra\", retmax=10, id=accession, linkname=\"sra_bioproject\")\n",
    "    record = Entrez.read(handle)\n",
    "    handle.close()\n",
    "    BID = [link[\"Id\"] for link in record[0][\"LinkSetDb\"][0][\"Link\"]][0]\n",
    "    #return BID\n",
    "\n",
    "    # SUMMARY FROM BIOSAMPLE ID TO TERM\n",
    "    handle = Entrez.esummary(db=\"bioproject\", id=BID)\n",
    "    record = Entrez.read(handle)\n",
    "    handle.close()\n",
    "    BioSample = record['DocumentSummarySet']['DocumentSummary'][0]['Project_Acc']\n",
    "    \n",
    "    return BioSample"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Link</th>\n",
       "      <th>Group</th>\n",
       "      <th>Method</th>\n",
       "      <th>Publication link</th>\n",
       "      <th>Publication title</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA615032</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>Icahn School of Medicine at Mount Sina</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>https://www.biorxiv.org/content/10.1101/2020.0...</td>\n",
       "      <td>SARS-CoV-2 launches a unique transcriptional s...</td>\n",
       "      <td>Looking at the response of human cells to COVID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA610428</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>UNIVERSITY OF WASHINGTON</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA613958</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>The Peter Doherty Institute for Infection and ...</td>\n",
       "      <td>PCR_ARTIC v1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA614546</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>Paragon Genomics</td>\n",
       "      <td>Mixed</td>\n",
       "      <td>https://www.biorxiv.org/content/10.1101/2020.0...</td>\n",
       "      <td>High sensitivity detection of coronavirus SARS...</td>\n",
       "      <td>Evaluating lots of things; worth following up....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA616446</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>Hubei Provincial Center for Disease Control an...</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>https://www.biorxiv.org/content/10.1101/2020.0...</td>\n",
       "      <td>Genome-wide data inferring the evolution and p...</td>\n",
       "      <td>Sequencing genomes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    BioProject                                               Link  \\\n",
       "0  PRJNA615032  https://www.ncbi.nlm.nih.gov/bioproject/?term=...   \n",
       "1  PRJNA610428  https://www.ncbi.nlm.nih.gov/bioproject/?term=...   \n",
       "2  PRJNA613958  https://www.ncbi.nlm.nih.gov/bioproject/?term=...   \n",
       "3  PRJNA614546  https://www.ncbi.nlm.nih.gov/bioproject/?term=...   \n",
       "4  PRJNA616446  https://www.ncbi.nlm.nih.gov/bioproject/?term=...   \n",
       "\n",
       "                                               Group        Method  \\\n",
       "0             Icahn School of Medicine at Mount Sina       RNA-seq   \n",
       "1                           UNIVERSITY OF WASHINGTON       RNA-seq   \n",
       "2  The Peter Doherty Institute for Infection and ...  PCR_ARTIC v1   \n",
       "3                                   Paragon Genomics         Mixed   \n",
       "4  Hubei Provincial Center for Disease Control an...       RNA-seq   \n",
       "\n",
       "                                    Publication link  \\\n",
       "0  https://www.biorxiv.org/content/10.1101/2020.0...   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3  https://www.biorxiv.org/content/10.1101/2020.0...   \n",
       "4  https://www.biorxiv.org/content/10.1101/2020.0...   \n",
       "\n",
       "                                   Publication title  \\\n",
       "0  SARS-CoV-2 launches a unique transcriptional s...   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3  High sensitivity detection of coronavirus SARS...   \n",
       "4  Genome-wide data inferring the evolution and p...   \n",
       "\n",
       "                                         Description  \n",
       "0    Looking at the response of human cells to COVID  \n",
       "1                                                NaN  \n",
       "2                                                NaN  \n",
       "3  Evaluating lots of things; worth following up....  \n",
       "4                                 Sequencing genomes  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val_loc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/validated_downloads.txt'\n",
    "meta_loc = '/home/mattolm/user_data/Covid_19/Pipeline/Metadata/metadata.xlsx'\n",
    "\n",
    "THREADS = 6\n",
    "\n",
    "VALIDATED_RUNS = []\n",
    "with open(val_loc, 'r') as o:\n",
    "    for line in o.readlines():\n",
    "        VALIDATED_RUNS.append(line.strip())\n",
    "VALIDATED_RUNS = set(VALIDATED_RUNS)\n",
    "\n",
    "BioMeta = pd.read_excel(meta_loc, sheet_name='BioProjects')\n",
    "BioMeta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get a list of all SRA samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/mattolm/miniconda3/lib/python3.7/site-packages/tqdm/std.py:658: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version\n",
      "  from pandas import Panel\n",
      "/home/mattolm/miniconda3/lib/python3.7/site-packages/tqdm/std.py:658: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version\n",
      "  from pandas import Panel\n",
      "All BioProjects are known\n"
     ]
    }
   ],
   "source": [
    "from datetime import date\n",
    "today = date.today()\n",
    "d4 = today.strftime(\"%d_%m_%Y\")\n",
    "dloc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA.tsv'.format(d4)\n",
    "\n",
    "SEARCH_STRING = '(\\\\\"Severe acute respiratory syndrome coronavirus 2\\\\\"[Organism] OR SARS-CoV-2[All Fields]) AND \\\\\"platform illumina\\\\\"[Properties]'\n",
    "\n",
    "! pysradb search \"$SEARCH_STRING\" --detailed --saveto \"$dloc\"\n",
    "\n",
    "SRdb = pd.read_csv(dloc, sep='\\t')\n",
    "SRdb['Run'] = SRdb['run_accession']\n",
    "SRdb = SRdb[~SRdb['run_accession'].isna()]\n",
    "\n",
    "# Add BioProject\n",
    "SRdb['BioProject'] = [SRA_to_bioproject(r) for r in SRdb['Run']]\n",
    "\n",
    "# Any unknown BioProjects?\n",
    "unk = set(SRdb['BioProject'].tolist()) - set(BioMeta['BioProject'].tolist())\n",
    "if len(unk) == 0:\n",
    "    print(\"All BioProjects are known\")\n",
    "else:\n",
    "    print(\"{0} new BioProjects with no info\".format(len(unk)))\n",
    "    print(\"\\n\".join(list(unk)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "90 new samples (and 77 already processed = 159 total (including irrelevant already processed))\n"
     ]
    }
   ],
   "source": [
    "# Filter out irrelevant bioprojects\n",
    "SRdb = SRdb[~SRdb['BioProject'].isin(BioMeta[BioMeta['Method'] == 'IRRELEVANT']['BioProject'])]\n",
    "NEW = set(SRdb['Run'].tolist())\n",
    "print(\"{0} new samples (and {1} already processed = {2} total (including irrelevant already processed))\".format(len(NEW - VALIDATED_RUNS), len(VALIDATED_RUNS), len(SRdb)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# START PROCESSING!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download samples that need downloading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "def check_sra_file(file):\n",
    "    cmd = \"/usr/bin/vdb-validate {0}\".format(file)\n",
    "    try:\n",
    "        output = subprocess.check_output(\n",
    "            cmd, stderr=subprocess.STDOUT, shell=True,\n",
    "            universal_newlines=True)\n",
    "        final_out = output.split('\\n')[-2]\n",
    "        return True\n",
    "    except Exception as exc:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running loop 0; downloading 23 new files\n"
     ]
    }
   ],
   "source": [
    "base_loc = '/home/mattolm/user_data/Covid_19/reads/'\n",
    "odir = '/home/mattolm/user_data/Covid_19/Pipeline/test/'\n",
    "\n",
    "SRdb['sra_file'] = [base_loc + x for x in SRdb['Run']]\n",
    "SRdb['sra_file_consistant'] = [(R in VALIDATED_RUNS) for R in SRdb['Run']]\n",
    "SRdb['sra_file_consistant'] = [check_sra_file(x) if y == False else y for x, y in zip(SRdb['sra_file'], SRdb['sra_file_consistant'])]\n",
    "\n",
    "TOTAL_LOOPS = 50\n",
    "loop = 0\n",
    "while len(SRdb[SRdb['sra_file_consistant'] == False]) > 0:\n",
    "    print('Running loop {0}; downloading {1} new files'.format(loop, len(SRdb[SRdb['sra_file_consistant'] == False])))\n",
    "    cmd_loc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/cmds/{0}_download.txt'.format(d4)\n",
    "    with open(cmd_loc, 'w') as o:\n",
    "        for i, row in SRdb[SRdb['sra_file_consistant'] == False].iterrows():\n",
    "            o.write(\"wget -c {0} -O {1}\\n\".format(row['sra_url'], row['sra_file']))\n",
    "    cmd = \"cat {0} | parallel -j {1}\".format(cmd_loc, THREADS)\n",
    "    call(cmd, shell=True)\n",
    "            \n",
    "    SRdb['sra_file_consistant'] = [check_sra_file(x) if y == False else y for x, y in zip(SRdb['sra_file'], SRdb['sra_file_consistant'])]\n",
    "    loop += 1\n",
    "    if loop == TOTAL_LOOPS:\n",
    "        break\n",
    "        \n",
    "if len(SRdb[SRdb['sra_file_consistant'] == False]) > 0:\n",
    "    print(\"THE FOLLOWING HAVE PROBLEMS:\")\n",
    "    for i, row in SRdb[SRdb['sra_file_consistant'] == False].iterrows():\n",
    "        print(\"/usr/bin/vdb-validate {0}\".format(row['sra_file']))\n",
    "        print(\"wget -c {0} -O {1}\\n\".format(row['sra_url'], row['sra_file']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All are good!\n"
     ]
    }
   ],
   "source": [
    "if len(SRdb[SRdb['sra_file_consistant'] == False]) > 0:\n",
    "    print(\"THE FOLLOWING HAVE PROBLEMS:\")\n",
    "    for i, row in SRdb[SRdb['sra_file_consistant'] == False].iterrows():\n",
    "        print(\"/usr/bin/vdb-validate {0}\".format(row['sra_file']))\n",
    "        print(\"wget -c {0} -O {1}\\n\".format(row['sra_url'], row['sra_file']))\n",
    "else:\n",
    "    print(\"All are good!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save new validations\n",
    "VALIDATED_RUNS = VALIDATED_RUNS.union(set(SRdb[SRdb['sra_file_consistant'] == True]['Run'].tolist()))\n",
    "with open('/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/validated_downloads.txt', 'w') as o:\n",
    "    for v in VALIDATED_RUNS:\n",
    "        o.write(v + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FastqQ Dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loop 0 - 90 samples are missing their fastq dumps\n"
     ]
    }
   ],
   "source": [
    "base_loc = '/home/mattolm/user_data/Covid_19/reads/'\n",
    "\n",
    "SRdb['fastq1'] = [x + '_1.fastq' for x in SRdb['sra_file']]\n",
    "SRdb['fastq2'] = [x + '_2.fastq' for x in SRdb['sra_file']]\n",
    "\n",
    "SRdb['fastq1_exists'] = [os.path.exists(x) for x in SRdb['fastq1']]\n",
    "SRdb['fastq2_exists'] = [os.path.exists(x) for x in SRdb['fastq2']]\n",
    "\n",
    "loop = 0\n",
    "MISSING_FASTA = SRdb[SRdb['fastq1_exists'] == False]\n",
    "while len(MISSING_FASTA) > 0:\n",
    "    print(\"Loop {0} - {1} samples are missing their fastq dumps\".format(loop, len(MISSING_FASTA)))\n",
    "    \n",
    "    cmd_loc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/cmds/{0}_dump.txt'.format(d4)\n",
    "    with open(cmd_loc, 'w') as o:\n",
    "        for i, row in MISSING_FASTA.iterrows():\n",
    "            o.write(\"fastq-dump {0} --split-files -O {1}\".format(row['sra_file'], base_loc) + '\\n')\n",
    "    cmd = \"cat {0} | parallel -j {1}\".format(cmd_loc, THREADS)\n",
    "    call(cmd, shell=True)\n",
    "\n",
    "    SRdb['fastq1_exists'] = [os.path.exists(x) for x in SRdb['fastq1']]\n",
    "    MISSING_FASTA = SRdb[SRdb['fastq1_exists'] == False]\n",
    "    loop += 1\n",
    "    \n",
    "SRdb['LibraryLayout'] = ['PAIRED' if t else 'SINGLE' for t in SRdb['fastq2_exists']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PAIRED\n",
      "True    55\n",
      "Name: fastq2_exists, dtype: int64\n",
      "SINGLE\n",
      "False    104\n",
      "Name: fastq2_exists, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "for x, db in SRdb.groupby('LibraryLayout'):\n",
    "    print(x)\n",
    "    print(db['fastq2_exists'].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "SRdb['filtered_fastq1'] = ['/home/mattolm/user_data/Covid_19/reads/filtered/' + os.path.basename(rp).replace('_1.fastq', '_rep1.fastq').replace('_rep1.fastq', '_bbduk_1.fastq') for rp in SRdb['fastq1']]\n",
    "SRdb['filtered_fastq2'] = ['/home/mattolm/user_data/Covid_19/reads/filtered/' + os.path.basename(rp).replace('_2.fastq', '_rep2.fastq').replace('_rep2.fastq', '_bbduk_2.fastq') for rp in SRdb['fastq2']]\n",
    "SRdb['filtered_fastqS'] = ['/home/mattolm/user_data/Covid_19/reads/filtered/' + os.path.basename(rp).replace('_1.fastq', '_rep1.fastq').replace('_rep1.fastq', '_bbduk.fastq') for rp in SRdb['fastq1']]\n",
    "\n",
    "SRdb['filtered_fastq1_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastq1']]\n",
    "SRdb['filtered_fastq2_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastq2']]\n",
    "SRdb['filtered_fastqS_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastqS']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11454606_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11454606_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11454607_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11454607_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11454608_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11454608_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11454609_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11454609_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11454610_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11454610_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11454611_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11454611_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11454612_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11454612_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11454613_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11454613_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11454614_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11454614_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11454615_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11454615_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412291_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412291_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412292_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412292_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412293_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412293_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412294_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412294_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412290_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412290_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412289_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412289_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412288_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412288_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412287_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412287_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412286_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412286_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412285_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412285_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412284_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412284_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412283_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412283_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412279_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412279_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412280_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412280_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412281_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412281_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412282_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412282_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412275_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412275_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412276_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412276_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412277_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412277_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412278_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412278_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412274_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412274_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412273_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412273_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412272_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412272_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412271_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412271_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412270_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412270_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412269_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412269_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412268_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412268_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412267_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412267_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412266_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412266_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412265_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412265_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412264_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412264_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412263_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412263_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412262_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412262_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412261_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412261_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412260_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412260_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412259_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412259_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412258_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412258_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412257_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412257_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412256_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412256_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412255_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412255_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412254_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412254_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412251_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412251_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412253_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412253_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412252_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412252_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412250_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412250_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412249_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412249_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412248_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412248_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412247_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412247_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412243_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412243_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412244_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412244_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412245_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412245_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412246_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412246_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412242_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412242_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412241_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412241_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412240_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412240_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412239_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412239_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412238_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412238_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412237_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412237_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412236_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412236_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412235_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412235_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412231_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412231_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412232_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412232_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412233_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412233_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412234_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412234_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412227_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412227_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412228_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412228_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412229_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412229_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412230_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412230_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412226_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412226_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412223_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412223_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412225_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412225_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412224_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412224_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412222_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412222_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412221_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412221_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412220_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412220_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412219_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412219_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412218_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412218_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412217_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412217_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412216_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412216_bbduk.fastq threads=6\n",
      "bbduk.sh in=/home/mattolm/user_data/Covid_19/reads/SRR11412215_1.fastq out=/home/mattolm/user_data/Covid_19/reads/filtered/SRR11412215_bbduk.fastq threads=6\n"
     ]
    }
   ],
   "source": [
    "from subprocess import call\n",
    "def process_reads(r1, r2, outfolder):\n",
    "    # repair\n",
    "    rr1 = outfolder + os.path.basename(r1).replace('_1.fastq', '_rep1.fastq')\n",
    "    rr2 = outfolder + os.path.basename(r2).replace('_2.fastq', '_rep2.fastq')\n",
    "    cmd = \"repair.sh in={0} in2={1} out={2} out2={3}\".format(r1, r2, rr1, rr2)\n",
    "    print(cmd)\n",
    "    call(cmd, shell=True)\n",
    "    \n",
    "    # bbduk\n",
    "    rb1 = rr1.replace('_rep1.fastq', '_bbduk_1.fastq')\n",
    "    rb2 = rr2.replace('_rep2.fastq', '_bbduk_2.fastq')\n",
    "    cmd = \"bbduk.sh in={0} in2={1} out={2} out2={3} threads={4}\".format(rr1, rr2, rb1, rb2, THREADS)\n",
    "    print(cmd)\n",
    "    call(cmd, shell=True)\n",
    "    \n",
    "    return rb1, rb2\n",
    "\n",
    "def process_reads_s(r1, outfolder):\n",
    "    # bbduk\n",
    "    rr1 = outfolder + os.path.basename(r1).replace('_1.fastq', '_rep1.fastq')\n",
    "    rb1 = rr1.replace('_rep1.fastq', '_bbduk.fastq')\n",
    "    cmd = \"bbduk.sh in={0} out={1} threads={2}\".format(r1, rb1, THREADS)\n",
    "    print(cmd)\n",
    "    call(cmd, shell=True)\n",
    "    \n",
    "    return rb1 \n",
    "\n",
    "SRdb['filtered_fastq1_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastq1']]\n",
    "SRdb['filtered_fastq2_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastq2']]\n",
    "SRdb['filtered_fastqS_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastqS']]\n",
    "\n",
    "# Process paired\n",
    "for i, row in SRdb[(SRdb['LibraryLayout'] == 'PAIRED') & ((SRdb['filtered_fastq1_exists'] == False) | (SRdb['filtered_fastq2_exists'] == False))].iterrows():\n",
    "    rb1, rb2 = process_reads(row['fastq1'], row['fastq2'], '/home/mattolm/user_data/Covid_19/reads/filtered/')\n",
    "    \n",
    "# Process unpaired\n",
    "for i, row in SRdb[(SRdb['LibraryLayout'] == 'SINGLE') & (SRdb['filtered_fastqS_exists'] == False)].iterrows():\n",
    "    b1 = process_reads_s(row['fastq1'], '/home/mattolm/user_data/Covid_19/reads/filtered/')\n",
    "    \n",
    "SRdb['filtered_fastq1_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastq1']]\n",
    "SRdb['filtered_fastq2_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastq2']]\n",
    "SRdb['filtered_fastqS_exists'] = [os.path.exists(x) for x in SRdb['filtered_fastqS']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True    159\n",
       "Name: mapping_done, dtype: int64"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BTL = '/home/mattolm/user_data/Covid_19/genomes/NC_045512.2.fasta.bt2'\n",
    "ML = '/home/mattolm/user_data/Covid_19/inStrain/mapping/'\n",
    "s2l = SRdb.set_index('Run')['LibraryLayout'].to_dict()\n",
    "\n",
    "SRdb['sam_file'] = [\"{0}{1}-vs-{2}.sam\".format(ML, os.path.basename(BTL), r) for r in SRdb['Run']]\n",
    "SRdb['mapping_done'] = [os.path.isfile(s) for s in SRdb['sam_file']]\n",
    "\n",
    "cmd_loc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/cmds/{0}_mapping.txt'.format(d4)\n",
    "re = open(cmd_loc, 'w')\n",
    "\n",
    "for i, row in SRdb[SRdb['mapping_done'] == False].iterrows():\n",
    "#for i, row in SRdb.iterrows():\n",
    "    sam_loc = row['sam_file']\n",
    "    if s2l[row['Run']] == 'PAIRED':\n",
    "        cmd = \"bowtie2 -x {3} -1 {0} -2 {1} --no-unal -S {2} -p {4} 2> {2}.log\".format(\n",
    "                row['filtered_fastq1'], row['filtered_fastq2'], sam_loc, BTL, THREADS)\n",
    "        re.write(cmd + '\\n')\n",
    "    else:\n",
    "        cmd = \"bowtie2 -x {3} -U {0} --no-unal -S {2} -p {4} 2> {2}.log\".format(\n",
    "                row['filtered_fastqS'], '', sam_loc, BTL, THREADS)\n",
    "        re.write(cmd + '\\n')\n",
    "re.close()\n",
    "\n",
    "cmd = \"cat {0} | bash\".format(cmd_loc)\n",
    "call(cmd, shell=True)\n",
    "\n",
    "SRdb['mapping_done'] = [os.path.isfile(s) for s in SRdb['sam_file']]\n",
    "SRdb['mapping_done'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse bowtie2 logs for information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_bt2(log):\n",
    "    table = defaultdict(list)\n",
    "    with open(log, 'r') as o:\n",
    "        for line in o.readlines():\n",
    "            line = line.strip()\n",
    "            if 'reads;' in line:\n",
    "                table['total_reads'].append(int(line.split()[0]))\n",
    "            if 'paired;' in line:\n",
    "                table['paired_reads'].append(int(line.split()[0]))\n",
    "            if 'overall alignment rate' in line:\n",
    "                table['percent_aligned'].append(float(line.split()[0][:-1]))\n",
    "    db = pd.DataFrame(table)\n",
    "    db['aligned_reads'] = db['total_reads'] * (db['percent_aligned'] / 100)\n",
    "    db['aligned_reads'] = db['aligned_reads'].astype(int)\n",
    "    return db\n",
    "\n",
    "dbs = []\n",
    "SRdb['mapping_log'] = SRdb['sam_file'] + '.log'\n",
    "SRdb['mapping_log_exists'] = [os.path.exists(x) for x in SRdb['mapping_log']]\n",
    "for i, row in SRdb.iterrows():\n",
    "    db = parse_bt2(row['mapping_log'])\n",
    "    db['Run'] = row['Run']\n",
    "    dbs.append(db)\n",
    "Mdb = pd.concat(dbs).reset_index(drop=True)\n",
    "for col in Mdb.columns:\n",
    "    if col in ['Run']:\n",
    "        continue\n",
    "    r2c = Mdb.set_index('Run')[col].to_dict()\n",
    "    SRdb[col] = SRdb['Run'].map(r2c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run inStrain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     123\n",
       "False     36\n",
       "Name: inStrain_succeeded, dtype: int64"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SRdb['inStrain_coverage'] = [\"/home/mattolm/user_data/Covid_19/inStrain/profiles/{0}.IS/output/{0}.IS_scaffold_info.tsv\".format(os.path.basename(r)) for r in SRdb['sam_file']]\n",
    "SRdb['inStrain_succeeded'] = [os.path.isfile(s) for s in SRdb['inStrain_coverage']]\n",
    "SRdb['inStrain_succeeded'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cat /home/mattolm/user_data/Covid_19/Pipeline/Jupyter/cmds/02_04_2020_instrain.txt | parallel -j 6\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True     123\n",
       "False     36\n",
       "Name: inStrain_succeeded, dtype: int64"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmd_loc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/cmds/{0}_instrain.txt'.format(d4)\n",
    "fasta = '/home/mattolm/user_data/Covid_19/genomes/NC_045512.2.fasta'\n",
    "genes ='/home/mattolm/user_data/Covid_19/genomes/NC_045512.2.gb'\n",
    "s2l = SRdb.set_index('Run')['LibraryLayout'].to_dict()\n",
    "\n",
    "re = open(cmd_loc, 'w')\n",
    "for i, row in SRdb[SRdb['inStrain_succeeded'] == False].iterrows():\n",
    "    if s2l[row['Run']] == 'PAIRED': \n",
    "        cmd = \"inStrain profile {0} {2} -o /home/mattolm/user_data/Covid_19/inStrain/profiles/{1}.IS -p 1 -g {3} --skip_mm_profiling\".format(row['sam_file'], os.path.basename(row['sam_file']), fasta, genes)\n",
    "        re.write(cmd + '\\n')\n",
    "    else:\n",
    "        cmd = \"inStrain profile {0} {2} -o /home/mattolm/user_data/Covid_19/inStrain/profiles/{1}.IS -p 1 --pairing_filter non_discordant -g {3} --skip_mm_profiling\".format(row['sam_file'], os.path.basename(row['sam_file']), fasta, genes)\n",
    "        re.write(cmd + '\\n')\n",
    "re.close()\n",
    "    \n",
    "cmd = \"cat {0} | parallel -j {1}\".format(cmd_loc, THREADS)\n",
    "print(cmd)\n",
    "call(cmd, shell=True)\n",
    "    \n",
    "SRdb['inStrain_succeeded'] = [os.path.isfile(s) for s in SRdb['inStrain_coverage']]\n",
    "SRdb['inStrain_succeeded'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 samples failed inStrain that should have passed\n"
     ]
    }
   ],
   "source": [
    "print(\"{0} samples failed inStrain that should have passed\".format(len(SRdb[(SRdb['inStrain_succeeded'] == False) & (SRdb['aligned_reads'] > 0)])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "['/'.join(x.split('/')[:-2]) + '/log/log.log' for x in SRdb[(SRdb['inStrain_succeeded'] == False) & (SRdb['aligned_reads'] > 0)]['inStrain_coverage'].tolist()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>experiment_desc</th>\n",
       "      <th>organism_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>instrument</th>\n",
       "      <th>total_spots</th>\n",
       "      <th>total_size</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>run_total_spots</th>\n",
       "      <th>run_total_bases</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>sra_url_alt</th>\n",
       "      <th>sra_url</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>isolate</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>host</th>\n",
       "      <th>host_disease</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>lat_lon</th>\n",
       "      <th>BioSampleModel</th>\n",
       "      <th>source_name</th>\n",
       "      <th>cell line</th>\n",
       "      <th>cell type</th>\n",
       "      <th>treatment</th>\n",
       "      <th>strain</th>\n",
       "      <th>time point</th>\n",
       "      <th>identification_method</th>\n",
       "      <th>culture_collection</th>\n",
       "      <th>host_description</th>\n",
       "      <th>passage_history</th>\n",
       "      <th>sample type</th>\n",
       "      <th>Laboratory Host</th>\n",
       "      <th>Extraction Method</th>\n",
       "      <th>ref_biomaterial</th>\n",
       "      <th>link_addit_analys</th>\n",
       "      <th>host_age</th>\n",
       "      <th>host_sex</th>\n",
       "      <th>country</th>\n",
       "      <th>sra_url_alt1</th>\n",
       "      <th>sra_url_alt2</th>\n",
       "      <th>host_disease_outcome</th>\n",
       "      <th>host_disease_stage</th>\n",
       "      <th>ena_fastq_url</th>\n",
       "      <th>ena_fastq_ftp</th>\n",
       "      <th>Run</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>sra_file</th>\n",
       "      <th>sra_exists</th>\n",
       "      <th>sra_file_consistant</th>\n",
       "      <th>fastq1</th>\n",
       "      <th>fastq2</th>\n",
       "      <th>fastq1_exists</th>\n",
       "      <th>fastq2_exists</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>filtered_fastq1</th>\n",
       "      <th>filtered_fastq2</th>\n",
       "      <th>filtered_fastqS</th>\n",
       "      <th>filtered_fastq1_exists</th>\n",
       "      <th>filtered_fastq2_exists</th>\n",
       "      <th>filtered_fastqS_exists</th>\n",
       "      <th>sam_file</th>\n",
       "      <th>mapping_done</th>\n",
       "      <th>mapping_log</th>\n",
       "      <th>mapping_log_exists</th>\n",
       "      <th>total_reads</th>\n",
       "      <th>paired_reads</th>\n",
       "      <th>percent_aligned</th>\n",
       "      <th>aligned_reads</th>\n",
       "      <th>inStrain_coverage</th>\n",
       "      <th>inStrain_succeeded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRP254688</td>\n",
       "      <td>SRX8032211</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>2697049.0</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>SRS6404546</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>5668472.0</td>\n",
       "      <td>489620824.0</td>\n",
       "      <td>SRR11454606</td>\n",
       "      <td>5668472.0</td>\n",
       "      <td>1.350169e+09</td>\n",
       "      <td>BetaCoV_Tianmen_HBCDC-HB-07_2020_R1_001.fastq.gz</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>human BetaCoV Tianmen HBCDC-HB-07/2020</td>\n",
       "      <td>Tianmen Center for Disease Control and Prevention</td>\n",
       "      <td>2020-02-08</td>\n",
       "      <td>China: Hubei</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Tianmen</td>\n",
       "      <td>not collected</td>\n",
       "      <td>Pathogen.cl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR11454606</td>\n",
       "      <td>PRJNA616446</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>5668472</td>\n",
       "      <td>5668472</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1700</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/prof...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRP254688</td>\n",
       "      <td>SRX8032210</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>2697049.0</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>SRS6404545</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>4407436.0</td>\n",
       "      <td>413461584.0</td>\n",
       "      <td>SRR11454607</td>\n",
       "      <td>4407436.0</td>\n",
       "      <td>1.157795e+09</td>\n",
       "      <td>BetaCoV_Wuhan_HBCDC-HB-06_2020_R1_001.fastq.gz</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>human BetaCoV Wuhan HBCDC-HB-06/2020</td>\n",
       "      <td>Wuhan Lung Hospital</td>\n",
       "      <td>2020-02-07</td>\n",
       "      <td>China: Hubei</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>not collected</td>\n",
       "      <td>Pathogen.cl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR11454607</td>\n",
       "      <td>PRJNA616446</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>4407436</td>\n",
       "      <td>4407436</td>\n",
       "      <td>0.25</td>\n",
       "      <td>11018</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/prof...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRP254688</td>\n",
       "      <td>SRX8032209</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>2697049.0</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>SRS6404544</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina iSeq 100</td>\n",
       "      <td>2690410.0</td>\n",
       "      <td>232046490.0</td>\n",
       "      <td>SRR11454608</td>\n",
       "      <td>2690410.0</td>\n",
       "      <td>7.258831e+08</td>\n",
       "      <td>BetaCoV_Jingzhou_HBCDC-HB-01_2020_R1_001.fastq.gz</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>human BetaCoV Jingzhou HBCDC-HB-01/2020</td>\n",
       "      <td>Jingzhou Center for Disease Control and Preven...</td>\n",
       "      <td>2020-01-08</td>\n",
       "      <td>China: Hubei</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Jingzhou</td>\n",
       "      <td>not collected</td>\n",
       "      <td>Pathogen.cl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR11454608</td>\n",
       "      <td>PRJNA616446</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>2690410</td>\n",
       "      <td>2690410</td>\n",
       "      <td>3.20</td>\n",
       "      <td>86093</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/prof...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRP254688</td>\n",
       "      <td>SRX8032208</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>2697049.0</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>SRS6404543</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>17121629.0</td>\n",
       "      <td>517518269.0</td>\n",
       "      <td>SRR11454609</td>\n",
       "      <td>17121629.0</td>\n",
       "      <td>1.284122e+09</td>\n",
       "      <td>BetaCoV_Wuhan_HBCDC-HB-02_2020_S10_R1_001.fast...</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>human BetaCoV Wuhan HBCDC-HB-02/2020</td>\n",
       "      <td>The Central Hospital Of Wuhan</td>\n",
       "      <td>2020-01-17</td>\n",
       "      <td>China: Hubei</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>not collected</td>\n",
       "      <td>Pathogen.cl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR11454609</td>\n",
       "      <td>PRJNA616446</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>17121629</td>\n",
       "      <td>17121629</td>\n",
       "      <td>0.37</td>\n",
       "      <td>63350</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/prof...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRP254688</td>\n",
       "      <td>SRX8032207</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>2697049.0</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>SRS6404542</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>14337950.0</td>\n",
       "      <td>413719955.0</td>\n",
       "      <td>SRR11454610</td>\n",
       "      <td>14337950.0</td>\n",
       "      <td>1.075346e+09</td>\n",
       "      <td>BetaCoV_Wuhan_HBCDC-HB-03_2020_S24_R1_001.fast...</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>https://sra-download.ncbi.nlm.nih.gov/traces/s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>human BetaCoV Wuhan HBCDC-HB-03/2020</td>\n",
       "      <td>Union Hospital of Tongji Medical College, Huaz...</td>\n",
       "      <td>2020-01-18</td>\n",
       "      <td>China: Hubei</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>not collected</td>\n",
       "      <td>Pathogen.cl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR11454610</td>\n",
       "      <td>PRJNA616446</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/SRR1145...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/reads/filtere...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/mapp...</td>\n",
       "      <td>True</td>\n",
       "      <td>14337950</td>\n",
       "      <td>14337950</td>\n",
       "      <td>0.87</td>\n",
       "      <td>124740</td>\n",
       "      <td>/home/mattolm/user_data/Covid_19/inStrain/prof...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession experiment_accession  \\\n",
       "0       SRP254688           SRX8032211   \n",
       "1       SRP254688           SRX8032210   \n",
       "2       SRP254688           SRX8032209   \n",
       "3       SRP254688           SRX8032208   \n",
       "4       SRP254688           SRX8032207   \n",
       "\n",
       "                                    experiment_title  \\\n",
       "0  RNA-Seq of Homo sapiens: hCov-19 infected pati...   \n",
       "1  RNA-Seq of Homo sapiens: hCov-19 infected pati...   \n",
       "2  RNA-Seq of Homo sapiens: hCov-19 infected pati...   \n",
       "3  RNA-Seq of Homo sapiens: hCov-19 infected pati...   \n",
       "4  RNA-Seq of Homo sapiens: hCov-19 infected pati...   \n",
       "\n",
       "                                     experiment_desc  organism_taxid   \\\n",
       "0  RNA-Seq of Homo sapiens: hCov-19 infected pati...        2697049.0   \n",
       "1  RNA-Seq of Homo sapiens: hCov-19 infected pati...        2697049.0   \n",
       "2  RNA-Seq of Homo sapiens: hCov-19 infected pati...        2697049.0   \n",
       "3  RNA-Seq of Homo sapiens: hCov-19 infected pati...        2697049.0   \n",
       "4  RNA-Seq of Homo sapiens: hCov-19 infected pati...        2697049.0   \n",
       "\n",
       "                                     organism_name library_strategy  \\\n",
       "0  Severe acute respiratory syndrome coronavirus 2          RNA-Seq   \n",
       "1  Severe acute respiratory syndrome coronavirus 2          RNA-Seq   \n",
       "2  Severe acute respiratory syndrome coronavirus 2          RNA-Seq   \n",
       "3  Severe acute respiratory syndrome coronavirus 2          RNA-Seq   \n",
       "4  Severe acute respiratory syndrome coronavirus 2          RNA-Seq   \n",
       "\n",
       "   library_source library_selection sample_accession  sample_title  \\\n",
       "0  TRANSCRIPTOMIC        RANDOM PCR       SRS6404546           NaN   \n",
       "1  TRANSCRIPTOMIC        RANDOM PCR       SRS6404545           NaN   \n",
       "2  TRANSCRIPTOMIC        RANDOM PCR       SRS6404544           NaN   \n",
       "3  TRANSCRIPTOMIC        RANDOM PCR       SRS6404543           NaN   \n",
       "4  TRANSCRIPTOMIC        RANDOM PCR       SRS6404542           NaN   \n",
       "\n",
       "          instrument  total_spots   total_size run_accession  run_total_spots  \\\n",
       "0   Illumina MiniSeq    5668472.0  489620824.0   SRR11454606        5668472.0   \n",
       "1   Illumina MiniSeq    4407436.0  413461584.0   SRR11454607        4407436.0   \n",
       "2  Illumina iSeq 100    2690410.0  232046490.0   SRR11454608        2690410.0   \n",
       "3        NextSeq 550   17121629.0  517518269.0   SRR11454609       17121629.0   \n",
       "4        NextSeq 550   14337950.0  413719955.0   SRR11454610       14337950.0   \n",
       "\n",
       "   run_total_bases                                          run_alias  \\\n",
       "0     1.350169e+09   BetaCoV_Tianmen_HBCDC-HB-07_2020_R1_001.fastq.gz   \n",
       "1     1.157795e+09     BetaCoV_Wuhan_HBCDC-HB-06_2020_R1_001.fastq.gz   \n",
       "2     7.258831e+08  BetaCoV_Jingzhou_HBCDC-HB-01_2020_R1_001.fastq.gz   \n",
       "3     1.284122e+09  BetaCoV_Wuhan_HBCDC-HB-02_2020_S10_R1_001.fast...   \n",
       "4     1.075346e+09  BetaCoV_Wuhan_HBCDC-HB-03_2020_S24_R1_001.fast...   \n",
       "\n",
       "                                         sra_url_alt  \\\n",
       "0  https://sra-download.ncbi.nlm.nih.gov/traces/s...   \n",
       "1  https://sra-download.ncbi.nlm.nih.gov/traces/s...   \n",
       "2  https://sra-download.ncbi.nlm.nih.gov/traces/s...   \n",
       "3  https://sra-download.ncbi.nlm.nih.gov/traces/s...   \n",
       "4  https://sra-download.ncbi.nlm.nih.gov/traces/s...   \n",
       "\n",
       "                                             sra_url experiment_alias  \\\n",
       "0  https://sra-download.ncbi.nlm.nih.gov/traces/s...              NaN   \n",
       "1  https://sra-download.ncbi.nlm.nih.gov/traces/s...              NaN   \n",
       "2  https://sra-download.ncbi.nlm.nih.gov/traces/s...              NaN   \n",
       "3  https://sra-download.ncbi.nlm.nih.gov/traces/s...              NaN   \n",
       "4  https://sra-download.ncbi.nlm.nih.gov/traces/s...              NaN   \n",
       "\n",
       "                                   isolate  \\\n",
       "0   human BetaCoV Tianmen HBCDC-HB-07/2020   \n",
       "1     human BetaCoV Wuhan HBCDC-HB-06/2020   \n",
       "2  human BetaCoV Jingzhou HBCDC-HB-01/2020   \n",
       "3     human BetaCoV Wuhan HBCDC-HB-02/2020   \n",
       "4     human BetaCoV Wuhan HBCDC-HB-03/2020   \n",
       "\n",
       "                                        collected_by collection_date  \\\n",
       "0  Tianmen Center for Disease Control and Prevention      2020-02-08   \n",
       "1                                Wuhan Lung Hospital      2020-02-07   \n",
       "2  Jingzhou Center for Disease Control and Preven...      2020-01-08   \n",
       "3                      The Central Hospital Of Wuhan      2020-01-17   \n",
       "4  Union Hospital of Tongji Medical College, Huaz...      2020-01-18   \n",
       "\n",
       "   geo_loc_name          host host_disease isolation_source        lat_lon  \\\n",
       "0  China: Hubei  Homo sapiens     COVID-19          Tianmen  not collected   \n",
       "1  China: Hubei  Homo sapiens     COVID-19            Wuhan  not collected   \n",
       "2  China: Hubei  Homo sapiens     COVID-19         Jingzhou  not collected   \n",
       "3  China: Hubei  Homo sapiens     COVID-19            Wuhan  not collected   \n",
       "4  China: Hubei  Homo sapiens     COVID-19            Wuhan  not collected   \n",
       "\n",
       "  BioSampleModel source_name cell line cell type treatment strain time point  \\\n",
       "0    Pathogen.cl         NaN       NaN       NaN       NaN    NaN        NaN   \n",
       "1    Pathogen.cl         NaN       NaN       NaN       NaN    NaN        NaN   \n",
       "2    Pathogen.cl         NaN       NaN       NaN       NaN    NaN        NaN   \n",
       "3    Pathogen.cl         NaN       NaN       NaN       NaN    NaN        NaN   \n",
       "4    Pathogen.cl         NaN       NaN       NaN       NaN    NaN        NaN   \n",
       "\n",
       "  identification_method culture_collection host_description passage_history  \\\n",
       "0                   NaN                NaN              NaN             NaN   \n",
       "1                   NaN                NaN              NaN             NaN   \n",
       "2                   NaN                NaN              NaN             NaN   \n",
       "3                   NaN                NaN              NaN             NaN   \n",
       "4                   NaN                NaN              NaN             NaN   \n",
       "\n",
       "  sample type Laboratory Host Extraction Method ref_biomaterial  \\\n",
       "0         NaN             NaN               NaN             NaN   \n",
       "1         NaN             NaN               NaN             NaN   \n",
       "2         NaN             NaN               NaN             NaN   \n",
       "3         NaN             NaN               NaN             NaN   \n",
       "4         NaN             NaN               NaN             NaN   \n",
       "\n",
       "  link_addit_analys  host_age host_sex country sra_url_alt1 sra_url_alt2  \\\n",
       "0               NaN       NaN      NaN     NaN          NaN          NaN   \n",
       "1               NaN       NaN      NaN     NaN          NaN          NaN   \n",
       "2               NaN       NaN      NaN     NaN          NaN          NaN   \n",
       "3               NaN       NaN      NaN     NaN          NaN          NaN   \n",
       "4               NaN       NaN      NaN     NaN          NaN          NaN   \n",
       "\n",
       "  host_disease_outcome host_disease_stage ena_fastq_url  ena_fastq_ftp  \\\n",
       "0                  NaN                NaN           NaN            NaN   \n",
       "1                  NaN                NaN           NaN            NaN   \n",
       "2                  NaN                NaN           NaN            NaN   \n",
       "3                  NaN                NaN           NaN            NaN   \n",
       "4                  NaN                NaN           NaN            NaN   \n",
       "\n",
       "           Run   BioProject  \\\n",
       "0  SRR11454606  PRJNA616446   \n",
       "1  SRR11454607  PRJNA616446   \n",
       "2  SRR11454608  PRJNA616446   \n",
       "3  SRR11454609  PRJNA616446   \n",
       "4  SRR11454610  PRJNA616446   \n",
       "\n",
       "                                            sra_file  sra_exists  \\\n",
       "0  /home/mattolm/user_data/Covid_19/reads/SRR1145...        True   \n",
       "1  /home/mattolm/user_data/Covid_19/reads/SRR1145...        True   \n",
       "2  /home/mattolm/user_data/Covid_19/reads/SRR1145...        True   \n",
       "3  /home/mattolm/user_data/Covid_19/reads/SRR1145...        True   \n",
       "4  /home/mattolm/user_data/Covid_19/reads/SRR1145...        True   \n",
       "\n",
       "   sra_file_consistant                                             fastq1  \\\n",
       "0                 True  /home/mattolm/user_data/Covid_19/reads/SRR1145...   \n",
       "1                 True  /home/mattolm/user_data/Covid_19/reads/SRR1145...   \n",
       "2                 True  /home/mattolm/user_data/Covid_19/reads/SRR1145...   \n",
       "3                 True  /home/mattolm/user_data/Covid_19/reads/SRR1145...   \n",
       "4                 True  /home/mattolm/user_data/Covid_19/reads/SRR1145...   \n",
       "\n",
       "                                              fastq2  fastq1_exists  \\\n",
       "0  /home/mattolm/user_data/Covid_19/reads/SRR1145...           True   \n",
       "1  /home/mattolm/user_data/Covid_19/reads/SRR1145...           True   \n",
       "2  /home/mattolm/user_data/Covid_19/reads/SRR1145...           True   \n",
       "3  /home/mattolm/user_data/Covid_19/reads/SRR1145...           True   \n",
       "4  /home/mattolm/user_data/Covid_19/reads/SRR1145...           True   \n",
       "\n",
       "   fastq2_exists LibraryLayout  \\\n",
       "0          False        SINGLE   \n",
       "1          False        SINGLE   \n",
       "2          False        SINGLE   \n",
       "3          False        SINGLE   \n",
       "4          False        SINGLE   \n",
       "\n",
       "                                     filtered_fastq1  \\\n",
       "0  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "1  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "2  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "3  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "4  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "\n",
       "                                     filtered_fastq2  \\\n",
       "0  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "1  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "2  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "3  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "4  /home/mattolm/user_data/Covid_19/reads/filtere...   \n",
       "\n",
       "                                     filtered_fastqS  filtered_fastq1_exists  \\\n",
       "0  /home/mattolm/user_data/Covid_19/reads/filtere...                   False   \n",
       "1  /home/mattolm/user_data/Covid_19/reads/filtere...                   False   \n",
       "2  /home/mattolm/user_data/Covid_19/reads/filtere...                   False   \n",
       "3  /home/mattolm/user_data/Covid_19/reads/filtere...                   False   \n",
       "4  /home/mattolm/user_data/Covid_19/reads/filtere...                   False   \n",
       "\n",
       "   filtered_fastq2_exists  filtered_fastqS_exists  \\\n",
       "0                   False                    True   \n",
       "1                   False                    True   \n",
       "2                   False                    True   \n",
       "3                   False                    True   \n",
       "4                   False                    True   \n",
       "\n",
       "                                            sam_file  mapping_done  \\\n",
       "0  /home/mattolm/user_data/Covid_19/inStrain/mapp...          True   \n",
       "1  /home/mattolm/user_data/Covid_19/inStrain/mapp...          True   \n",
       "2  /home/mattolm/user_data/Covid_19/inStrain/mapp...          True   \n",
       "3  /home/mattolm/user_data/Covid_19/inStrain/mapp...          True   \n",
       "4  /home/mattolm/user_data/Covid_19/inStrain/mapp...          True   \n",
       "\n",
       "                                         mapping_log  mapping_log_exists  \\\n",
       "0  /home/mattolm/user_data/Covid_19/inStrain/mapp...                True   \n",
       "1  /home/mattolm/user_data/Covid_19/inStrain/mapp...                True   \n",
       "2  /home/mattolm/user_data/Covid_19/inStrain/mapp...                True   \n",
       "3  /home/mattolm/user_data/Covid_19/inStrain/mapp...                True   \n",
       "4  /home/mattolm/user_data/Covid_19/inStrain/mapp...                True   \n",
       "\n",
       "   total_reads  paired_reads  percent_aligned  aligned_reads  \\\n",
       "0      5668472       5668472             0.03           1700   \n",
       "1      4407436       4407436             0.25          11018   \n",
       "2      2690410       2690410             3.20          86093   \n",
       "3     17121629      17121629             0.37          63350   \n",
       "4     14337950      14337950             0.87         124740   \n",
       "\n",
       "                                   inStrain_coverage  inStrain_succeeded  \n",
       "0  /home/mattolm/user_data/Covid_19/inStrain/prof...                True  \n",
       "1  /home/mattolm/user_data/Covid_19/inStrain/prof...                True  \n",
       "2  /home/mattolm/user_data/Covid_19/inStrain/prof...                True  \n",
       "3  /home/mattolm/user_data/Covid_19/inStrain/prof...                True  \n",
       "4  /home/mattolm/user_data/Covid_19/inStrain/prof...                True  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Full\n",
    "SRdb.to_csv('/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA_full_info.csv'.format(d4), index=False)\n",
    "SRdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>experiment_desc</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>instrument</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>total_reads</th>\n",
       "      <th>paired_reads</th>\n",
       "      <th>percent_aligned</th>\n",
       "      <th>aligned_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11454606</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>SRS6404546</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>Tianmen Center for Disease Control and Prevention</td>\n",
       "      <td>2020-02-08</td>\n",
       "      <td>Tianmen</td>\n",
       "      <td>5668472</td>\n",
       "      <td>5668472</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11454607</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>SRS6404545</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>Wuhan Lung Hospital</td>\n",
       "      <td>2020-02-07</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>4407436</td>\n",
       "      <td>4407436</td>\n",
       "      <td>0.25</td>\n",
       "      <td>11018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11454608</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>SRS6404544</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>Illumina iSeq 100</td>\n",
       "      <td>Jingzhou Center for Disease Control and Preven...</td>\n",
       "      <td>2020-01-08</td>\n",
       "      <td>Jingzhou</td>\n",
       "      <td>2690410</td>\n",
       "      <td>2690410</td>\n",
       "      <td>3.20</td>\n",
       "      <td>86093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11454609</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>SRS6404543</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>The Central Hospital Of Wuhan</td>\n",
       "      <td>2020-01-17</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>17121629</td>\n",
       "      <td>17121629</td>\n",
       "      <td>0.37</td>\n",
       "      <td>63350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR11454610</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>SRS6404542</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RANDOM PCR</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>Union Hospital of Tongji Medical College, Huaz...</td>\n",
       "      <td>2020-01-18</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>14337950</td>\n",
       "      <td>14337950</td>\n",
       "      <td>0.87</td>\n",
       "      <td>124740</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Run                                   experiment_title  \\\n",
       "0  SRR11454606  RNA-Seq of Homo sapiens: hCov-19 infected pati...   \n",
       "1  SRR11454607  RNA-Seq of Homo sapiens: hCov-19 infected pati...   \n",
       "2  SRR11454608  RNA-Seq of Homo sapiens: hCov-19 infected pati...   \n",
       "3  SRR11454609  RNA-Seq of Homo sapiens: hCov-19 infected pati...   \n",
       "4  SRR11454610  RNA-Seq of Homo sapiens: hCov-19 infected pati...   \n",
       "\n",
       "                                     experiment_desc sample_accession  \\\n",
       "0  RNA-Seq of Homo sapiens: hCov-19 infected pati...       SRS6404546   \n",
       "1  RNA-Seq of Homo sapiens: hCov-19 infected pati...       SRS6404545   \n",
       "2  RNA-Seq of Homo sapiens: hCov-19 infected pati...       SRS6404544   \n",
       "3  RNA-Seq of Homo sapiens: hCov-19 infected pati...       SRS6404543   \n",
       "4  RNA-Seq of Homo sapiens: hCov-19 infected pati...       SRS6404542   \n",
       "\n",
       "  library_strategy  library_source library_selection LibraryLayout  \\\n",
       "0          RNA-Seq  TRANSCRIPTOMIC        RANDOM PCR        SINGLE   \n",
       "1          RNA-Seq  TRANSCRIPTOMIC        RANDOM PCR        SINGLE   \n",
       "2          RNA-Seq  TRANSCRIPTOMIC        RANDOM PCR        SINGLE   \n",
       "3          RNA-Seq  TRANSCRIPTOMIC        RANDOM PCR        SINGLE   \n",
       "4          RNA-Seq  TRANSCRIPTOMIC        RANDOM PCR        SINGLE   \n",
       "\n",
       "          instrument                                       collected_by  \\\n",
       "0   Illumina MiniSeq  Tianmen Center for Disease Control and Prevention   \n",
       "1   Illumina MiniSeq                                Wuhan Lung Hospital   \n",
       "2  Illumina iSeq 100  Jingzhou Center for Disease Control and Preven...   \n",
       "3        NextSeq 550                      The Central Hospital Of Wuhan   \n",
       "4        NextSeq 550  Union Hospital of Tongji Medical College, Huaz...   \n",
       "\n",
       "  collection_date isolation_source  total_reads  paired_reads  \\\n",
       "0      2020-02-08          Tianmen      5668472       5668472   \n",
       "1      2020-02-07            Wuhan      4407436       4407436   \n",
       "2      2020-01-08         Jingzhou      2690410       2690410   \n",
       "3      2020-01-17            Wuhan     17121629      17121629   \n",
       "4      2020-01-18            Wuhan     14337950      14337950   \n",
       "\n",
       "   percent_aligned  aligned_reads  \n",
       "0             0.03           1700  \n",
       "1             0.25          11018  \n",
       "2             3.20          86093  \n",
       "3             0.37          63350  \n",
       "4             0.87         124740  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Basic\n",
    "PMdb = SRdb[['Run', 'experiment_title', 'experiment_desc', 'sample_accession', \n",
    "             'library_strategy', 'library_source', 'library_selection', 'LibraryLayout', \n",
    "             'instrument', 'collected_by', 'collection_date', 'isolation_source',\n",
    "             'total_reads',\t'paired_reads', 'percent_aligned', 'aligned_reads']]#, 'BioSample']]\n",
    "PMdb.to_csv('/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA_basic_info.csv'.format(d4), index=False)\n",
    "PMdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "59 of 159 samples have >=50 coverage\n",
      "60 of 159 samples have >=0.9 breadth\n",
      "50 of 159 samples have >=50 coverage and >=0.9 breadth\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Center_ID</th>\n",
       "      <th>Group</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Method</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>instrument</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>total_reads</th>\n",
       "      <th>percent_aligned</th>\n",
       "      <th>coverage</th>\n",
       "      <th>breadth</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>experiment_desc</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>Sufficient_cov</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11454606</td>\n",
       "      <td>Hubei_China</td>\n",
       "      <td>Hubei Provincial Center for Disease Control an...</td>\n",
       "      <td>PRJNA616446</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>5668472</td>\n",
       "      <td>0.03</td>\n",
       "      <td>6.168746</td>\n",
       "      <td>0.965522</td>\n",
       "      <td>SRS6404546</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>2020-02-08</td>\n",
       "      <td>Tianmen</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11454607</td>\n",
       "      <td>Hubei_China</td>\n",
       "      <td>Hubei Provincial Center for Disease Control an...</td>\n",
       "      <td>PRJNA616446</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>4407436</td>\n",
       "      <td>0.25</td>\n",
       "      <td>40.258502</td>\n",
       "      <td>0.997559</td>\n",
       "      <td>SRS6404545</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>2020-02-07</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11454608</td>\n",
       "      <td>Hubei_China</td>\n",
       "      <td>Hubei Provincial Center for Disease Control an...</td>\n",
       "      <td>PRJNA616446</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Illumina iSeq 100</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>2690410</td>\n",
       "      <td>3.20</td>\n",
       "      <td>364.855165</td>\n",
       "      <td>0.998428</td>\n",
       "      <td>SRS6404544</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>2020-01-08</td>\n",
       "      <td>Jingzhou</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11454609</td>\n",
       "      <td>Hubei_China</td>\n",
       "      <td>Hubei Provincial Center for Disease Control an...</td>\n",
       "      <td>PRJNA616446</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>17121629</td>\n",
       "      <td>0.37</td>\n",
       "      <td>139.966692</td>\n",
       "      <td>0.998763</td>\n",
       "      <td>SRS6404543</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>2020-01-17</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR11454610</td>\n",
       "      <td>Hubei_China</td>\n",
       "      <td>Hubei Provincial Center for Disease Control an...</td>\n",
       "      <td>PRJNA616446</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>14337950</td>\n",
       "      <td>0.87</td>\n",
       "      <td>284.849948</td>\n",
       "      <td>0.998729</td>\n",
       "      <td>SRS6404542</td>\n",
       "      <td>RNA-Seq of Homo sapiens: hCov-19 infected pati...</td>\n",
       "      <td>2020-01-18</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Run    Center_ID  \\\n",
       "0  SRR11454606  Hubei_China   \n",
       "1  SRR11454607  Hubei_China   \n",
       "2  SRR11454608  Hubei_China   \n",
       "3  SRR11454609  Hubei_China   \n",
       "4  SRR11454610  Hubei_China   \n",
       "\n",
       "                                               Group   BioProject   Method  \\\n",
       "0  Hubei Provincial Center for Disease Control an...  PRJNA616446  RNA-seq   \n",
       "1  Hubei Provincial Center for Disease Control an...  PRJNA616446  RNA-seq   \n",
       "2  Hubei Provincial Center for Disease Control an...  PRJNA616446  RNA-seq   \n",
       "3  Hubei Provincial Center for Disease Control an...  PRJNA616446  RNA-seq   \n",
       "4  Hubei Provincial Center for Disease Control an...  PRJNA616446  RNA-seq   \n",
       "\n",
       "  library_strategy  library_source         instrument LibraryLayout  \\\n",
       "0          RNA-Seq  TRANSCRIPTOMIC   Illumina MiniSeq        SINGLE   \n",
       "1          RNA-Seq  TRANSCRIPTOMIC   Illumina MiniSeq        SINGLE   \n",
       "2          RNA-Seq  TRANSCRIPTOMIC  Illumina iSeq 100        SINGLE   \n",
       "3          RNA-Seq  TRANSCRIPTOMIC        NextSeq 550        SINGLE   \n",
       "4          RNA-Seq  TRANSCRIPTOMIC        NextSeq 550        SINGLE   \n",
       "\n",
       "   total_reads  percent_aligned    coverage   breadth sample_accession  \\\n",
       "0      5668472             0.03    6.168746  0.965522       SRS6404546   \n",
       "1      4407436             0.25   40.258502  0.997559       SRS6404545   \n",
       "2      2690410             3.20  364.855165  0.998428       SRS6404544   \n",
       "3     17121629             0.37  139.966692  0.998763       SRS6404543   \n",
       "4     14337950             0.87  284.849948  0.998729       SRS6404542   \n",
       "\n",
       "                                     experiment_desc collection_date  \\\n",
       "0  RNA-Seq of Homo sapiens: hCov-19 infected pati...      2020-02-08   \n",
       "1  RNA-Seq of Homo sapiens: hCov-19 infected pati...      2020-02-07   \n",
       "2  RNA-Seq of Homo sapiens: hCov-19 infected pati...      2020-01-08   \n",
       "3  RNA-Seq of Homo sapiens: hCov-19 infected pati...      2020-01-17   \n",
       "4  RNA-Seq of Homo sapiens: hCov-19 infected pati...      2020-01-18   \n",
       "\n",
       "  isolation_source  Sufficient_cov  \n",
       "0          Tianmen           False  \n",
       "1            Wuhan           False  \n",
       "2         Jingzhou            True  \n",
       "3            Wuhan            True  \n",
       "4            Wuhan            True  "
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parsed\n",
    "\n",
    "PLdb = SRdb.copy()\n",
    "\n",
    "# Add from BioMeta\n",
    "PLdb = pd.merge(PLdb, BioMeta[['BioProject', 'Group', 'Method', 'Center_ID']], on='BioProject', how='left')\n",
    "\n",
    "VARIABLES = ['Center_ID', 'Group', 'BioProject', 'Method', 'library_strategy', 'library_source', 'instrument', 'LibraryLayout']\n",
    "for v in VARIABLES:\n",
    "    if len(PLdb[PLdb[v].isna()]) > 0:\n",
    "        print(\"{0} samples have no {1}\".format(len(PLdb[PLdb[v].isna()]), v))\n",
    "\n",
    "# Add basic instrain\n",
    "dbs = []\n",
    "for i, row in PLdb[PLdb['inStrain_succeeded'] == True].iterrows():\n",
    "    db = pd.read_csv(row['inStrain_coverage'], sep='\\t')\n",
    "    db['Run'] = row['Run']\n",
    "    dbs.append(db)\n",
    "COdb = pd.concat(dbs).reset_index(drop=True)\n",
    "\n",
    "PLdb = pd.merge(PLdb, COdb[['Run', 'coverage', 'breadth']], how='left', on='Run')\n",
    "assert len(PLdb[(PLdb['coverage'].isna()) & (PLdb['aligned_reads'] > 0)]) == 0\n",
    "\n",
    "# Filter\n",
    "DESC_VARS = ['sample_accession', 'experiment_desc', 'collection_date', 'isolation_source']\n",
    "INFO_VARS = ['total_reads', 'percent_aligned', 'coverage', 'breadth']\n",
    "PLdb = PLdb[['Run'] + VARIABLES + INFO_VARS + DESC_VARS]\n",
    "\n",
    "# Print a little description\n",
    "COV_LIM = 50\n",
    "BRE_LIM = 0.9\n",
    "\n",
    "fdb = PLdb[(PLdb['coverage'] >= COV_LIM)]\n",
    "print(\"{0} of {1} samples have >={2} coverage\".format(len(fdb), len(PLdb), COV_LIM, BRE_LIM))\n",
    "\n",
    "fdb = PLdb[(PLdb['breadth'] >= BRE_LIM)]\n",
    "print(\"{0} of {1} samples have >={3} breadth\".format(len(fdb), len(PLdb), COV_LIM, BRE_LIM))\n",
    "\n",
    "fdb = PLdb[(PLdb['coverage'] >= COV_LIM) & (PLdb['breadth'] >= BRE_LIM)]\n",
    "print(\"{0} of {1} samples have >={2} coverage and >={3} breadth\".format(len(fdb), len(PLdb), COV_LIM, BRE_LIM))\n",
    "fdb['Center_ID'].value_counts()\n",
    "\n",
    "PLdb['Sufficient_cov'] = [((coverage >= COV_LIM) & (breadth >= BRE_LIM)) for coverage, breadth in zip(PLdb['coverage'], PLdb['breadth'])]\n",
    "PLdb.to_csv('/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA_parsed_info.csv'.format(d4), index=False)\n",
    "PLdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    109\n",
       "True      50\n",
       "Name: Sufficient_cov, dtype: int64"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "PLdb['Sufficient_cov'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
