{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import scipy\n",
    "import sklearn\n",
    "import matplotlib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from collections import defaultdict\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set_style('whitegrid')\n",
    "pd.set_option('display.max_rows', 100)\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "pd.set_option('display.max_columns', 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATE = '02_04_2020'\n",
    "PLdb = pd.read_csv('/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA_full_info.csv'.format(DATE))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True    95\n",
       "Name: inStrain_succeeded, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PLdb = PLdb[PLdb['aligned_reads'] > 0]\n",
    "PLdb['inStrain_succeeded'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adjust metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "saveloc = '/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_tables/'.format(DATE)\n",
    "if not os.path.isdir(saveloc):\n",
    "    os.mkdir(saveloc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PLdb = pd.read_csv('/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA_full_info.csv'.format(DATE))\n",
    "\n",
    "# VARIABLES = ['library_strategy', 'library_source', 'instrument', 'collected_by']\n",
    "# for v in VARIABLES:\n",
    "#     print(\"{0} samples have no {1}\".format(len(PLdb[PLdb[v].isna()]), v))\n",
    "# print()\n",
    "    \n",
    "# # Add collected_by\n",
    "# PLdb['collected_by'] = ['NCBI (GEO)' if 'ACE2 library' in str(e) else c for e, c in zip(PLdb['experiment_title'], PLdb['collected_by'])]\n",
    "# for v in VARIABLES:\n",
    "#     print(\"{0} samples have no {1}\".format(len(PLdb[PLdb[v].isna()]), v))\n",
    "    \n",
    "# # Add center_ID\n",
    "# c2c = { 'Paragon Genomics':'Paragon_Genomics', \n",
    "#         'University of Washington Virology laboratory':'Univ_Washington_Vir',\n",
    "#         'Victorian Infectious Diseases Reference Laboratory (VIDRL)':'VIDRL',\n",
    "#         'Wuhan Jinyintan Hospital':\"Wuhan_Jinyintan\",\n",
    "#         \"Utah Public Health Laboratory\":'Utah_PublicHealth',\n",
    "#         'NCBI (GEO)':'NCBI (GEO)',\n",
    "#         'Institute of Pathogen Biology, Chinese Academy of Medical Sciences and Peking Union Medical College':'Peking_Beijing',\n",
    "#         'Wisconsin State Lab of Hygiene':'Wisconsin_StateLab',\n",
    "#         'BEI Resources/American Type and Culture Collection (ATCC)':'ATCC',\n",
    "#         'University of Washington Medical Center department of Laboratory Medicine, Virology':'Univ_Washington_Med',\n",
    "#         'Andersen Lab, The Scripps Research Institute':'Scripps',\n",
    "#         'National Influenza Centre, National Public health Laboratory, Kathmandu, Nepal':'Univ_tech_Pereira'\n",
    "#       }\n",
    "# PLdb['Center_ID'] = PLdb['collected_by'].map(c2c)\n",
    "# PLdb[PLdb['Center_ID'].isna()]['collected_by'].value_counts()\n",
    "\n",
    "# # Save\n",
    "# PLdb.to_csv('/home/mattolm/user_data/Covid_19/Pipeline/Jupyter/{0}_SRA_parsed_info.csv'.format(DATE), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and save coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>length</th>\n",
       "      <th>breadth</th>\n",
       "      <th>coverage</th>\n",
       "      <th>median_cov</th>\n",
       "      <th>std_cov</th>\n",
       "      <th>bases_w_0_coverage</th>\n",
       "      <th>mean_clonality</th>\n",
       "      <th>median_clonality</th>\n",
       "      <th>mean_microdiversity</th>\n",
       "      <th>median_microdiversity</th>\n",
       "      <th>rarefied_mean_microdiversity</th>\n",
       "      <th>rarefied_median_microdiversity</th>\n",
       "      <th>unmaskedBreadth</th>\n",
       "      <th>rarefied_breadth</th>\n",
       "      <th>expected_breadth</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>Reference_SNPs</th>\n",
       "      <th>BiAllelic_SNPs</th>\n",
       "      <th>MultiAllelic_SNPs</th>\n",
       "      <th>consensus_SNPs</th>\n",
       "      <th>population_SNPs</th>\n",
       "      <th>conANI</th>\n",
       "      <th>popANI</th>\n",
       "      <th>Run</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.965522</td>\n",
       "      <td>6.168746</td>\n",
       "      <td>5</td>\n",
       "      <td>4.292021</td>\n",
       "      <td>1031</td>\n",
       "      <td>0.997119</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.002881</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.617664</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.995691</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>0.999675</td>\n",
       "      <td>0.999729</td>\n",
       "      <td>SRR11454606</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.997559</td>\n",
       "      <td>40.258502</td>\n",
       "      <td>37</td>\n",
       "      <td>20.940226</td>\n",
       "      <td>73</td>\n",
       "      <td>0.996198</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.003802</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003515</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996622</td>\n",
       "      <td>0.289570</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>69</td>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>0.999832</td>\n",
       "      <td>0.999866</td>\n",
       "      <td>SRR11454607</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.998428</td>\n",
       "      <td>364.855165</td>\n",
       "      <td>265</td>\n",
       "      <td>337.567648</td>\n",
       "      <td>47</td>\n",
       "      <td>0.995800</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.004200</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004157</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.995887</td>\n",
       "      <td>0.971809</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>81</td>\n",
       "      <td>5</td>\n",
       "      <td>72</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.999832</td>\n",
       "      <td>0.999832</td>\n",
       "      <td>SRR11454608</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.998763</td>\n",
       "      <td>139.966692</td>\n",
       "      <td>115</td>\n",
       "      <td>114.470238</td>\n",
       "      <td>37</td>\n",
       "      <td>0.995933</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.004067</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.998261</td>\n",
       "      <td>0.869645</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>114</td>\n",
       "      <td>2</td>\n",
       "      <td>105</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>0.999766</td>\n",
       "      <td>0.999933</td>\n",
       "      <td>SRR11454609</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.998729</td>\n",
       "      <td>284.849948</td>\n",
       "      <td>248</td>\n",
       "      <td>183.518004</td>\n",
       "      <td>38</td>\n",
       "      <td>0.996902</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.003098</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002892</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.998428</td>\n",
       "      <td>0.976792</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>50</td>\n",
       "      <td>2</td>\n",
       "      <td>45</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.999933</td>\n",
       "      <td>0.999933</td>\n",
       "      <td>SRR11454610</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      scaffold  length   breadth    coverage  median_cov     std_cov  \\\n",
       "0  NC_045512.2   29903  0.965522    6.168746           5    4.292021   \n",
       "1  NC_045512.2   29903  0.997559   40.258502          37   20.940226   \n",
       "2  NC_045512.2   29903  0.998428  364.855165         265  337.567648   \n",
       "3  NC_045512.2   29903  0.998763  139.966692         115  114.470238   \n",
       "4  NC_045512.2   29903  0.998729  284.849948         248  183.518004   \n",
       "\n",
       "   bases_w_0_coverage  mean_clonality  median_clonality  mean_microdiversity  \\\n",
       "0                1031        0.997119               1.0             0.002881   \n",
       "1                  73        0.996198               1.0             0.003802   \n",
       "2                  47        0.995800               1.0             0.004200   \n",
       "3                  37        0.995933               1.0             0.004067   \n",
       "4                  38        0.996902               1.0             0.003098   \n",
       "\n",
       "   median_microdiversity  rarefied_mean_microdiversity  \\\n",
       "0                    0.0                           NaN   \n",
       "1                    0.0                      0.003515   \n",
       "2                    0.0                      0.004157   \n",
       "3                    0.0                      0.003715   \n",
       "4                    0.0                      0.002892   \n",
       "\n",
       "   rarefied_median_microdiversity  unmaskedBreadth  rarefied_breadth  \\\n",
       "0                             NaN         0.617664          0.000000   \n",
       "1                             0.0         0.996622          0.289570   \n",
       "2                             0.0         0.995887          0.971809   \n",
       "3                             0.0         0.998261          0.869645   \n",
       "4                             0.0         0.998428          0.976792   \n",
       "\n",
       "   expected_breadth  SNPs  Reference_SNPs  BiAllelic_SNPs  MultiAllelic_SNPs  \\\n",
       "0          0.995691    14               5               9                  0   \n",
       "1          1.000000    69               4              64                  1   \n",
       "2          1.000000    81               5              72                  4   \n",
       "3          1.000000   114               2             105                  7   \n",
       "4          1.000000    50               2              45                  3   \n",
       "\n",
       "   consensus_SNPs  population_SNPs    conANI    popANI          Run  \\\n",
       "0               6                5  0.999675  0.999729  SRR11454606   \n",
       "1               5                4  0.999832  0.999866  SRR11454607   \n",
       "2               5                5  0.999832  0.999832  SRR11454608   \n",
       "3               7                2  0.999766  0.999933  SRR11454609   \n",
       "4               2                2  0.999933  0.999933  SRR11454610   \n",
       "\n",
       "       genome  \n",
       "0  MT039887.1  \n",
       "1  MT039887.1  \n",
       "2  MT039887.1  \n",
       "3  MT039887.1  \n",
       "4  MT039887.1  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import inStrain\n",
    "import inStrain.SNVprofile\n",
    "\n",
    "dbs = []\n",
    "for i, row in PLdb[PLdb['inStrain_succeeded'] == True].iterrows():\n",
    "    db = pd.read_csv(row['inStrain_coverage'], sep='\\t')\n",
    "    db['Run'] = row['Run']\n",
    "    dbs.append(db)\n",
    "\n",
    "COdb = pd.concat(dbs).reset_index(drop=True)\n",
    "COdb['genome'] = 'MT039887.1'\n",
    "# for c in ['detected_scaffolds', 'true_scaffolds']:\n",
    "#     del COdb[c]\n",
    "    \n",
    "COdb.to_csv(saveloc + 'COVID_genome_coverage.csv', index=False)\n",
    "COdb.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>length</th>\n",
       "      <th>breadth</th>\n",
       "      <th>coverage</th>\n",
       "      <th>median_cov</th>\n",
       "      <th>std_cov</th>\n",
       "      <th>bases_w_0_coverage</th>\n",
       "      <th>mean_clonality</th>\n",
       "      <th>median_clonality</th>\n",
       "      <th>mean_microdiversity</th>\n",
       "      <th>median_microdiversity</th>\n",
       "      <th>rarefied_mean_microdiversity</th>\n",
       "      <th>rarefied_median_microdiversity</th>\n",
       "      <th>unmaskedBreadth</th>\n",
       "      <th>rarefied_breadth</th>\n",
       "      <th>expected_breadth</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>Reference_SNPs</th>\n",
       "      <th>BiAllelic_SNPs</th>\n",
       "      <th>MultiAllelic_SNPs</th>\n",
       "      <th>consensus_SNPs</th>\n",
       "      <th>population_SNPs</th>\n",
       "      <th>conANI</th>\n",
       "      <th>popANI</th>\n",
       "      <th>Run</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.965522</td>\n",
       "      <td>6.168746</td>\n",
       "      <td>5</td>\n",
       "      <td>4.292021</td>\n",
       "      <td>1031</td>\n",
       "      <td>0.997119</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.002881</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.617664</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.995691</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>0.999675</td>\n",
       "      <td>0.999729</td>\n",
       "      <td>SRR11454606</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.997559</td>\n",
       "      <td>40.258502</td>\n",
       "      <td>37</td>\n",
       "      <td>20.940226</td>\n",
       "      <td>73</td>\n",
       "      <td>0.996198</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.003802</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003515</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.996622</td>\n",
       "      <td>0.289570</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>69</td>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>0.999832</td>\n",
       "      <td>0.999866</td>\n",
       "      <td>SRR11454607</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.998428</td>\n",
       "      <td>364.855165</td>\n",
       "      <td>265</td>\n",
       "      <td>337.567648</td>\n",
       "      <td>47</td>\n",
       "      <td>0.995800</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.004200</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004157</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.995887</td>\n",
       "      <td>0.971809</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>81</td>\n",
       "      <td>5</td>\n",
       "      <td>72</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.999832</td>\n",
       "      <td>0.999832</td>\n",
       "      <td>SRR11454608</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.998763</td>\n",
       "      <td>139.966692</td>\n",
       "      <td>115</td>\n",
       "      <td>114.470238</td>\n",
       "      <td>37</td>\n",
       "      <td>0.995933</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.004067</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.998261</td>\n",
       "      <td>0.869645</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>114</td>\n",
       "      <td>2</td>\n",
       "      <td>105</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>0.999766</td>\n",
       "      <td>0.999933</td>\n",
       "      <td>SRR11454609</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.998729</td>\n",
       "      <td>284.849948</td>\n",
       "      <td>248</td>\n",
       "      <td>183.518004</td>\n",
       "      <td>38</td>\n",
       "      <td>0.996902</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.003098</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002892</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.998428</td>\n",
       "      <td>0.976792</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>50</td>\n",
       "      <td>2</td>\n",
       "      <td>45</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.999933</td>\n",
       "      <td>0.999933</td>\n",
       "      <td>SRR11454610</td>\n",
       "      <td>MT039887.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      scaffold  length   breadth    coverage  median_cov     std_cov  \\\n",
       "0  NC_045512.2   29903  0.965522    6.168746           5    4.292021   \n",
       "1  NC_045512.2   29903  0.997559   40.258502          37   20.940226   \n",
       "2  NC_045512.2   29903  0.998428  364.855165         265  337.567648   \n",
       "3  NC_045512.2   29903  0.998763  139.966692         115  114.470238   \n",
       "4  NC_045512.2   29903  0.998729  284.849948         248  183.518004   \n",
       "\n",
       "   bases_w_0_coverage  mean_clonality  median_clonality  mean_microdiversity  \\\n",
       "0                1031        0.997119               1.0             0.002881   \n",
       "1                  73        0.996198               1.0             0.003802   \n",
       "2                  47        0.995800               1.0             0.004200   \n",
       "3                  37        0.995933               1.0             0.004067   \n",
       "4                  38        0.996902               1.0             0.003098   \n",
       "\n",
       "   median_microdiversity  rarefied_mean_microdiversity  \\\n",
       "0                    0.0                           NaN   \n",
       "1                    0.0                      0.003515   \n",
       "2                    0.0                      0.004157   \n",
       "3                    0.0                      0.003715   \n",
       "4                    0.0                      0.002892   \n",
       "\n",
       "   rarefied_median_microdiversity  unmaskedBreadth  rarefied_breadth  \\\n",
       "0                             NaN         0.617664          0.000000   \n",
       "1                             0.0         0.996622          0.289570   \n",
       "2                             0.0         0.995887          0.971809   \n",
       "3                             0.0         0.998261          0.869645   \n",
       "4                             0.0         0.998428          0.976792   \n",
       "\n",
       "   expected_breadth  SNPs  Reference_SNPs  BiAllelic_SNPs  MultiAllelic_SNPs  \\\n",
       "0          0.995691    14               5               9                  0   \n",
       "1          1.000000    69               4              64                  1   \n",
       "2          1.000000    81               5              72                  4   \n",
       "3          1.000000   114               2             105                  7   \n",
       "4          1.000000    50               2              45                  3   \n",
       "\n",
       "   consensus_SNPs  population_SNPs    conANI    popANI          Run  \\\n",
       "0               6                5  0.999675  0.999729  SRR11454606   \n",
       "1               5                4  0.999832  0.999866  SRR11454607   \n",
       "2               5                5  0.999832  0.999832  SRR11454608   \n",
       "3               7                2  0.999766  0.999933  SRR11454609   \n",
       "4               2                2  0.999933  0.999933  SRR11454610   \n",
       "\n",
       "       genome  \n",
       "0  MT039887.1  \n",
       "1  MT039887.1  \n",
       "2  MT039887.1  \n",
       "3  MT039887.1  \n",
       "4  MT039887.1  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import inStrain\n",
    "import inStrain.SNVprofile\n",
    "\n",
    "dbs = []\n",
    "for i, row in PLdb[PLdb['inStrain_succeeded'] == True].iterrows():\n",
    "    loc = row['inStrain_coverage'].replace('_genomeWide_scaffold_info.tsv', '_SNVs.tsv')\n",
    "    try:\n",
    "        db = pd.read_csv(loc, sep='\\t')\n",
    "        db['Run'] = row['Run']\n",
    "        dbs.append(db)\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "\n",
    "Sdb = pd.concat(dbs).reset_index(drop=True)\n",
    "Sdb['genome'] = 'MT039887.1'\n",
    "\n",
    "Sdb.to_csv(saveloc + 'COVID_raw_SNVs.csv', index=False)\n",
    "Sdb.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and save positional coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>position</th>\n",
       "      <th>coverage</th>\n",
       "      <th>normalized_coverage</th>\n",
       "      <th>microdiversity</th>\n",
       "      <th>normalized_microdiversity</th>\n",
       "      <th>rarefied_microdiversity</th>\n",
       "      <th>normalized_rarefied_microdiversity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11454606</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11454606</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11454606</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11454606</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR11454606</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Run  position  coverage  normalized_coverage  microdiversity  \\\n",
       "0  SRR11454606         0         0                  0.0             NaN   \n",
       "1  SRR11454606         1         0                  0.0             NaN   \n",
       "2  SRR11454606         2         0                  0.0             NaN   \n",
       "3  SRR11454606         3         0                  0.0             NaN   \n",
       "4  SRR11454606         4         0                  0.0             NaN   \n",
       "\n",
       "   normalized_microdiversity  rarefied_microdiversity  \\\n",
       "0                        NaN                      NaN   \n",
       "1                        NaN                      NaN   \n",
       "2                        NaN                      NaN   \n",
       "3                        NaN                      NaN   \n",
       "4                        NaN                      NaN   \n",
       "\n",
       "   normalized_rarefied_microdiversity  \n",
       "0                                 NaN  \n",
       "1                                 NaN  \n",
       "2                                 NaN  \n",
       "3                                 NaN  \n",
       "4                                 NaN  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TRUE_LENGTH = 29903\n",
    "SCAFFOLD = 'NC_045512.2'\n",
    "bldb = pd.DataFrame(list(range(0, TRUE_LENGTH))).rename(columns={0:'position'})\n",
    "\n",
    "dbs = []\n",
    "for i, row in PLdb[PLdb['inStrain_succeeded'] == True].iterrows():\n",
    "    IS_loc = '/'.join(row['inStrain_coverage'].split('/')[:-2])\n",
    "    IS = inStrain.SNVprofile.SNVprofile(IS_loc)\n",
    "    \n",
    "    # Load positional coverage\n",
    "    covT = IS.get('covT')\n",
    "    cov = inStrain.profileUtilities._mm_counts_to_counts_shrunk(covT[SCAFFOLD])\n",
    "    db = pd.DataFrame(cov).reset_index(drop=False).rename(columns={'index':'position', 0:'coverage'})\n",
    "    db['coverage'] = db['coverage'].astype(int)\n",
    "    \n",
    "    # Load positional microdiversity\n",
    "    covT = IS.get('clonT')\n",
    "    cov = inStrain.plottingUtilities._get_basewise_clons3(covT[SCAFFOLD], 100)\n",
    "    cdb = cov.reset_index(drop=False).rename(columns={'index':'position', 0:'clonality'})\n",
    "    cdb['clonality'] = cdb['clonality'].astype(float)\n",
    "    cdb['microdiversity'] = 1 - cdb['clonality']\n",
    "    cdb['normalized_microdiversity'] = sklearn.preprocessing.normalize(cdb['microdiversity'].values.reshape(-1, 1), axis=0) # Normalize clonality before NaN values are introduced\n",
    "    \n",
    "    # Load rarefied positional microdiversity\n",
    "    covTR = IS.get('clonTR')\n",
    "    cov = inStrain.plottingUtilities._get_basewise_clons3(covTR[SCAFFOLD], 100)\n",
    "    crdb = cov.reset_index(drop=False).rename(columns={'index':'position', 0:'rarefied_clonality'})\n",
    "    crdb['rarefied_clonality'] = crdb['rarefied_clonality'].astype(float)\n",
    "    crdb['rarefied_microdiversity'] = 1 - crdb['rarefied_clonality']\n",
    "    try:\n",
    "        crdb['normalized_rarefied_microdiversity'] = sklearn.preprocessing.normalize(crdb['rarefied_microdiversity'].values.reshape(-1, 1), axis=0) # Normalize clonality before NaN values are introduced\n",
    "    except ValueError:\n",
    "        crdb['normalized_rarefied_microdiversity'] = np.nan\n",
    "    \n",
    "    # Merge\n",
    "    db = pd.merge(bldb, db, on='position', how='outer')\n",
    "    db = pd.merge(db, cdb, on='position', how='outer')\n",
    "    db = pd.merge(db, crdb, on='position', how='outer')\n",
    "    db['Run'] = row['Run']\n",
    "    db = db.sort_values('position')\n",
    "    \n",
    "    # Backfill coverage with 0s and normalize\n",
    "    db['coverage'] = db['coverage'].fillna(0).astype(int)\n",
    "    db['normalized_coverage'] = sklearn.preprocessing.normalize(db['coverage'].values.reshape(-1, 1), axis=0)\n",
    "    \n",
    "    # Add mean and adjust\n",
    "#     db['mean_coverage'] = db['coverage'].mean()\n",
    "#     db['adjusted_coverage'] = db['coverage'] / db['mean_coverage']\n",
    "    \n",
    "#     db['mean_microdiversity'] = db['microdiversity'].mean()\n",
    "#     db['adjusted_microdiversity'] = db['coverage'] / db['mean_microdiversity']\n",
    "    \n",
    "#     db['mean_rarefied_microdiversity'] = db['rarefied_microdiversity'].mean()\n",
    "#     db['adjusted_rarefied_microdiversity'] = db['coverage'] / db['mean_rarefied_microdiversity']\n",
    "    \n",
    "    dbs.append(db)\n",
    "    \n",
    "# Re-order columns\n",
    "Odb = pd.concat(dbs).reset_index(drop=True)\n",
    "order = ['Run', 'position']\n",
    "for item in ['coverage', 'microdiversity', 'rarefied_microdiversity']:\n",
    "    order += [item, 'normalized_' + item]\n",
    "Odb = Odb[order]\n",
    "\n",
    "# Save\n",
    "Odb.to_csv(saveloc + 'COVID_positional_coverage_v2.csv.gz', index=False)\n",
    "Odb.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene-level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>gene</th>\n",
       "      <th>coverage</th>\n",
       "      <th>normalized_microdiversity</th>\n",
       "      <th>normalized_microdiversity_rank</th>\n",
       "      <th>coverage_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11059940</td>\n",
       "      <td>E</td>\n",
       "      <td>1.921053</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11059940</td>\n",
       "      <td>M</td>\n",
       "      <td>21.856502</td>\n",
       "      <td>0.000929</td>\n",
       "      <td>7.0</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11059940</td>\n",
       "      <td>N</td>\n",
       "      <td>35.115873</td>\n",
       "      <td>0.001697</td>\n",
       "      <td>17.0</td>\n",
       "      <td>29.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11059940</td>\n",
       "      <td>ORF10</td>\n",
       "      <td>17.744361</td>\n",
       "      <td>0.002567</td>\n",
       "      <td>26.0</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR11059940</td>\n",
       "      <td>ORF3a</td>\n",
       "      <td>9.589372</td>\n",
       "      <td>0.000642</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Run   gene   coverage  normalized_microdiversity  \\\n",
       "0  SRR11059940      E   1.921053                   0.000000   \n",
       "1  SRR11059940      M  21.856502                   0.000929   \n",
       "2  SRR11059940      N  35.115873                   0.001697   \n",
       "3  SRR11059940  ORF10  17.744361                   0.002567   \n",
       "4  SRR11059940  ORF3a   9.589372                   0.000642   \n",
       "\n",
       "   normalized_microdiversity_rank  coverage_rank  \n",
       "0                             1.0            2.0  \n",
       "1                             7.0           25.0  \n",
       "2                            17.0           29.0  \n",
       "3                            26.0           21.0  \n",
       "4                             4.0            8.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ON = 'normalized_microdiversity'\n",
    "MIN_COV = 0\n",
    "\n",
    "POdb = pd.read_csv('/home/mattolm/user_data/Covid_19/datatables/COVID_genes_positional_v2.2.csv')\n",
    "\n",
    "fdb = pd.merge(Odb, POdb, on='position', how='left').merge(COdb, how='left', on='Run')\n",
    "fdb = fdb[fdb['coverage_y'] >= MIN_COV]\n",
    "\n",
    "table = defaultdict(list)\n",
    "for run, rdb in fdb.groupby('Run'):\n",
    "    for gene, db in rdb.groupby('protein'):\n",
    "        table['Run'].append(run)\n",
    "        table['gene'].append(gene)\n",
    "        table['coverage'].append(db['coverage_x'].mean())\n",
    "        table[ON].append(db[ON].mean())\n",
    "        \n",
    "    for gene, db in rdb.groupby('gene'):\n",
    "        table['Run'].append(run)\n",
    "        table['gene'].append(gene)\n",
    "        table['coverage'].append(db['coverage_x'].mean())\n",
    "        table[ON].append(db[ON].mean())\n",
    "\n",
    "run = 'all'\n",
    "rdb = fdb\n",
    "for gene, db in rdb.groupby('protein'):\n",
    "    table['Run'].append(run)\n",
    "    table['gene'].append(gene)\n",
    "    table['coverage'].append(db['coverage_x'].mean())\n",
    "    table[ON].append(db[ON].mean())\n",
    "\n",
    "for gene, db in rdb.groupby('gene'):\n",
    "    table['Run'].append(run)\n",
    "    table['gene'].append(gene)\n",
    "    table['coverage'].append(db['coverage_x'].mean())\n",
    "    table[ON].append(db[ON].mean())\n",
    "    \n",
    "MGdb = pd.DataFrame(table)\n",
    "\n",
    "# Add rank order\n",
    "dbs = defaultdict(list)\n",
    "ITEMS = [ON, 'coverage']\n",
    "for run, db in MGdb.groupby('Run'):\n",
    "    for item in ITEMS:\n",
    "        db['{0}_rank'.format(item)] = db[item].rank(method='min')#.astype(int)\n",
    "        dbs[item].append(db)\n",
    "for item in ITEMS:\n",
    "    db = pd.concat(dbs[item]).reset_index(drop=True)\n",
    "    MGdb = pd.merge(MGdb, db, how='left', on=list(MGdb.columns))\n",
    "    \n",
    "# Save\n",
    "MGdb.to_csv(saveloc + 'COVID_gene_coverage_v2.csv', index=False)\n",
    "MGdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
